---
title: "cis-meQTLs Analysis"
subtitle: "Dexamethasone-Stimulated Human Array Project"
author: 
  - name: Anastasiia Hryhorzhevska
    email: anastasiia_hry@psych.mpg.de
    url: https://github.com/ahryho
    affiliation: Max Planck Institute of Psychiatry
    affiliation_url: https://www.psych.mpg.de/2664393/medizinische-genomforschung
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    css: ../style.css
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
options(digits = 4, width = 100, stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE,
                      tidy.opts = list(width.cutoff = 100),
                      tidy=TRUE,
                      fig.pos = "H",
                      dpi = 200,
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE, 
                      cache.lazy = FALSE)

# define DT as the standard printing method for data.frame
library(knitr)
library(DT)
knit_print.data.frame <- function(x, ...) {
  knit_print(DT::datatable(x,
                           filter="top",
                           rownames = FALSE,
                           extensions = "FixedColumns",
                           options = list(
                             scrollX = TRUE,
                             pageLength = 5
                             
                           )), ...)
}

registerS3method("knit_print", "data.frame", knit_print.data.frame)
```

# __Functional CpG annotation__
***

```{r, include = F, eval = T}
source("~/bio/code/mpip/dex-stim-human-array/code/util.R", chdir = TRUE)

library(data.table)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(scales)
library(factoextra)
library(viridis)
library(knitr)

library(splitstackshape)
library(reshape2)
library(IRanges)
library(ggplot2)
library(biomaRt)

require(foreign)
library(parallel)
library(foreach)
library(doParallel)

library(gUtils)

out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/"

# meqtl.dex.fn   <- paste0(out.dir.pre, "me-qtl_cis_indp_result_dex_with_delta_fdr_005.csv")
# meqtl.veh.fn   <- paste0(out.dir.pre, "me-qtl_cis_indp_result_veh_fdr_005.csv")
# meqtl.delta.fn <- paste0(out.dir.pre, "me-qtl_cis_indp_result_delta_fdr_005.csv")

# col.names <-  c("SNP", "CpG_ID", "beta", "t-stat", "p-value", "fdr")

# ind.meqtl.dex.df   <- fread(meqtl.dex.fn, col.names = col.names)
# ind.meqtl.veh.df   <- fread(meqtl.veh.fn, col.names = col.names)
# ind.meqtl.delta.df <- fread(meqtl.delta.fn, col.names = col.names)
```

```{r include = F, eval = F}
meqtls.cpg.anno.df <- fread(paste0(out.dir.pre, "meqtls_cpg_annotated_delta.csv"), sep = "\t")

# Prepare genes' ranges

# Split gene names 
tmp.df <- meqtls.cpg.anno.df[, .(Name, UCSC_RefGene_Name)]
cpg.gene.df <- cSplit(indt = tmp.df, splitCols = 'UCSC_RefGene_Name', sep = ';', type.convert = F, drop = F) %>%
    melt(id.vars = c("Name", "UCSC_RefGene_Name"), value.name = "GeneSymbol") %>%
    dplyr::select(Name, GeneSymbol) 
  
# Split gene group column
tmp.df <- meqtls.cpg.anno.df[, .(Name, UCSC_RefGene_Group)]
cpg.gene.group.df <- cSplit(indt = tmp.df, splitCols = 'UCSC_RefGene_Group', sep = ';', type.convert = F, drop = F) %>%
    melt(id.vars = c("Name", "UCSC_RefGene_Group"), value.name = "GeneGroup") %>%
    dplyr::select(Name, GeneGroup) 
  
cpg.gene.df <- cbind(cpg.gene.df, cpg.gene.group.df[, "GeneGroup"]) %>% 
    unique() # %>% na.omit()

colnames(cpg.gene.df) <- c("CpG_ID", "GeneSymbol", "GeneGroup")

chr.list <- meqtls.cpg.anno.df$chr %>% unique() %>% sub("chr", "", .)

# Get Genes' coordinates and create IRanges
gene.list <- unique(na.omit(cpg.gene.df$GeneSymbol))

ensembl       <- biomaRt::useEnsembl('ensembl', dataset = 'hsapiens_gene_ensembl', GRCh = 37)
gene.coord.df <- getBM(attributes = c('hgnc_symbol', 'chromosome_name', 'start_position', 'end_position'),
                         filters = 'hgnc_symbol',
                         values = gene.list,
                         mart = ensembl) %>% unique()
  
gene.coord.df <- gene.coord.df[gene.coord.df$chromosome_name %in% chr.list,]
colnames(gene.coord.df) <- c("GeneSymbol", "GeneChr", "GeneStartPos", "GeneEndPos")

gene.coord.range <- makeGRangesFromDataFrame(gene.coord.df, 
                                             start.field = "GeneStartPos", 
                                             end.field = "GeneEndPos", 
                                             seqnames.field = c("GeneChr"))
names(gene.coord.range) <- gene.coord.df$GeneSymbol


# Prepare methyl ranges

cpg.coord.df    <- meqtls.cpg.anno.df[, .(Name, chr, pos)] %>% dplyr:: mutate(chr = sub("chr", "", chr))
cpg.coord.range <- makeGRangesFromDataFrame(cpg.coord.df, 
                                            start.field = "pos", 
                                            end.field = "pos", 
                                            seqnames.field = c("chr"))
names(cpg.coord.range) <- cpg.coord.df$Name

# Calculate distances

chr.list     <- cpg.coord.df$chr %>% unique()
registerDoParallel(cores = 50)

distances <- foreach(chr = chr.list, .combine = rbind, .packages = c('gUtils', 'dplyr')) %dopar% {
  
  r1 <- cpg.coord.range[cpg.coord.range@seqnames == chr] # GRanges object with 740357 ranges and 0 metadata columns
  r2 <- gene.coord.range[gene.coord.range@seqnames == chr] # GRanges object with 12841 ranges and 0 metadata columns
  
  dist.to.all <- gr.dist(r1, r2)
  rownames(dist.to.all) <- names(r1)
  colnames(dist.to.all) <- names(r2)
  
  dist.to.all.melt <- dist.to.all %>% 
    reshape2::melt() %>% 
    na.omit() %>% 
    unique()
  colnames(dist.to.all.melt) <- c("CpG_ID", "Gene_Symbol", "CpG_Gene_DIST")
  
  return(dist.to.all.melt)
}

stopImplicitCluster()

colnames(distances) <- c("PROBE_ID", "ILMN_ID", "CG_GENE_DIST")

cpg.gene.coord.df <- distances %>% setDT()

cpg.gene.coord.df <- left_join(cpg.gene.coord.df, cpg.coord.df, by = "PROBE_ID")
cpg.gene.coord.df <- left_join(cpg.gene.coord.df, unique(gene.coord.df), by = c("ILMN_ID" = "GeneSymbol"))
cpg.gene.coord.df <- cpg.gene.coord.df %>% dplyr::select(PROBE_ID, CG_CHR = chr, CG_POS = pos,
                                                  ILMN_ID, GENE_CHR = GeneChr, GENE_START_POS = GeneStartPos, GENE_END_POS = GeneEndPos,
                                                  CG_GENE_DIST)

  
```

```{r include = F, eval = F}
GetDistances <- function(dmps.anno.df, ensembl){
  
  # Split Gene Name and Gene Group columns 
  library(splitstackshape)
  tmp.df <- dmps.anno.df[, .(PROBE_ID, UCSC_RefGene_Name)]
  cpg.gene.df <- cSplit(indt = tmp.df, splitCols = 'UCSC_RefGene_Name', sep = ';', type.convert = F, drop = F) %>%
    melt(id.vars = c("PROBE_ID", "UCSC_RefGene_Name"), value.name = "GeneSymbol") %>%
    dplyr::select(PROBE_ID, GeneSymbol) 
  
  # Split gene group column
  tmp.df <- dmps.anno.df[, .(PROBE_ID, UCSC_RefGene_Group)]
  cpg.gene.group.df <- cSplit(indt = tmp.df, splitCols = 'UCSC_RefGene_Group', sep = ';', type.convert = F, drop = F) %>%
    melt(id.vars = c("PROBE_ID", "UCSC_RefGene_Group"), value.name = "GeneGroup") %>%
    dplyr::select(PROBE_ID, GeneGroup) 
  
  cpg.gene.df <- cbind(cpg.gene.df, cpg.gene.group.df[, "GeneGroup"]) %>% 
    unique() # %>% na.omit()
  colnames(cpg.gene.df) <- c("PROBE_ID", "GeneSymbol", "GeneGroup")


  # Get Genes' coordinates and create IRanges
  gene.list <- unique(na.omit(cpg.gene.df$GeneSymbol))
  gene.coord.df <- getBM(attributes = c('hgnc_symbol', 'chromosome_name', 'start_position', 'end_position'),
                         filters = 'hgnc_symbol',
                         values = gene.list,
                         mart = ensembl) %>%
    unique()

  chr.list <- dmps.anno.df$chr %>% unique() %>% sub("chr", "", .)
  gene.coord.df <- gene.coord.df[gene.coord.df$chromosome_name %in% chr.list,]
  colnames(gene.coord.df) <- c("GeneSymbol", "GeneChr", "GeneStartPos", "GeneEndPos")
  
  # CpG Gene Cord DF
  cpg.coord.df <- dmps.anno.df[, .(PROBE_ID, chr, pos)] %>% dplyr:: mutate(chr = sub("chr", "", chr))
  
  cpg.gene.coord.df <- left_join(cpg.gene.df, cpg.coord.df, by = "PROBE_ID")
 
    cpg.gene.coord.df <- left_join(cpg.gene.coord.df, gene.coord.df, by = "GeneSymbol")
    cpg.gene.coord.df[["CG_GENE_DIST"]] <- ifelse(cpg.gene.coord.df$GeneStartPos > cpg.gene.coord.df$pos,
                                                  cpg.gene.coord.df$GeneStartPos - cpg.gene.coord.df$pos,
                                                  ifelse(cpg.gene.coord.df$pos > cpg.gene.coord.df$GeneEndPos,
                                                       cpg.gene.coord.df$pos - cpg.gene.coord.df$GeneEndPos - 1,
                                                       0))
  
  #   cpg.gene.coord.df[["CG_GENE_DIST"]] <- lapply(cpg.gene.coord.df, function(x)
  #   #ifelse((x["pos"] > x["GeneStartPos"] ) & (x["pos"] < x["GeneEndPos"]),
  #   #       0,
  #          min(abs(x["pos"] - x["GeneStartPos"], abs(x["pos"] - x["GeneEndPos"])))
  # )
    
    
    # ifelse((cpg.gene.coord.df$pos > cpg.gene.coord.df$GeneStartPos ) &
    #                                               (cpg.gene.coord.df$pos < cpg.gene.coord.df$GeneEndPos),
    #                                             0,
    #                                             min(abs(cpg.gene.coord.df$pos - cpg.gene.coord.df$GeneStartPos),
    #                                                 abs(cpg.gene.coord.df$pos - cpg.gene.coord.df$GeneEndPos)))

    cpg.gene.coord.df <- cpg.gene.coord.df %>% 
      dplyr::select( PROBE_ID,  chr, pos, GeneSymbol, GeneChr, GeneStartPos, GeneEndPos, GeneGroup, CG_GENE_DIST)
  
    cpg.gene.coord.df <- cpg.gene.coord.df[!is.na(cpg.gene.coord.df$CG_GENE_DIST),] # %>% setDT()
  
  cpg.na.gene.df    <- tmp.df[UCSC_RefGene_Group == "", "PROBE_ID"] %>% unique()
  
  cpg.na.gene.df[["chr"]]   <- NA
  cpg.na.gene.df[["pos"]]   <- NA
  cpg.na.gene.df[["GeneSymbol"]]   <- NA
  cpg.na.gene.df[["GeneChr"]]   <- NA
  cpg.na.gene.df[["GeneStartPos"]]   <- NA
  cpg.na.gene.df[["GeneEndPos"]]   <- NA
  cpg.na.gene.df[["GeneGroup"]]    <- NA
  cpg.na.gene.df[["CG_GENE_DIST"]] <- NA
  
  cpg.gene.coord.df <- rbind(cpg.gene.coord.df, cpg.na.gene.df)
  
  return(cpg.gene.coord.df)
}

meqtls.cpg.anno.df <- fread(paste0(out.dir.pre, "meqtls_cpg_annotated_delta.csv"), sep = "\t")
colnames(meqtls.cpg.anno.df)[4] <- "PROBE_ID"

chr.list <- meqtls.cpg.anno.df$chr %>% unique()# %>% sub("chr", "", .)
ensembl  <- biomaRt::useEnsembl('ensembl', dataset = 'hsapiens_gene_ensembl', GRCh = 37)

registerDoParallel(cores = 50)

# distances <- foreach(chrom = chr.list[1], .combine = rbind, .packages = c('dplyr', 'reshape2', 'biomaRt')) %dopar% {
distances <- foreach(chrom = chr.list, .combine = rbind, .packages = c('dplyr', 'reshape2', 'biomaRt')) %dopar% {
  df <- meqtls.cpg.anno.df[meqtls.cpg.anno.df$chr == chrom,]
  GetDistances(df, ensembl)
}

stopImplicitCluster()

cpg.gene.coord.df <- distances %>% setDT()

cpg.gene.coord.df <- fread('~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/ind_me-qtl_cis/delta/annotation/me-qtl_delta_cpgs_annotated_gene_distances.csv')

# write.csv2(cpg.gene.coord.df,
#            '~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/ind_me-qtl_cis/delta/annotation/me-qtl_delta_cpgs_annotated_gene_distances.csv',
#            quote = F, row.names = F)

cpg.na.gene.df   <- cpg.gene.coord.df[is.na(CG_GENE_DIST), .(PROBE_ID, GeneSymbol, GeneGroup, CG_GENE_DIST)]
cpg.gene.dist.df <- cpg.gene.coord.df[, .(PROBE_ID, GeneSymbol, CG_GENE_DIST)]
plt.df           <- cpg.gene.dist.df[ , .SD[which.min(CG_GENE_DIST)], by = PROBE_ID, ]

# Check
plt.df[PROBE_ID == "cg08975523"]
cpg.gene.coord.df[PROBE_ID == "cg08975523"]

plt.df[PROBE_ID == "cg06562291"]
cpg.gene.coord.df[PROBE_ID == "cg06562291"]
```

```{r load-cpg-gene-distances, include = F, eval = F}

# DataFrame with closest annotated genes, all columns

plt.df3          <- left_join(plt.df, cpg.gene.coord.df)
plt.df3          <- rbind(plt.df3, cpg.gene.coord.df[is.na(CG_GENE_DIST), ])

plt.df3[PROBE_ID == "cg06562291"]
plt.df[PROBE_ID == "cg06562291"]
cpg.gene.coord.df[PROBE_ID == "cg06562291"]

write.csv2(plt.df3,
           '~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/ind_me-qtl_cis/delta/annotation/me-qtl_delta_cpgs_annotated_closest_genes_distances.csv',
           quote = F, row.names = F)
```

```{r include = F, eval = T}
# code: ../code/methylation/02_dma/01_dmp/03_get_cpg_gene_anno_dist
cpg.closest.genes.dist.fn <-"~/bio/code/mpip/dex-stim-human-array/output/data/methylation/02_dmp/dex_cpgs_annotated_closest_genes_distances.csv"
cpg.closest.genes.dist.df <- fread(cpg.closest.genes.dist.fn)

delta.meqtl.cpg.anno.df               <- fread(paste0(out.dir.pre, "meqtls_cpg_annotated_delta.csv"), sep = "\t")
delta.meqtl.cpg.closest.genes.dist.df <- cpg.closest.genes.dist.df[PROBE_ID %in% delta.meqtl.cpg.anno.df$Name]

# delta.meqtl.cpg.closest.genes.dist.fn <-"~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/ind_me-qtl_cis/delta/annotation/me-qtl_delta_cpgs_annotated_closest_genes_distances.csv"
# delta.meqtl.cpg.closest.genes.dist.df <- fread(cpg.closest.genes.dist.fn)
```

## Illumina {.tabset .tabset-fade .tabset-pills}

### Chromosomes

```{r out.width = "90%", fig.cap = "CpGs distribution across chromosomes", include = T, eval = T}
library(GenomicFeatures)
chr.length.tbl <- getChromInfoFromUCSC("hg19")[, c(1, 2)] %>% setDT()
chr.length.tbl <- chr.length.tbl[chrom %in% paste0("chr", 1:22)] #, "X", "Y"))]
colnames(chr.length.tbl) <- c("chr", "chr_size")
chr.length.tbl$chr <- as.factor(chr.length.tbl$chr)

chr.order <- paste("chr", c(1:22), sep = "")
meth      <- delta.meqtl.cpg.anno.df
meth$chr  <- factor(meth$chr,levels = chr.order)
meth      <- left_join(meth, chr.length.tbl)

library(janitor)
chr.cpgs.cnt <- tabyl(meth$chr, sort = T)
colnames(chr.cpgs.cnt) <- c("chr", "cnt_cpg", "freq_cpg")

chr.cpgs.cnt <- left_join(chr.cpgs.cnt, chr.length.tbl, by = "chr")
chr.cpgs.cnt$chr  <- factor(chr.cpgs.cnt$chr,levels = chr.order)
chr.cpgs.cnt[["freq_chr"]] <- chr.cpgs.cnt$chr_size / sum(chr.cpgs.cnt$chr_size)
chr.cpgs.cnt[["cpg_chr_cnt"]] <- chr.cpgs.cnt$cnt_cpg / chr.cpgs.cnt$freq_chr
chr.cpgs.cnt[["cpg_chr_freq"]] <- chr.cpgs.cnt$cpg_chr_cnt / sum(chr.cpgs.cnt$cpg_chr_cnt)

# Relative freq = (Number of CpGs on the chromosome (i) * (length of chromosomes{1:22} / length of the chromosome(i)) ) in %

ggplot(chr.cpgs.cnt, aes(x = chr, y = cpg_chr_freq, fill = chr)) + 
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_text(aes(label = scales::percent(cpg_chr_freq, accuracy = 0.1), y = cpg_chr_freq), 
            stat = "identity", vjust = -.5, size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(title = "Delta meQTLs CpGs distribution across the chromosomes wieghted by the chromosomes' length",
       x = "Chromosome",
       y = "Relative frequency") +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 8),
         axis.title = element_text(size = 8),
         axis.text.x = element_text(angle = 20, hjust = 0.5),
         legend.position = "none")
```

### Genomic regions

```{r out.width = "90%", fig.cap = "The distribution of CpGs in different genomic regions", include = T, eval = T}
cbPalette <- c(  "#009E73", "#E69F00", "#F0E442", "#D55E00", "#CC79A7", "#56B4E9", "#999999")

delta.meqtl.cpg.anno.df$Relation_to_Island <- factor(delta.meqtl.cpg.anno.df$Relation_to_Island, 
                                                              levels = c("N_Shelf", "N_Shore", "Island", "S_Shore", "S_Shelf", "OpenSea"))

ggplot(delta.meqtl.cpg.anno.df, aes(x = Relation_to_Island)) +
  geom_bar(aes(y = (..count..)/sum(..count..), fill = Relation_to_Island), position = position_dodge()) +
  geom_text(aes(label = scales::percent((..count..)/sum(..count..), accuracy = 0.1), y = (..count..)/sum(..count..)),
            stat = "count", vjust = -.5,  position = position_dodge(1), size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "Relation to Island",
       y = "Frequency",
       title = "") +
  theme(legend.title = element_blank(),
        legend.position = "none",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 45, hjust = 0.5)) +
   scale_fill_manual(values = cbPalette)
```

### Gene regions

```{r out.width = "90%", fig.cap = "The distribution of CpGs in different gene regions", include = T, eval = T}
cbPalette <- c(  "#009E73", "#E69F00", "#F0E442", "#D55E00", "#CC79A7", "#56B4E9", "#999999")

# delta.meqtl.cpg.closest.genes.dist.df$GeneGroup <- factor(delta.meqtl.cpg.closest.genes.dist.df$GeneGroup, 
#                                                               levels = c("1-5kb", "promoters", "5UTRs", "introns", "exons", "3UTRs"))

ggplot(delta.meqtl.cpg.closest.genes.dist.df, aes(x = GeneGroup)) + 
  geom_bar(aes(y = (..count..)/sum(..count..), fill = GeneGroup), position = position_dodge()) + 
  geom_text(aes(label = scales::percent((..count..)/sum(..count..), accuracy = 0.1), y = (..count..)/sum(..count..)), 
            stat = "count", vjust = -.5,  position = position_dodge(1), size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "Relation to Gene Region",
       y = "Frequency", 
       title = "From IlluminaHumanMethylationEPICanno.ilm10b4.hg19 package ") +
  theme(legend.title = element_blank(), 
        legend.position = "none", #c(.1,.8),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 45, hjust = 0.5)) +
   scale_fill_manual(values = cbPalette)
```

```{r eval = F, include = F}
get_df <- function(d) {
  ids <- d[, 1]
  means <- rowMeans(d[, -1], na.rm=T)
  # sds <- rowSds(d, na.rm=T)

  data.frame(id = ids, mean = means) #, sd=sds)
}
```

```{r eval = F, include = F}
output.eqtm.pre <- "~/bio/code/mpip/dex-stim-human-array/data/integrative/matrixEQTL/"

methyl.mtrx.veh <- fread(paste0(output.eqtm.pre, "methyl_beta_mtrx_veh.csv"))
methyl.mtrx.dex <- fread(paste0(output.eqtm.pre, "methyl_beta_mtrx_dex.csv"))
methyl.mtrx <- data.frame(cbind(methyl.mtrx.veh, methyl.mtrx.dex[, -1]))

```

```{r eval = F, include = F}
methyl.mtrx.veh.delta <- methyl.mtrx.veh[CpG_ID %in% delta.meqtl.cpg.anno.df$Name]

methyl.mtrx.veh.stat <- get_df(methyl.mtrx.veh.delta)
methyl.mtrx.veh.stat <- left_join(methyl.mtrx.veh.stat, delta.meqtl.cpg.anno.df[, c("Name", "Relation_to_Island")],
                                  by = c("CpG_ID" = "Name"))

ggplot(methyl.mtrx.veh.stat, aes(x = mean)) + 
  geom_histogram(aes(y = (..count..)/sum(..count..), fill = Relation_to_Island)) +
  geom_text(aes(label = scales::percent((..count..)/sum(..count..), accuracy = 0.1), y = (..count..)/sum(..count..)), 
            stat = "count", vjust = -.5,  position = position_dodge(1), size = 1) +
  scale_y_continuous(labels = scales::percent) 
```

```{r eval = F, include = F}
methyl.mtrx.dex.delta <- methyl.mtrx.dex[CpG_ID %in% delta.meqtl.cpg.anno.df$Name]

methyl.mtrx.dex.stat <- get_df(methyl.mtrx.dex.delta)
methyl.mtrx.dex.stat <- left_join(methyl.mtrx.dex.stat, delta.meqtl.cpg.anno.df[, c("Name", "Relation_to_Island")],
                                  by = c("CpG_ID" = "Name"))

ggplot(methyl.mtrx.dex.stat, aes(x = mean)) + 
  geom_histogram(aes(y = (..count..)/sum(..count..), fill = Relation_to_Island)) +
  geom_text(aes(label = scales::percent((..count..)/sum(..count..), accuracy = 0.1), y = (..count..)/sum(..count..)), 
            stat = "count", vjust = -.5,  position = position_dodge(1), size = 1) +
  scale_y_continuous(labels = scales::percent) 
```

```{r include = F, eval = T}
# Prepare methyl ranges

cpg.coord.df    <- delta.meqtl.cpg.anno.df[, .(Name, chr, pos)]# %>% dplyr:: mutate(chr = sub("chr", "", chr))
cpg.coord.range <- makeGRangesFromDataFrame(cpg.coord.df, 
                                            start.field = "pos", 
                                            end.field = "pos", 
                                            seqnames.field = c("chr"))
names(cpg.coord.range) <- cpg.coord.df$Name
```

```{r include = F, eval = T}
library(annotatr)
library(AnnotationHub)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)

annots <- c('hg19_cpgs', 'hg19_basicgenes') #, 'hg19_genes_intergenic', 'hg19_genes_intronexonboundaries')
gene.annots <- c('hg19_basicgenes') #, 'hg19_genes_intergenic', 'hg19_genes_intronexonboundaries')
    
annotations <- build_annotations(
  genome = "hg19",
  annotations = gene.annots
)
```

```{r include = F, eval = T}
delta.meqtls.cpg.annotated <- annotate_regions(
    regions = cpg.coord.range,
    annotations = annotations,
    ignore.strand = TRUE,
    quiet = FALSE)

# print(delta.meqtls.cpg.annotated)
# plot_annotation(annotated_regions = delta.meqtls.cpg.annotated)

delta.meqtls.cpg.annotated.df <- data.frame(delta.meqtls.cpg.annotated) %>% setDT()

delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_1to5kb", annot.type := "1-5kb"]
delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_3UTRs", annot.type := "3UTRs"]
delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_5UTRs", annot.type := "5UTRs"]
delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_exons", annot.type := "exons"]
delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_introns", annot.type := "introns"]
delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_promoters", annot.type := "promoters"]
# head(delta.meqtls.cpg.annotated.df)

delta.meqtls.cpg.annotated.df <- delta.meqtls.cpg.annotated.df[, .(seqnames, start, end, annot.type)] %>% unique()

delta.meqtls.cpg.annotated.df$annot.type <- factor(delta.meqtls.cpg.annotated.df$annot.type, 
                                                              levels = c("1-5kb", "promoters", "5UTRs", "introns", "exons", "3UTRs"))
```

```{r out.width = "90%", fig.cap = "The distribution of CpGs in different gene regions", include = T, eval = T}
ggplot(delta.meqtls.cpg.annotated.df, aes(x = annot.type)) + 
  geom_bar(aes(y = (..count..)/sum(..count..), fill = annot.type), position = position_dodge()) + 
  geom_text(aes(label = scales::percent((..count..)/sum(..count..), accuracy = 0.1), y = (..count..)/sum(..count..)), 
            stat = "count", vjust = -.5,  position = position_dodge(1), size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "Gene Region",
       y = "Percentage of CpG", 
       title = "From UCSC Genome Browser") +
  theme(legend.title = element_blank(), 
        legend.position = "none", #c(.1,.8),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 45, hjust = 0.5)) +
   scale_fill_manual(values = cbPalette)
```

## Chromatin States {.tabset .tabset-fade .tabset-pills}

### All cell types

```{r include = F, eval = T}
# hmm.annotations <- build_annotations(
#   genome = c("hg19"),
#   annotations = "hg19_Hsmm-chromatin"
# )

chromhmm.all.states <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromHMM_all_states.Rds")
colnames(values(chromhmm.all.states))[1] <- "type"

chromhmm.states     <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromhmm_states.Rds") %>% setDT()
chromhmm.states[COLOR.NAME == "Silver", COLOR.NAME := "Grey"] 
chromhmm.states[COLOR.NAME == "White", COLOR.NAME := "Snow"] 

chromhmm.epigenomes <- read.csv("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/epigenomes.tsv", sep = "\t")

delta.meqtls.cpg.chromhmm.annotated <- annotate_regions(
    regions = cpg.coord.range,
    annotations = chromhmm.all.states,
    ignore.strand = TRUE,
    quiet = FALSE)

# print(delta.meqtls.cpg.chromhmm.annotated)
# plot_annotation(annotated_regions = delta.meqtls.cpg.chromhmm.annotated)
```

```{r include = F, eval = T}
cbPalette <- chromhmm.states$COLOR.NAME
```

```{r func-plot-chromhmm-states, include = F, eval = T}
GetChromHMMPlot <- function(delta.meqtls.cpg.chromhmm.annotated.df, cbPalette = NULL, plot.title = NULL){
  
  delta.meqtls.cpg.chromhmm.annotated.df$annot.type <- factor(delta.meqtls.cpg.chromhmm.annotated.df$annot.type, 
                                                              levels = chromhmm.states$STATE)
  
  ggplot(delta.meqtls.cpg.chromhmm.annotated.df, aes(x = annot.type)) + 
    geom_bar(aes(y = (..count..)/sum(..count..), fill = annot.type), position = position_dodge()) + 
    geom_text(aes(label = scales::percent((..count..)/sum(..count..), accuracy = 0.1), y = (..count..)/sum(..count..)), 
              stat = "count", vjust = -.5,  position = position_dodge(1), size = 3) +
    scale_y_continuous(labels = scales::percent) +
    labs(x = "ChromHMM states",
         y = "Percentage of CpG", 
         title = plot.title) +
    theme(legend.title = element_blank(), 
          legend.position = "none", #c(.1,.8),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 8),
          axis.title = element_text(size = 8),
          axis.text.x = element_text(angle = 45, hjust = 0.5)) +
    scale_fill_manual(values = cbPalette)
}
```

```{r out.width = "90%", fig.cap = "The distribution of CpGs in 15 chromatin states from 127 cell types", include = T, eval = T}
delta.meqtls.cpg.chromhmm.annotated.df <- data.frame(delta.meqtls.cpg.chromhmm.annotated)

delta.meqtls.cpg.chromhmm.annotated.df <- delta.meqtls.cpg.chromhmm.annotated.df[, c("seqnames", "start", "end", "annot.type")] %>% unique()

plot.title <- "15 chromatin states from 127 cell types"

GetChromHMMPlot(delta.meqtls.cpg.chromhmm.annotated.df, cbPalette, plot.title)
```

### Blood and T-cell

```{r out.width = "90%", fig.cap = "The distribution of CpGs in 15 chromatin states from 14 blood and T-cell types", include = T, eval = T}
delta.meqtls.cpg.chromhmm.annotated.df <- data.frame(delta.meqtls.cpg.chromhmm.annotated) %>% setDT()

chromhmm.blood.eids <- chromhmm.epigenomes[chromhmm.epigenomes$GROUP == "Blood & T-cell", "EID"]

delta.meqtls.cpg.chromhmm.annotated.df <- delta.meqtls.cpg.chromhmm.annotated.df[annot.code %in% chromhmm.blood.eids,
                                                                                 .(seqnames, start, end, annot.type)] %>% unique()

plot.title <- "15 chromatin states from 14 blood and T-cell types"

GetChromHMMPlot(delta.meqtls.cpg.chromhmm.annotated.df, cbPalette, plot.title)
```

### Brain

```{r out.width = "90%", fig.cap = "The distribution of CpGs in 15 chromatin states from 10 brain types", include = T, eval = T}
delta.meqtls.cpg.chromhmm.annotated.df <- data.frame(delta.meqtls.cpg.chromhmm.annotated) %>% setDT()

chromhmm.brain.eids <- chromhmm.epigenomes[chromhmm.epigenomes$GROUP == "Brain", "EID"]

delta.meqtls.cpg.chromhmm.annotated.df <- delta.meqtls.cpg.chromhmm.annotated.df[annot.code %in% chromhmm.brain.eids,
                                                                                 .(seqnames, start, end, annot.type)] %>% unique()

plot.title <- "15 chromatin states from 10 brain types"

GetChromHMMPlot(delta.meqtls.cpg.chromhmm.annotated.df, cbPalette, plot.title)
```
