---
title: "cis-meQTLs-eQTMs Analysis"
subtitle: "Dexamethasone-Stimulated Human Array Project"
author: 
  - name: Anastasiia Hryhorzhevska
    email: anastasiia_hry@psych.mpg.de
    url: https://github.com/ahryho
    affiliation: Max Planck Institute of Psychiatry
    affiliation_url: https://www.psych.mpg.de/2664393/medizinische-genomforschung
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    css: ../style.css
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
options(digits = 4, width = 100, stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE,
                      tidy.opts = list(width.cutoff = 100),
                      tidy=TRUE,
                      fig.pos = "H",
                      dpi = 200,
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE, 
                      cache.lazy = FALSE)

# define DT as the standard printing method for data.frame
library(knitr)
library(DT)
knit_print.data.frame <- function(x, ...) {
  knit_print(DT::datatable(x,
                           filter="top",
                           rownames = FALSE,
                           extensions = "FixedColumns",
                           options = list(
                             scrollX = TRUE,
                             pageLength = 5
                             
                           )), ...)
}

registerS3method("knit_print", "data.frame", knit_print.data.frame)
```
```{r include=FALSE}
library(data.table)
library(dplyr)

library(grid)
library(gridExtra)
library(ggplot2)
library(ggrepel)
library(scales)
library(GGally)
library(factoextra)
library(viridis)

library(UpSetR)
library(eulerr)
library(ggVennDiagram)
library(corrplot)
```

```{r load-func, include = F, eval = T}
setwd("~/bio/code/mpip/dex-stim-human-array/")
source("code/integrative/util.R")
cbPalette <- c( "#0072B2", "#009E73", "#E69F00", "#F0E442", "#D55E00", "#CC79A7", "#56B4E9", "#999999")
```

***
# __Overlaps: meQTLs and eQTMs__
***

```{r load-data, include = F, eval = T}
src.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/data/integrative/matrixEQTL/"
out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/"

meqtl.dex.fn   <- paste0(out.dir.pre, "meqtls/me-eqtl_cis_results_06122021/me-qtl_cis_result_dex_fdr_005.csv")  # "me-qtl_cis_result_dex_fdr_005.csv")
meqtl.veh.fn   <- paste0(out.dir.pre, "meqtls//me-eqtl_cis_results_06122021/me-qtl_cis_result_veh_fdr_005.csv") # "me-qtl_cis_result_veh_fdr_005.csv")
meqtl.delta.fn <- paste0(out.dir.pre, "meqtls/me-qtl_cis_result_delta_fdr_005.csv") # "me-qtl_cis_result_delta_beta.csv")# 

col.names <- c("SNP", "CpG_ID", "beta_meqtl", "t-stat_meqtl", "p-value_meqtl", "fdr_meqtl")

meqtl.dex.df   <- fread(meqtl.dex.fn, col.names = col.names)
meqtl.veh.df   <- fread(meqtl.veh.fn, col.names = col.names)
meqtl.delta.df <- fread(meqtl.delta.fn, col.names = col.names)

meqtl.all.full.df <- fread()
# Load eQTMs

eqtm.veh.nom.fn   <- paste0(out.dir.pre, "eqtms/eqtm_cis_result_veh_beta.csv") 
eqtm.dex.nom.fn   <- paste0(out.dir.pre, "eqtms/eqtm_cis_result_dex_beta.csv") 
eqtm.delta.nom.fn <- paste0(out.dir.pre, "eqtms/eqtm_cis_result_delta_beta.csv") 

col.names       <- c("CpG_ID", "ENSG_ID", "beta_eqtm", "t-stat_eqtm", "p-value_eqtm", "fdr_eqtm")

eqtm.veh.nom.df   <- fread(eqtm.veh.nom.fn, col.names = col.names)
eqtm.dex.nom.df   <- fread(eqtm.dex.nom.fn, col.names = col.names)
eqtm.delta.nom.df <- fread(eqtm.delta.nom.fn, col.names = col.names)

eqtm.all.nom.df   <- rbind(eqtm.veh.nom.df, eqtm.dex.nom.df, eqtm.delta.nom.df)

# Load publicly available data from Josine L.Min at el. 2021
# 
# public.meqtl.df <- fread("data/public_data/GoDMC_mQTL/assoc_meta_all.csv")
# public.meqtl.sub.df <- public.meqtl.df[, .(cpg, snp, pval)]
```

```{r, include = F, eval = T}
meqtl.all.full.df <- rbind(meqtl.dex.df[, treatment := "dex"],
                           meqtl.veh.df[, treatment := "veh"],
                           meqtl.delta.df[, treatment := "delta"])

meqtl.all.full.df <- meqtl.all.full.df[!is.na(meqtl.all.full.df$FDR),]
meqtl.all.full.df[["meQTL_ID"]] <- paste(meqtl.all.full.df$SNP, meqtl.all.full.df$CpG_ID, sep = "-")
```


## Baseline

```{r out.width = "95%", include = T, eval = T, fig.cap = "Number of intersections, veh CpGs"}
GetVennPlt(meqtl.df = meqtl.veh.df, eqtm.df = eqtm.veh.nom.df, cbPal.col = "#E69F00")
```

## Dex-stimutated

```{r out.width = "95%", include = T, eval = T, fig.cap = "Number of intersections, dex CpGs"}
GetVennPlt(meqtl.df = meqtl.dex.df, eqtm.df = eqtm.dex.nom.df, cbPal.col = "#009E73")
```

## Delta {.tabset .tabset-fade .tabset-pills}

### Dex eQTMs

```{r out.width = "95%", include = T, eval = T, fig.cap = "Number of intersections, delta CpGs"}
GetVennPlt(meqtl.df = meqtl.delta.df, eqtm.df = eqtm.dex.nom.df)
# gridExtra::grid.arrange(venn.dex, venn.delta, nrow = 1, heights = 10, 
#                         left = "delta meQTLs with dex eQTMs",
#                         right = "delta meQTLs with delta eQTMs")
```

### Delta eQTMs

```{r out.width = "95%", include = T, eval = T, fig.cap = "Number of intersections, delta CpGs"}
GetVennPlt(meqtl.df = meqtl.delta.df, eqtm.df = eqtm.delta.nom.df)
```

## __meQTLs groups based on allelic direction__

```{r load-omics}
meth.beta.dex.mtrx <- LoadMethylBeta("dex")
meth.beta.veh.mtrx <- LoadMethylBeta("veh")
gex.dex.mtrx       <- LoadGEX("dex")
gex.veh.mtrx       <- LoadGEX("veh")
snp.mtrx           <- LoadGenotype()
```

```{r eval = T, include = F}
out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/meqtls/primary_meqtl_results/"

meqtl.parallel.fc.grp.veh.dex.gr <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_parallel_fc_grp_veh_dex_gr.rds"))
meqtl.parallel.fc.grp.delta.gr   <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_parallel_fc_grp_delta_gr.rds"))
meqtl.opposite.fc.gr             <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_opposite_fc_grp_gr.rds"))
```

```{r include = F, eval = T}
parallel.fc.grp.veh.dex.cpg <- list(CpG_ID = names(meqtl.parallel.fc.grp.veh.dex.gr@anno))
parallel.fc.grp.delta.cpg   <- list(CpG_ID = names(meqtl.parallel.fc.grp.delta.gr@anno))
opposite.fc.grp.cpg         <- list(CpG_ID = names(meqtl.opposite.fc.gr@anno))
```

```{r load-meqtl-fc-groups}
dir.pre <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/meqtls/meqtl_parallel_and_opposite_fc_groups/"

meqtl.parallel.fc.grp.veh.dex.df <- fread(paste0(dir.pre, "meqtl_parallel_fc_gr_veh_dex_df.csv"))
meqtl.parallel.fc.grp.delta.df   <- fread(paste0(dir.pre, "meqtl_parallel_fc_gr_delta_df.csv"))
meqtl.opposite.fc.df             <- fread(paste0(dir.pre, "meqtl_opposite_fc_gr_df.csv"))
```

### Parallel FC veh & dex meQTLs & veh eQTMs

```{r eval = T, include = F}
eqtm.veh.nom.df[["eQTM_ID"]] <- paste(eqtm.veh.nom.df$CpG_ID, eqtm.veh.nom.df$ENSG_ID, sep = "-")
eqtm.dex.nom.df[["eQTM_ID"]] <- paste(eqtm.dex.nom.df$CpG_ID, eqtm.dex.nom.df$ENSG_ID, sep = "-")
intersect.dex.veh.eqtms      <- intersect(eqtm.veh.nom.df$eQTM_ID, eqtm.dex.nom.df$eQTM_ID)  
```
Intersection between the group of baseline meQTLs with the same allelic (FC) directions and the set of baseline eQTMs which are not intersected with dex eQTMs 
<!-- The direction of an allelic effect is the direction (add or subtract) that an allele has on a phenotype. -->
```{r out.width = "95%", include = T, eval = T, fig.cap = "Number of intersections, CpGs \n dex + veh with parallel FC"}
venn.rslt <- GetVennPlt(meqtl.df = parallel.fc.grp.veh.dex.cpg, eqtm.df = eqtm.veh.nom.df, cbPal.col = "#999999")
venn.rslt$venn.plot
```

#### Counts {.tabset .tabset-fade .tabset-pills}

```{r}
venn.eqtm.df  <- eqtm.veh.nom.df[CpG_ID %in% venn.rslt$cpgs, .(CpG_ID, ENSG_ID)] 
venn.meqtl.df <- meqtl.parallel.fc.grp.veh.dex.df[CpG_ID %in% venn.rslt$cpgs][, .(meQTL_ID, CpG_ID, SNP)] %>% unique()
# plots.cnt     <- HistoPlotCntPerProbes(venn.eqtm.df)
```

```{r}
print(paste("eQTMs:", nrow(venn.eqtm.df), "meQTLs:", length(venn.meqtl.df$meQTL_ID),
             "CpGs:", length(unique(venn.meqtl.df$CpG_ID)), "SNPs:", length(unique(venn.meqtl.df$SNP)), 
             "ENSGs:", length(unique(venn.eqtm.df$ENSG_ID)), sep = " "))
```

##### Numebr of CpG per GEX 

```{r incldue = F, eval = T}
 num.ensg.per.cpg <- venn.eqtm.df %>% 
    group_by(CpG_ID) %>% tally()
  
 plot.gex.cnt <- ggplot(num.ensg.per.cpg, aes(n)) + 
    geom_bar(position = position_dodge()) +
    scale_x_continuous(limits = c(0, max(num.ensg.per.cpg$n) + 5), 
                       breaks = seq(0, max(num.ensg.per.cpg$n) + 5, 10), 
                       expand = c(0, 0)) + 
    theme(
      legend.title = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank(),
      axis.text.y = element_blank(),
      plot.title = element_text(size = 10),
      axis.title = element_text(size = 8),
      axis.ticks.y = element_blank()) +
    labs(title = "Number of GEX probes per CpG", 
         x = "Number of GEX probes", 
         y = "")
```

```{r}
plot.gex.cnt
```

##### Numebr of GEX per CpG 

```{r include = F, eval = T}
 num.cpg.per.gex <- venn.eqtm.df %>% 
    group_by(ENSG_ID) %>% tally()
  
 plot.cpg.cnt <- ggplot(num.cpg.per.gex, aes(n)) + 
    geom_bar(position = position_dodge()) +
    scale_x_continuous(limits = c(0, max(num.ensg.per.cpg$n) + 5), 
                       breaks = seq(0, max(num.ensg.per.cpg$n) + 5, 10), 
                       expand = c(0, 0)) + 
    theme(legend.title = element_blank(),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          axis.text.y = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.ticks.y = element_blank()) +
    labs(title = "Number of CpGs per GEX probe", 
         x = "Number of CpG sites", 
         y = "")
```

```{r}
plot.cpg.cnt
```
#### Heatmap{.tabset .tabset-fade .tabset-pills}

```{r }
n.top.eqtm <- 200

top.eqtm       <- eqtm.veh.nom.df[CpG_ID %in% venn.rslt$cpgs][1:n.top.eqtm]
top.eqtm.cpgs  <- top.eqtm$CpG_ID %>% unique()
top.eqtm.ensgs <- top.eqtm$ENSG_ID %>% unique()

meth.sub.df   <- meth.beta.veh.mtrx[CpG_ID %in% top.eqtm.cpgs]
gex.sub.df    <- gex.veh.mtrx[ENSG_ID %in% top.eqtm.ensgs]

meth.cor.mtrx <- cor(t(meth.sub.df[, -1]))
gex.cor.mtrx  <- cor(t(gex.sub.df[, -1]))

meth.gex.sub.df   <- rbind(meth.sub.df[, -1], gex.sub.df[, -1])
meth.gex.cor.mtrx <- cor(t(meth.gex.sub.df))
```

##### CpG & GEX

```{r out.width = "95%", include = T, eval = T, fig.cap = "Correlation within cis-window of 1Mbp, CpG and GEX \n delta with parallel FC"}
corrplot(meth.gex.cor.mtrx, 
         method = "color", cl.pos = "r", 
         type = "full",
         tl.pos = "n",
         title = paste0("Top ", n.top.eqtm, " eQTMs, ", "CpG & GEX"))
```

##### CpG

```{r out.width = "95%", include = T, eval = T, fig.cap = "Correlation within cis-window of 1Mbp, CpGs \n delta with parallel FC"}
corrplot(meth.cor.mtrx, 
         method = "color", cl.pos = "r", 
         type = "full",
         tl.pos = "n")
```

##### GEX

```{r out.width = "95%", include = T, eval = T, fig.cap = "Correlation within cis-window of 1Mbp, GEX \n delta with parallel FC"}
corrplot(gex.cor.mtrx, 
         method = "color", cl.pos = "r", 
         type = "full",
         tl.pos = "n")
```

#### Examples

```{r include = T, eval = T}
# paste0("Scatter plot for the independent significant at FDR <= ", fdr.thr, " meQTLs")
eqtm.id       <- eqtm.veh.nom.df[eQTM_ID %in% top.eqtm$eQTM_ID][fdr_eqtm == min(fdr_eqtm), eQTM_ID]
selected.eqtm <- eqtm.veh.nom.df[eQTM_ID %in% eqtm.id[1]] # beta == max(beta) cg10186336

selected.meqtl <- meqtl.parallel.fc.grp.veh.dex.df[CpG_ID %in% selected.eqtm$CpG_ID, 
                                                 .(CpG_ID, SNP, beta, `p-value`, fdr, treatment)] %>% unique()

selected.eqtl <- data.frame(ENSG_ID = selected.eqtm$ENSG_ID, SNP = selected.meqtl$SNP) %>% unique()

data.table(selected.eqtm %>% 
        dplyr::select(CpG_ID, ENSG_ID, FC = beta_eqtm, t_stat = `t-stat_eqtm`, 
                      p_val = `p-value_eqtm`, FDR = fdr_eqtm) %>% 
        mutate_if(is.numeric, funs(as.character(signif(., 3)))))

data.table(selected.meqtl %>% 
        dplyr::select(CpG_ID, SNP, FC = beta, 
                      p_val = `p-value`, FDR = fdr, treatment) %>% 
        mutate_if(is.numeric, funs(as.character(signif(., 3)))))
```

```{r, eval = T, include = F}
plot.title <- paste0("Parallel FC: veh x dex: (", selected.meqtl$CpG_ID[1], 
                     ", ", selected.meqtl$SNP[1], ", ", selected.eqtm$ENSG_ID, ")")
```

```{r out.width = "95%", include = T, eval = T}
GetScatterPlot(meqtl.all.full.df, selected.meqtl, fdr.thr = 0.05,
               plot.title = plot.title)
```

```{r out.width = "95%", include = T, eval = T}
ScatterPlotGEXvsDNAm(meth.beta.dex.mtrx, meth.beta.veh.mtrx,
                     gex.dex.mtrx, gex.veh.mtrx,
                     cpg.id = selected.eqtm$CpG_ID, ensg.id = selected.eqtm$ENSG_ID)
```

```{r out.width = "95%", include = T, eval = T}
ProcessGetBoxPlot(meth.beta.veh.mtrx, meth.beta.dex.mtrx, snp.mtrx, selected.meqtl[1], fdr.thr = 0.05, 
                  plot.title = plot.title)
```

```{r out.width = "95%", include = T, eval = T}
ProcessGetBoxPlot(gex.veh.mtrx, gex.dex.mtrx, snp.mtrx, selected.eqtl[1,], fdr.thr = 0.05, 
                  plot.title = plot.title)
```

### Parallel FC delta meQTLs & dex eQTMs

```{r out.width = "95%", include = T, eval = T, fig.cap = "Number of intersections, CpGs \n delta with parallel FC"}
venn.rslt <- GetVennPlt(meqtl.df = parallel.fc.grp.delta.cpg, eqtm.df = eqtm.dex.nom.df, cbPal.col = "#999999")
venn.rslt$venn.plot
```

#### Counts {.tabset .tabset-fade .tabset-pills}

```{r}
venn.eqtm.df  <- eqtm.dex.nom.df[CpG_ID %in% venn.rslt$cpgs, .(CpG_ID, ENSG_ID)] 
venn.meqtl.df <- meqtl.parallel.fc.grp.delta.df[CpG_ID %in% venn.rslt$cpgs][, .(meQTL_ID, CpG_ID, SNP)] %>% unique()
plots.cnt     <- HistoPlotCntPerProbes(venn.eqtm.df)
```

```{r}
print(paste("eQTMs:", nrow(venn.eqtm.df), "meQTLs:", length(venn.meqtl.df$meQTL_ID),
             "CpGs:", length(unique(venn.meqtl.df$CpG_ID)), "SNPs:", unique(length(venn.meqtl.df$SNP)), 
             "ENSGs:", length(unique(venn.eqtm.df$ENSG_ID)), sep = " "))
```

##### Numebr of CpG per GEX 

```{r}
plots.cnt$cpg
```

##### Numebr of GEX per CpG 

```{r}
plots.cnt$gex
```


#### Heatmap{.tabset .tabset-fade .tabset-pills}

```{r }
n.top.eqtm <- 200

top.eqtm       <- eqtm.dex.nom.df[CpG_ID %in% venn.rslt$cpgs][1:n.top.eqtm]
top.eqtm.cpgs  <- top.eqtm$CpG_ID %>% unique()
top.eqtm.ensgs <- top.eqtm$ENSG_ID %>% unique()

meth.sub.df   <- meth.beta.dex.mtrx[CpG_ID %in% top.eqtm.cpgs]
gex.sub.df    <- gex.dex.mtrx[ENSG_ID %in% top.eqtm.ensgs]

meth.cor.mtrx <- cor(t(meth.sub.df[, -1]))
gex.cor.mtrx  <- cor(t(gex.sub.df[, -1]))

meth.gex.sub.df   <- rbind(meth.sub.df[, -1], gex.sub.df[, -1])
meth.gex.cor.mtrx <- cor(t(meth.gex.sub.df))
```

##### CpG & GEX

```{r out.width = "95%", include = T, eval = T, fig.cap = "Correlation within cis-window of 1Mbp, CpG and GEX \n delta with parallel FC"}
corrplot(meth.gex.cor.mtrx, 
         method = "color", cl.pos = "r", 
         type = "full",
         tl.pos = "n",
         title = paste0("Top ", n.top.eqtm, " eQTMs, ", "CpG & GEX"))
```

##### CpG

```{r out.width = "95%", include = T, eval = T, fig.cap = "Correlation within cis-window of 1Mbp, CpGs \n delta with parallel FC"}
corrplot(meth.cor.mtrx, 
         method = "color", cl.pos = "r", 
         type = "full",
         tl.pos = "n")
```

##### GEX

```{r out.width = "95%", include = T, eval = T, fig.cap = "Correlation within cis-window of 1Mbp, GEX \n delta with parallel FC"}
corrplot(gex.cor.mtrx, 
         method = "color", cl.pos = "r", 
         type = "full",
         tl.pos = "n")
```

#### Examples

```{r include = T, eval = T}
# paste0("Scatter plot for the independent significant at FDR <= ", fdr.thr, " meQTLs")
eqtm.id       <- eqtm.dex.nom.df[eQTM_ID %in% intersect.dex.veh.eqtms][fdr_eqtm == min(fdr_eqtm), eQTM_ID]
selected.eqtm <- eqtm.dex.nom.df[eQTM_ID %in% eqtm.id] # beta == max(beta) cg18193195

selected.meqtl <- meqtl.parallel.fc.grp.delta.df[CpG_ID %in% selected.eqtm$CpG_ID, 
                                                 .(CpG_ID, SNP, beta, `p-value`, fdr, treatment)] %>% unique()

selected.eqtl <- data.frame(ENSG_ID = selected.eqtm$ENSG_ID, SNP = selected.meqtl$SNP) %>% unique()

data.table(selected.eqtm %>% 
        dplyr::select(CpG_ID, ENSG_ID, FC = beta_eqtm, t_stat = `t-stat_eqtm`, 
                      p_val = `p-value_eqtm`, FDR = fdr_eqtm) %>% 
        mutate_if(is.numeric, funs(as.character(signif(., 3)))))

data.table(selected.meqtl %>% 
        dplyr::select(CpG_ID, SNP, FC = beta, 
                      p_val = `p-value`, FDR = fdr, treatment) %>% 
        mutate_if(is.numeric, funs(as.character(signif(., 3)))))
```

```{r, eval = T, include = F}
plot.title <- paste0("Parallel FC: veh x dex x delta: (", selected.meqtl$CpG_ID[1], 
                     ", ", selected.meqtl$SNP[1], ", ", selected.eqtm$ENSG_ID, ")")
```

```{r out.width = "95%", include = T, eval = T}

ScatterPlotGEXvsDNAm(meth.beta.dex.mtrx, meth.beta.veh.mtrx,
                     gex.dex.mtrx, gex.veh.mtrx,
                     cpg.id = selected.eqtm$CpG_ID, ensg.id = selected.eqtm$ENSG_ID)
```

```{r out.width = "95%", include = T, eval = T}
ProcessGetBoxPlot(meth.beta.veh.mtrx, meth.beta.dex.mtrx, snp.mtrx, selected.meqtl[1], fdr.thr = 0.05, 
                  plot.title = plot.title)
```

```{r out.width = "95%", include = T, eval = T}
ProcessGetBoxPlot(gex.veh.mtrx, gex.dex.mtrx, snp.mtrx, selected.eqtl[1,], fdr.thr = 0.05, 
                  plot.title = plot.title)
```

### Opposite FC meQTLs & dex eQTMs

```{r out.width = "95%", include = T, eval = T, fig.cap = "Number of intersections, CpGs \n opposite FC"}
venn.rslt <- GetVennPlt(meqtl.df = opposite.fc.grp.cpg, eqtm.df = eqtm.dex.nom.df, cbPal.col = "#999999")
venn.rslt$venn.plot
```

#### Counts {.tabset .tabset-fade .tabset-pills}

```{r}
venn.eqtm.df  <- eqtm.dex.nom.df[CpG_ID %in% venn.rslt$cpgs, .(CpG_ID, ENSG_ID)] 
venn.meqtl.df <- meqtl.opposite.fc.df[CpG_ID %in% venn.rslt$cpgs][, .(meQTL_ID, CpG_ID, SNP)] %>% unique()
plots.cnt     <- HistoPlotCntPerProbes(venn.eqtm.df)
```

```{r}
print(paste("eQTMs:", nrow(venn.eqtm.df), "meQTLs:", length(venn.meqtl.df$meQTL_ID),
             "CpGs:", length(unique(venn.meqtl.df$CpG_ID)), "SNPs:", length(unique(venn.meqtl.df$SNP)), 
             "ENSGs:", length(unique(venn.eqtm.df$ENSG_ID)), sep = " "))
```

##### Numebr of CpG per GEX 

```{r}
plots.cnt$cpg
```

##### Numebr of GEX per CpG 

```{r}
plots.cnt$gex
```

#### Heatmap{.tabset .tabset-fade .tabset-pills}

```{r }
n.top.eqtm <- 200

top.eqtm       <- eqtm.dex.nom.df[CpG_ID %in% venn.rslt$cpgs][1:n.top.eqtm]
top.eqtm.cpgs  <- top.eqtm$CpG_ID %>% unique()
top.eqtm.ensgs <- top.eqtm$ENSG_ID %>% unique()

meth.sub.df   <- meth.beta.dex.mtrx[CpG_ID %in% top.eqtm.cpgs]
gex.sub.df    <- gex.dex.mtrx[ENSG_ID %in% top.eqtm.ensgs]

meth.cor.mtrx <- cor(t(meth.sub.df[, -1]))
gex.cor.mtrx  <- cor(t(gex.sub.df[, -1]))

meth.gex.sub.df   <- rbind(meth.sub.df[, -1], gex.sub.df[, -1])
meth.gex.cor.mtrx <- cor(t(meth.gex.sub.df))
```

##### CpG & GEX

```{r out.width = "95%", include = T, eval = T, fig.cap = "Correlation within cis-window of 1Mbp, CpG and GEX \n delta with parallel FC"}
corrplot(meth.gex.cor.mtrx, 
         method = "color", cl.pos = "r", 
         type = "full",
         tl.pos = "n",
         title = paste0("Top ", n.top.eqtm, " eQTMs, ", "CpG & GEX"))
```

##### CpG

```{r out.width = "95%", include = T, eval = T, fig.cap = "Correlation within cis-window of 1Mbp, CpGs \n delta with parallel FC"}
corrplot(meth.cor.mtrx, 
         method = "color", cl.pos = "r", 
         type = "full",
         tl.pos = "n")
```

##### GEX

```{r out.width = "95%", include = T, eval = T, fig.cap = "Correlation within cis-window of 1Mbp, GEX \n delta with parallel FC"}
corrplot(gex.cor.mtrx, 
         method = "color", cl.pos = "r", 
         type = "full",
         tl.pos = "n")
```

#### Examples

```{r include = T, eval = T}
# paste0("Scatter plot for the independent significant at FDR <= ", fdr.thr, " meQTLs")
eqtm.id       <- eqtm.dex.nom.df[eQTM_ID %in% top.eqtm$eQTM_ID][fdr_eqtm == min(fdr_eqtm), eQTM_ID]
selected.eqtm <- eqtm.dex.nom.df[eQTM_ID %in% eqtm.id] # beta == max(beta) cg08880347

selected.meqtl <- meqtl.parallel.fc.grp.delta.df[CpG_ID %in% selected.eqtm$CpG_ID, 
                                                 .(CpG_ID, SNP, beta, `p-value`, fdr, treatment)] %>% unique()

selected.eqtl <- data.frame(ENSG_ID = selected.eqtm$ENSG_ID, SNP = selected.meqtl$SNP) %>% unique()

data.table(selected.eqtm %>% 
        dplyr::select(CpG_ID, ENSG_ID, FC = beta_eqtm, t_stat = `t-stat_eqtm`, 
                      p_val = `p-value_eqtm`, FDR = fdr_eqtm) %>% 
        mutate_if(is.numeric, funs(as.character(signif(., 3)))))

data.table(selected.meqtl %>% 
        dplyr::select(CpG_ID, SNP, FC = beta, 
                      p_val = `p-value`, FDR = fdr, treatment) %>% 
        mutate_if(is.numeric, funs(as.character(signif(., 3)))))
```

```{r, eval = T, include = F}
plot.title <- paste0("Opposite FC: (", selected.meqtl$CpG_ID[1], 
                     ", ", selected.meqtl$SNP[1], ", ", selected.eqtm$ENSG_ID, ")")
```

```{r out.width = "95%", include = T, eval = T}

ScatterPlotGEXvsDNAm(meth.beta.dex.mtrx, meth.beta.veh.mtrx,
                     gex.dex.mtrx, gex.veh.mtrx,
                     cpg.id = selected.eqtm$CpG_ID, ensg.id = selected.eqtm$ENSG_ID)
```

```{r out.width = "95%", include = T, eval = T}
ProcessGetBoxPlot(meth.beta.veh.mtrx, meth.beta.dex.mtrx, snp.mtrx, selected.meqtl[1], fdr.thr = 0.05, 
                  plot.title = plot.title)
```

```{r out.width = "95%", include = T, eval = T}
ProcessGetBoxPlot(gex.veh.mtrx, gex.dex.mtrx, snp.mtrx, selected.eqtl[1,], fdr.thr = 0.05, 
                  plot.title = plot.title)
```