---
title: "cis-meQTLs Analysis"
subtitle: "Dexamethasone-Stimulated Human Array Project"
author: 
  - name: Anastasiia Hryhorzhevska
    email: anastasiia_hry@psych.mpg.de
    url: https://github.com/ahryho
    affiliation: Max Planck Institute of Psychiatry
    affiliation_url: https://www.psych.mpg.de/2664393/medizinische-genomforschung
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    css: style.css
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
options(digits = 4, width = 100, stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE,
                      tidy.opts = list(width.cutoff = 100),
                      tidy=TRUE,
                      fig.pos = "H",
                      dpi = 200,
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE, 
                      cache.lazy = FALSE)

# define DT as the standard printing method for data.frame
library(knitr)
library(DT)
knit_print.data.frame <- function(x, ...) {
  knit_print(DT::datatable(x,
                           filter="top",
                           rownames = FALSE,
                           extensions = "FixedColumns",
                           options = list(
                             scrollX = TRUE,
                             pageLength = 5
                             
                           )), ...)
}

registerS3method("knit_print", "data.frame", knit_print.data.frame)

library(data.table)
library(arrow)
library(dplyr)
library(eulerr)
```

```{r}
setwd("~/bio/code/mpip/dex-stim-human-array/")
source("code/integrative/util.R")
```

```{r load-and-prepare-pub-data, eval=F}
# Load publicly available data from Josine L.Min at el. 2021
public.meqtl.df <- fread("data/public_data/GoDMC_mQTL/assoc_meta_all.csv", select = c("cpg", "snp", "pval", "beta_a1", "cistrans"))
public.meqtl.cis.sub.df <- public.meqtl.df[cistrans == T] # 71'616'458
```

```{r load-public-meqtls, eval = T}
public.meqtl.cis.sub.interest.df <- fread("data/public_data/GoDMC_mQTL/assoc_meta_cis_meqtl.csv")
```

```{r load-veh-meqtls, eval = T}
col.names    <- c("SNP", "CpG_ID", "beta", "t-stat", "p-value", "FDR")
meqtl.veh.df <- fread("~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/meqtls/with_missingness/me-qtl_cis_result_snps_with_na_veh_beta_fdr_005.csv", col.names = col.names) 
```

```{r load-dnam-veh}
methyl.beta.mtrx <- LoadMethylBeta("veh")
all.veh.cpgs     <- methyl.beta.mtrx$CpG_ID
rm(methyl.beta.mtrx)
```

```{r load-snp-data}
# Genotype data

snp.loc.df <- fread("data/integrative/matrixEQTL/snp_locations.csv")

head(snp.loc.df)
dim(snp.loc.df)

snp.loc.df[["SNP_ID"]] <- paste0("chr", snp.loc.df$chr, ":", snp.loc.df$pos, ":", "SNP")
```

# Overlap at meCpG level: significant meQTLs @ FDR <= 0.05
 
```{r}
veh.mecpgs <- meqtl.veh.df$CpG_ID %>% unique() # 137'537
pub.mecpgs <- public.meqtl.cis.sub.interest.df$cpg %>% unique() # 232'477

intersect.all.veh.pub.cpgs <- intersect(all.veh.cpgs, pub.mecpgs) # 203'696
intersect.mecpgs.fdr       <- intersect(veh.mecpgs, intersect.all.veh.pub.cpgs) #53'891

length(intersect.mecpgs.fdr) / length(veh.mecpgs) # 39.18%
length(intersect.mecpgs.fdr) / length(intersect.all.veh.pub.cpgs) # 26.46 %
```

# Extract public meQTLs with overlapped with baseline CpGs 

```{r eval = F}
public.meqtl.cis.sub.interest.df      <- public.meqtl.cis.sub.df[cpg %in% intersect.all.veh.pub.cpgs, ] # 59'720'788

public.meqtl.cis.sub.interest.df[["meQTL_ID"]] <- paste(public.meqtl.cis.sub.interest.df$cpg, public.meqtl.cis.sub.interest.df$snp, sep = "-")

fwrite(public.meqtl.cis.sub.interest.df, 
       "data/public_data/GoDMC_mQTL/assoc_meta_cis_meqtl.csv",
       quote = F, row.names = F, sep = ";")
```

# Extract baseline meQTLs with overlapped public meQTL CpGs 

```{r eval = F}
public.meqtl.cis.sub.interest.df <- fread("data/public_data/GoDMC_mQTL/assoc_meta_cis_meqtl.csv")
intersect.all.veh.pub.cpgs       <- pub.mecpgs
meqtl.veh.df                     <- meqtl.veh.df[CpG_ID %in% intersect.all.veh.pub.cpgs, ] # 2'985'831
```

# Join baseline meQTL db with SNP ID
```{r}
meqtl.veh.df  <- left_join(meqtl.veh.df, snp.loc.df)
meqtl.veh.lst <- paste(meqtl.veh.df$CpG_ID, meqtl.veh.df$SNP_ID, sep = "-") #  3,007,281
length(meqtl.veh.lst)
```

# Overlap at meQTL level: significant veh meQTLs @ FDR <= 0.05 and public meQTL @ p-value < 1e-5

```{r}
intersect.meqtls <- intersect(meqtl.veh.lst, public.meqtl.cis.sub.interest.df$meQTL_ID) # 2'861'232
length(intersect.meqtls); length(intersect.meqtls) / length(meqtl.veh.lst) # 95.14 %
```

```{r}
meqtls.meqtls <- list(baseline = unique(meqtl.veh.lst), 
                      GoDMC = unique(public.meqtl.cis.sub.interest.df$meQTL_ID))

perc.olap <- scales::percent(length(intersect.meqtls) / length(meqtl.veh.lst), accuracy = 0.1)
```

Number of intersections between baseline and GoDMC meQTLs

```{r out.width = "95%", include = T, eval = T}
euler.tbl <- euler(meqtls.meqtls)

# For visualisation purposes:
euler.tbl$ellipses["baseline", c("a", "b")] <- euler.tbl$ellipses["baseline", c("a", "b")] * 2

plot(euler.tbl,
     quantities = list(type = c("counts")),
     fills = alpha(c("grey", "pink"), 0.5),
     col = alpha(c("grey", "pink"), 1),
     labels =  c("Baseline", "GoDMC"))
```


## Direction of meQTLs: intersection (veh meQTLs; public)

### Prepare data 
```{r}
pub.meqtl.intersect.with.veh.df <- public.meqtl.cis.sub.interest.df[meQTL_ID %in% intersect.meqtls]

colnames(pub.meqtl.intersect.with.veh.df) <- c("CpG_ID", "SNP_ID", "p-value_pub", "beta_pub", "cistrans", "meQTL_ID")
colnames(meqtl.veh.df)                    <- c("rsSNP", "CpG_ID", "beta_veh", "t-stat_veh",  "p-value_veh", "FDR_veh", "CHR_SNP", "POS_SNP", "SNP_ID")

pub.meqtl.intersect.with.veh.df <- left_join(pub.meqtl.intersect.with.veh.df, meqtl.veh.df, by = c("CpG_ID", "SNP_ID")) # 2'851'678 x 13
```

### Check the FC directions and save results

```{r}
pub.meqtl.intersect.with.veh.df[, FC_direction := sign(beta_pub * beta_veh)]

fwrite(pub.meqtl.intersect.with.veh.df, 
       "data/public_data/GoDMC_mQTL/godmc_veh_olap_meqtl_df.csv",
       quote = F, row.names = F, sep = ";")

pub.meqtl.intersect.with.veh.df <- fread("data/public_data/GoDMC_mQTL/godmc_veh_olap_meqtl_df.csv")
```

```{r}
fc.dir.tbl <- table(pub.meqtl.intersect.with.veh.df$FC_direction)
fc.dir.tbl 
#     -1       1 
# 2768486   83192
```

```{r}
fc.dir.tbl <- table(pub.meqtl.intersect.with.veh.df$FC_direction)

print(paste0("% of meQTLs with opposite allelic durections: ", signif(fc.dir.tbl["-1"] / sum(fc.dir.tbl), 4) * 100, " %")) # 97.08%
print(paste0("% of meQTLs with parallel allelic durections: ", signif(fc.dir.tbl["1"] / sum(fc.dir.tbl), 4) * 100, " %"))  # 2.92 %
```

# Overlap at meQTL level: significant veh meQTLs @ FDR <= 0.05 and public meQTL @ p-value < 0.05

```{r}
pub.meqtl.intersect.with.veh.pubfdr.df <- pub.meqtl.intersect.with.veh.df[`p-value_pub` <= 0.05] # 2'849'887

nrow(pub.meqtl.intersect.with.veh.pubfdr.df) / length(meqtl.veh.lst) # 95.44 %
```