---
title: "cis-meQTLs Analysis"
subtitle: "Dexamethasone-Stimulated Human Array Project"
author: 
  - name: Anastasiia Hryhorzhevska
    email: anastasiia_hry@psych.mpg.de
    url: https://github.com/ahryho
    affiliation: Max Planck Institute of Psychiatry
    affiliation_url: https://www.psych.mpg.de/2664393/medizinische-genomforschung
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    css: ../../../style.css
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
options(digits = 4, width = 100, stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE,
                      tidy.opts = list(width.cutoff = 100),
                      tidy=TRUE,
                      fig.pos = "H",
                      dpi = 150,
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE, 
                      include = F,
                      eval = T,
                      cache.lazy = FALSE)

# define DT as the standard printing method for data.frame
library(knitr)
library(DT)
knit_print.data.frame <- function(x, ...) {
  knit_print(DT::datatable(x,
                           filter="top",
                           rownames = FALSE,
                           extensions = "FixedColumns",
                           options = list(
                             scrollX = TRUE,
                             pageLength = 5
                             
                           )), ...)
}

registerS3method("knit_print", "data.frame", knit_print.data.frame)
```

# __Functional CpG annotation__
***

```{r, include = F, eval = T}
source("~/bio/code/mpip/dex-stim-human-array/code/util.R", chdir = TRUE)

library(data.table)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(scales)
library(GGally)
library(factoextra)
library(viridis)

library(splitstackshape)
library(reshape2)
library(IRanges)
library(ggplot2)
library(biomaRt)

require(foreign)
library(parallel)
library(foreach)
library(doParallel)

library(gUtils)

library(officer)

dir.pre      <- "~/bio/code/mpip/dex-stim-human-array/"
# dir.pre      <- "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/"        
out.dir.pre  <- paste0(dir.pre, "output/data/integrative/matrixEQTL/meqtls/region_wise_independent_snps/")

# meqtl.dex.fn   <- paste0(out.dir.pre, "me-qtl_cis_indp_result_dex_with_delta_fdr_005.csv")
# meqtl.veh.fn   <- paste0(out.dir.pre, "me-qtl_cis_indp_result_veh_fdr_005.csv")
# meqtl.delta.fn <- paste0(out.dir.pre, "me-qtl_cis_indp_result_delta_fdr_005.csv")

# col.names <-  c("SNP", "CpG_ID", "beta", "t-stat", "p-value", "fdr")

# ind.meqtl.dex.df   <- fread(meqtl.dex.fn, col.names = col.names)
# ind.meqtl.veh.df   <- fread(meqtl.veh.fn, col.names = col.names)
# ind.meqtl.delta.df <- fread(meqtl.delta.fn, col.names = col.names)
```

```{r}
setwd("~/bio/code/mpip/dex-stim-human-array/")

file.sources <- c("code/util.R",
                  "code/integrative/meqtl/04_me-qtl_analysis/04_04_me-qtl_snp_annotation/PlotChromHMMEnrichment.R",
                  "code/integrative/meqtl/04_me-qtl_analysis/04_04_me-qtl_snp_annotation/SavePPTXChromHMMEnrichmentPlot.R")
                  
sapply(file.sources, source, .GlobalEnv)
```

```{r include = F, eval = T}
# code: ../code/methylation/02_dma/01_dmp/03_get_cpg_gene_anno_dist
cpg.closest.genes.dist.fn <-paste0(dir.pre, "output/data/methylation/02_dmp/dex_cpgs_annotated_closest_genes_distances.csv")
cpg.closest.genes.dist.df <- fread(cpg.closest.genes.dist.fn)

delta.meqtl.cpg.anno.df               <- fread(paste0(out.dir.pre, "meqtls_cpg_annotated_delta.csv"), sep = "\t")
delta.meqtl.cpg.closest.genes.dist.df <- cpg.closest.genes.dist.df[PROBE_ID %in% delta.meqtl.cpg.anno.df$Name]

veh.meqtl.cpg.anno.df                 <- fread(paste0(out.dir.pre, "meqtls_cpg_annotated_veh.csv"), sep = "\t")
veh.meqtl.cpg.closest.genes.dist.df   <- cpg.closest.genes.dist.df[PROBE_ID %in% veh.meqtl.cpg.anno.df$Name]

dex.meqtl.cpg.anno.df                 <- fread(paste0(out.dir.pre, "meqtls_cpg_annotated_dex.csv"), sep = "\t")
dex.meqtl.cpg.closest.genes.dist.df   <- cpg.closest.genes.dist.df[PROBE_ID %in% dex.meqtl.cpg.anno.df$Name]
```

## Illumina {.tabset .tabset-fade .tabset-pills}

### Example of annotated CpGs {.tabset .tabset-fade .tabset-pills}

#### Full anno

```{r include = T, eval = T}
delta.meqtl.cpg.anno.df$Relation_to_Island <- factor(delta.meqtl.cpg.anno.df$Relation_to_Island, 
                                                     levels = c("N_Shelf", "N_Shore", "Island", "S_Shore", "S_Shelf", "OpenSea"))
delta.meqtl.cpg.anno.df[["Model"]] <- as.factor("Delta")

veh.meqtl.cpg.anno.df$Relation_to_Island <- factor(veh.meqtl.cpg.anno.df$Relation_to_Island,
                                                   levels = c("N_Shelf", "N_Shore", "Island", "S_Shore", "S_Shelf", "OpenSea"))
veh.meqtl.cpg.anno.df[["Model"]] <- as.factor("Baseline")

dex.meqtl.cpg.anno.df$Relation_to_Island <- factor(dex.meqtl.cpg.anno.df$Relation_to_Island,
                                                   levels = c("N_Shelf", "N_Shore", "Island", "S_Shore", "S_Shelf", "OpenSea"))
dex.meqtl.cpg.anno.df[["Model"]] <- as.factor("Dex")

meqtl.cpg.anno.df <- rbind(delta.meqtl.cpg.anno.df, veh.meqtl.cpg.anno.df, dex.meqtl.cpg.anno.df)

head(meqtl.cpg.anno.df, 100)
```

#### Closest genes

```{r include = T, eval = T}
head(cpg.closest.genes.dist.df, 100)
```

### Chromosomes

```{r out.width = "90%", fig.cap = "CpGs distribution across chromosomes", include = T, eval = T}
library(GenomicFeatures)
chr.length.tbl <- getChromInfoFromUCSC("hg19")[, c(1, 2)] %>% setDT()
chr.length.tbl <- chr.length.tbl[chrom %in% paste0("chr", 1:22)] #, "X", "Y"))]
colnames(chr.length.tbl) <- c("chr", "chr_size")
chr.length.tbl$chr <- as.factor(chr.length.tbl$chr)

chr.order <- paste("chr", c(1:22), sep = "")
meth      <- delta.meqtl.cpg.anno.df
meth$chr  <- factor(meth$chr,levels = chr.order)
meth      <- left_join(meth, chr.length.tbl)

library(janitor)
chr.cpgs.cnt <- tabyl(meth$chr, sort = T)
colnames(chr.cpgs.cnt) <- c("chr", "cnt_cpg", "freq_cpg")

chr.cpgs.cnt <- left_join(chr.cpgs.cnt, chr.length.tbl, by = "chr")
chr.cpgs.cnt$chr  <- factor(chr.cpgs.cnt$chr,levels = chr.order)
chr.cpgs.cnt[["freq_chr"]] <- chr.cpgs.cnt$chr_size / sum(chr.cpgs.cnt$chr_size)
chr.cpgs.cnt[["cpg_chr_cnt"]] <- chr.cpgs.cnt$cnt_cpg / chr.cpgs.cnt$freq_chr
chr.cpgs.cnt[["cpg_chr_freq"]] <- chr.cpgs.cnt$cpg_chr_cnt / sum(chr.cpgs.cnt$cpg_chr_cnt)

# Relative freq = (Number of CpGs on the chromosome (i) * (length of chromosomes{1:22} / length of the chromosome(i)) ) in %

ggplot(chr.cpgs.cnt, aes(x = chr, y = cpg_chr_freq, fill = chr)) + 
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_text(aes(label = scales::percent(cpg_chr_freq, accuracy = 0.1), y = cpg_chr_freq), 
            stat = "identity", vjust = -.5, size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(title = "Delta meQTLs CpGs distribution across the chromosomes wieghted by the chromosomes' length",
       x = "Chromosome",
       y = "Relative frequency") +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 8),
         axis.title = element_text(size = 8),
         axis.text.x = element_text(angle = 20, hjust = 0.5),
         legend.position = "none")
```

### Genomic regions

```{r out.width = "90%", fig.cap = "Distribution of locations of the CpGs in different genomic regions based on Illumina EPIC annotation", include = T, eval = T}
cbPalette <- c( "#CC6677", "#DDCC77", "#117733", "#AA4499", "#44AA99", "#888888")
plt.axis.size <- 8

ggplot(meqtl.cpg.anno.df, aes(x = Model, fill = Relation_to_Island)) + 
  geom_bar( position = "fill") + 
  geom_text(aes(by = Model, # scales::percent(..count../tapply(..count.., ..x.., sum), accuracy = 0.1), 
                y = (..count..)/sum(..count..)), 
            stat = "prop", 
            position = position_fill(vjust = .5), size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "Percentage of CpGs", 
       title = "") +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5)) +
   scale_fill_manual("Feature", values = cbPalette)
```

```{r out.width = "90%", fig.cap = "Distribution of locations of the CpGs in different genomic regions based on Illumina EPIC annotation", include = T, eval = T}
cbPalette <- c( "#CC6677", "#DDCC77", "#117733", "#AA4499", "#44AA99", "#888888")

ggplot(meqtl.cpg.anno.df[Model %in% c("Delta", "Baseline")], aes(x = Model, fill = Relation_to_Island)) + 
  geom_bar( position = "fill") + 
  geom_text(aes(by = Model, # scales::percent(..count../tapply(..count.., ..x.., sum), accuracy = 0.1), 
                y = (..count..)/sum(..count..)), 
            stat = "prop", 
            position = position_fill(vjust = .5), size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "Percentage of CpGs", 
       title = "") +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5)) +
   scale_fill_manual("Feature", values = cbPalette)
```

```{r out.width = "90%", fig.cap = "Distribution of locations of delta meQTL CpGs in different genomic regions based on Illumina EPIC annotation", include = T, eval = T}
cbPalette <- c( "#CC6677", "#DDCC77", "#117733", "#AA4499", "#44AA99", "#888888")

ggplot(meqtl.cpg.anno.df[Model %in% c("Delta")], aes(x = Model, fill = Relation_to_Island)) + 
  geom_bar( stat = "count") + 
  geom_text(aes(by = Model, # scales::percent(..count../tapply(..count.., ..x.., sum), accuracy = 0.1), 
                label = scales::percent((..count..)/sum(..count..), accuracy = 0.1)), 
                stat = "prop", 
                position = position_stack(vjust = 0.6, ), size = plt.axis.size / 2, color = "white") +
  coord_polar(theta = 'y', start = 0) +
  labs(x = " ",
       y = "Percentage of CpGs", 
       title = "") +
  theme(legend.position = "right",
        legend.title = element_blank(),
        legend.text = element_text(size = plt.axis.size),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = plt.axis.size),
        axis.text.x = element_blank()) +
  scale_fill_manual("Feature", values = cbPalette)
```

```{r}
meqtls.cpgs <- list(delta = delta.meqtl.cpg.anno.df$Name, 
                   veh = veh.meqtl.cpg.anno.df$Name,
                   dex = dex.meqtl.cpg.anno.df$Name)

delta <- setdiff(setdiff(meqtls.cpgs$delta, meqtls.cpgs$dex), meqtls.cpgs$veh)
veh <- setdiff(setdiff(meqtls.cpgs$veh, meqtls.cpgs$dex), meqtls.cpgs$delta)
dex <- setdiff(setdiff(meqtls.cpgs$dex, meqtls.cpgs$veh), meqtls.cpgs$delta)

meqtl.unique.cpg.anno.df <- rbind(  
  delta.meqtl.cpg.anno.df[Name %in% delta,],
  veh.meqtl.cpg.anno.df[Name %in% veh,],
  dex.meqtl.cpg.anno.df[Name %in% dex,]
)
```

```{r out.width = "90%", fig.cap = "Distribution of locations of the meCpGs in different genomic regions based on Illumina EPIC annotation", include = T, eval = T}
cbPalette <- c( "#CC6677", "#DDCC77", "#117733", "#AA4499", "#44AA99", "#888888")

ggplot(meqtl.unique.cpg.anno.df, aes(x = Model, fill = Relation_to_Island)) + 
  geom_bar( position = "fill") + 
  geom_text(aes(by = Model, # scales::percent(..count../tapply(..count.., ..x.., sum), accuracy = 0.1), 
                y = (..count..)/sum(..count..)), 
            stat = "prop", 
            position = position_fill(vjust = .5), size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "Percentage of CpGs", 
       title = "Unique meCpGs") +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 10),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5)) +
   scale_fill_manual("Feature", values = cbPalette)
```

### Gene regions

```{r include = F, eval = T}
delta.meqtl.cpg.closest.genes.dist.df$GeneGroup <- as.factor(delta.meqtl.cpg.closest.genes.dist.df$GeneGroup)

delta.meqtl.cpg.closest.genes.dist.df[["Model"]] <- as.factor("Delta")
veh.meqtl.cpg.closest.genes.dist.df[["Model"]]   <- as.factor("Baseline")
dex.meqtl.cpg.closest.genes.dist.df[["Model"]]   <- as.factor("Dex")

meqtl.cpg.closest.genes.dist.df <- rbind(delta.meqtl.cpg.closest.genes.dist.df,
                                         veh.meqtl.cpg.closest.genes.dist.df,
                                         dex.meqtl.cpg.closest.genes.dist.df)

meqtl.cpg.closest.genes.dist.df[is.na(meqtl.cpg.closest.genes.dist.df$GeneGroup), GeneGroup := "#NA"]
```

```{r out.width = "90%", fig.cap = "Distribution of gene-centric locations of the CpGs based on Illumina EPIC annotation", include = F, eval = F}
cbPalette <- c("#88CCEE", "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888")

ggplot(meqtl.cpg.closest.genes.dist.df, aes(x = Model, fill = GeneGroup)) + 
  geom_bar( position = "fill") + 
  geom_text(aes(by = Model), # label = scales::percent(..count../tapply(..count.., ..x.., sum), accuracy = 0.1)),
            stat = "prop", 
            position = position_fill(vjust = .5), size = 3) +
 # scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "Percentage of CpGs", 
       title = "Annotation to the closest gene based on IlluminaHumanMethylationEPICanno.ilm10b4.hg19 package") +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5)) +
   scale_fill_manual("Type", values = cbPalette)
```

```{r include = F, eval = T}
# Prepare methyl ranges

cpg.coord.df    <- delta.meqtl.cpg.anno.df[, .(Name, chr, pos)]# %>% dplyr:: mutate(chr = sub("chr", "", chr))
cpg.coord.range <- makeGRangesFromDataFrame(cpg.coord.df, 
                                            start.field = "pos", 
                                            end.field = "pos", 
                                            seqnames.field = c("chr"))
names(cpg.coord.range) <- cpg.coord.df$Name
```

Annotation with Chipseeker

```{r include = F, eval = T}
delta.meqtl.snp.anno.rds              <- readRDS(paste0(out.dir.pre, "meqtls_cpg_annotated_withChIPseeker_delta.rds"))
veh.meqtl.snp.anno.rds                <- readRDS(paste0(out.dir.pre, "meqtls_cpg_annotated_withChIPseeker_veh.rds"))
dex.meqtl.snp.anno.rds                <- readRDS(paste0(out.dir.pre, "meqtls_cpg_annotated_withChIPseeker_dex.rds"))
```

```{r include = F, eval = T}
delta.meqtl.snp.anno.stat.df <- delta.meqtl.snp.anno.rds@annoStat
veh.meqtl.snp.anno.stat.df   <- veh.meqtl.snp.anno.rds@annoStat
dex.meqtl.snp.anno.stat.df   <- dex.meqtl.snp.anno.rds@annoStat

delta.meqtl.snp.anno.stat.df[["Model"]] <- as.factor("Delta")
veh.meqtl.snp.anno.stat.df[["Model"]]   <- as.factor("Baseline")
dex.meqtl.snp.anno.stat.df[["Model"]]   <- as.factor("Dex")

meqtl.snp.anno.stat.df <- rbind(delta.meqtl.snp.anno.stat.df,
                                veh.meqtl.snp.anno.stat.df,
                                dex.meqtl.snp.anno.stat.df)

meqtl.snp.anno.stat.df$Feature   <- factor(meqtl.snp.anno.stat.df$Feature)
meqtl.snp.anno.stat.df$Frequency <- signif(meqtl.snp.anno.stat.df$Frequency, 3)
```

```{r out.width = "90%", fig.cap = "Distribution of gene-centric locations of the meCpGs based on Illumina EPIC annotation", include = T, eval = T}
cbPalette <- c("#88CCEE", "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888")

ggplot(meqtl.snp.anno.stat.df, aes(x = Model, y = Frequency, fill = Feature)) + 
  geom_bar( position = "fill", stat = "identity") + 
  geom_text(data = subset(meqtl.snp.anno.stat.df, Frequency > 2),
            aes(by = Model, label = Frequency), 
            position = position_fill(vjust = .5), size = 3) +
  labs(x = "",
       y = "Percentage of CpGs", 
       title = "meCpGs annotation from UCSC for the hg19 genome build using TxDb.Hsapiens.UCSC.hg19.knownGene and ChIPseeker Bioconductor R packages") +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5)) +
   scale_fill_manual("Feature", values = cbPalette)
```


```{r out.width = "90%", fig.cap = "Distribution of gene-centric locations of the delta meCpGs based on Illumina EPIC annotation", include = T, eval = T}
cbPalette <- c("#88CCEE", "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888")

ggplot(meqtl.snp.anno.stat.df[meqtl.snp.anno.stat.df$Model == "Delta", ], aes(x = Model, y = Frequency, fill = Feature)) + 
  geom_col() + 
  geom_text(
    data = subset(meqtl.snp.anno.stat.df[meqtl.snp.anno.stat.df$Model == "Delta", ], Frequency > 2),
    aes(by = Model, label = paste0(Frequency, "%")), position = position_stack(vjust = 0.8, ), size = plt.axis.size / 2, color = "white") +
  coord_polar(theta = 'y', start = 0) +
  labs(x = " ",
       y = "Percentage of CpGs", 
       title = "") +
  theme(legend.position = "right",
        legend.title = element_blank(),
        legend.text = element_text(size = plt.axis.size),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_blank()) +
  scale_fill_manual("Feature", values = cbPalette)
```

# __Enrichment of genomic locations in relation to island__
***

## Delta vs Baseline
```{r}
gen.loc.enrich.perm.rslt <- fread(paste0(dir.pre, "output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/meqtl_cpgs_relation_to_island_enrichment_perm.csv"))
```

```{r out.width = "90%", fig.cap = "Genomic region enrichment for the delta meCpGs.  The Y-axis denotes the fold enrichment/depletion as compared to baseline meCpGs. Green bars indicate significant enrichment/depletion, grey bars non-significant differences based on permutation Fisher-tests empirical P <= 0.05. ", eval = T, include = T}

gen.loc.enrich.perm.rslt[["is_sign"]] <- ifelse(gen.loc.enrich.perm.rslt$p_val_emp < 0.05, "Significant", "Non-significant")

ggplot(gen.loc.enrich.perm.rslt, aes(x = Feature, y = or.odds.ratio, fill = is_sign)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
  geom_hline(yintercept = 1, color = "red") + 
  labs(x = "",
       y = "Odds ratio", 
       title = "") +
  theme(legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = plt.axis.size),
        axis.text.x = element_text(angle = 0, hjust = 0.5, size = plt.axis.size),
        axis.text.y = element_text(size = plt.axis.size),
        legend.text = element_text(size = plt.axis.size)) +
   scale_fill_manual("", values = c( "#999999", "#009E73")) 
```

## Delta vs All
```{r}
gen.loc.enrich.perm.rslt <- fread("~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/meqtl_cpgs_relation_to_island_enrichment_perm_delta_vs_all.csv")
```

```{r out.width = "90%", fig.cap = "Genomic region enrichment for the delta meCpGs.  The Y-axis denotes the fold enrichment/depletion as compared to baseline meCpGs. Green bars indicate significant enrichment/depletion, grey bars non-significant differences based on permutation Fisher-tests empirical P <= 0.05. ", eval = T, include = T}

gen.loc.enrich.perm.rslt[["is_sign"]] <- ifelse(gen.loc.enrich.perm.rslt$p_val_emp < 0.05, "Significant", "Non-significant")

ggplot(gen.loc.enrich.perm.rslt, aes(x = Feature, y = or.odds.ratio, fill = is_sign)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
  geom_hline(yintercept = 1, color = "red") + 
  labs(x = "",
       y = "Odds ratio", 
       title = "") +
  theme(legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5)) +
   scale_fill_manual("", values = c( "#999999", "#009E73")) 
```

## Baseline vs All
```{r}
gen.loc.enrich.perm.rslt <- fread("~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/meqtl_cpgs_relation_to_island_enrichment_perm_veh_vs_all.csv")
```

```{r out.width = "90%", fig.cap = "Genomic region enrichment for the delta meCpGs.  The Y-axis denotes the fold enrichment/depletion as compared to baseline meCpGs. Green bars indicate significant enrichment/depletion, grey bars non-significant differences based on permutation Fisher-tests empirical P <= 0.05. ", eval = T, include = T}

gen.loc.enrich.perm.rslt[["is_sign"]] <- ifelse(gen.loc.enrich.perm.rslt$p_val_emp < 0.05, "Significant", "Non-significant")

ggplot(gen.loc.enrich.perm.rslt, aes(x = Feature, y = or.odds.ratio, fill = is_sign)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
  geom_hline(yintercept = 1, color = "red") + 
  labs(x = "",
       y = "Odds ratio", 
       title = "") +
  theme(legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5)) +
   scale_fill_manual("", values = c( "#999999", "#009E73")) 
```

# __Enrichment of gene-centric locations__
***

## Delta vs Baseline
```{r}
gen.loc.enrich.perm.rslt <- fread(paste0(dir.pre, "output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/meqtl_cpgs_chipseeker_enrichment_perm_delta_vs_veh.csv"))
```

```{r out.width = "90%", fig.cap = "Genomic region enrichment for the delta meCpGs.  The Y-axis denotes the fold enrichment/depletion as compared to baseline meCpGs. Green bars indicate significant enrichment/depletion, grey bars non-significant differences based on permutation Fisher-tests empirical P <= 0.05. ", eval = T, include = T}

gen.loc.enrich.perm.rslt[["is_sign"]] <- ifelse(gen.loc.enrich.perm.rslt$p_val_emp < 0.05, "Significant", "Non-significant")

ggplot(gen.loc.enrich.perm.rslt, aes(x = Feature, y = odds.ratio, fill = is_sign)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
  geom_hline(yintercept = 1, color = "red") + 
  labs(x = "",
       y = "Odds ratio", 
       title = "") +
  theme(legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = plt.axis.size),
        axis.text.x = element_text(angle = 20, hjust = 0.5, size = plt.axis.size),
        axis.text.y = element_text(size = plt.axis.size),
        legend.text = element_text(size = plt.axis.size))  +
   scale_fill_manual("", values = c( "#999999", "#009E73")) 
```

## Delta vs All
```{r}
gen.loc.enrich.perm.rslt <- fread(paste0(dir.pre, "output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/meqtl_cpgs_chipseeker_enrichment_perm_delta_vs_all.csv"))
```

```{r out.width = "90%", fig.cap = "Genomic region enrichment for the delta meCpGs.  The Y-axis denotes the fold enrichment/depletion as compared to baseline meCpGs. Green bars indicate significant enrichment/depletion, grey bars non-significant differences based on permutation Fisher-tests empirical P <= 0.05. ", eval = T, include = T}

gen.loc.enrich.perm.rslt[["is_sign"]] <- ifelse(gen.loc.enrich.perm.rslt$p_val_emp < 0.05, "Significant", "Non-significant")

ggplot(gen.loc.enrich.perm.rslt, aes(x = Feature, y = odds.ratio, fill = is_sign)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
  geom_hline(yintercept = 1, color = "red") + 
  labs(x = "",
       y = "Odds ratio", 
       title = "") +
  theme(legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = plt.axis.size),
        axis.text.x = element_text(angle = 20, hjust = 0.5, size = plt.axis.size),
        axis.text.y = element_text(size = plt.axis.size),
        legend.text = element_text(size = plt.axis.size))  +
   scale_fill_manual("", values = c( "#999999", "#009E73")) 
```

## Baseline vs All
```{r}
gen.loc.enrich.perm.rslt <- fread(paste0(dir.pre, "output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/meqtl_cpgs_chipseeker_enrichment_perm_veh_vs_all.csv"))
```

```{r out.width = "90%", fig.cap = "Genomic region enrichment for the delta meCpGs.  The Y-axis denotes the fold enrichment/depletion as compared to baseline meCpGs. Green bars indicate significant enrichment/depletion, grey bars non-significant differences based on permutation Fisher-tests empirical P <= 0.05. ", eval = T, include = T}

gen.loc.enrich.perm.rslt[["is_sign"]] <- ifelse(gen.loc.enrich.perm.rslt$p_val_emp < 0.05, "Significant", "Non-significant")

ggplot(gen.loc.enrich.perm.rslt, aes(x = Feature, y = odds.ratio, fill = is_sign)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
  geom_hline(yintercept = 1, color = "red") + 
  labs(x = "",
       y = "Odds ratio", 
       title = "") +
  theme(legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = plt.axis.size),
        axis.text.x = element_text(angle = 20, hjust = 0.5, size = plt.axis.size),
        axis.text.y = element_text(size = plt.axis.size),
        legend.text = element_text(size = plt.axis.size))  +
   scale_fill_manual("", values = c( "#999999", "#009E73")) 
```

# __ChromHMM Enrichment__
***

```{r eval = T, include = F}
chromhmm.blood.states <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromHMM_blood_states.Rds") 
chromhmm.brain.states <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromHMM_brain_states.Rds") 

states.lst <- elementMetadata(chromhmm.blood.states)[, "type"] %>% unique() %>% sort()

epigenome <- fread("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/epigenomes.tsv",
                   col.names = c("EID", "ORDER", "GROUP", "MNEMONIC", "NAME"))

chromhmm.states     <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromhmm_states.Rds") %>% setDT()

chromhmm.states[COLOR.NAME == "Silver", COLOR.NAME := "Grey"] 
chromhmm.states[COLOR.NAME == "White", COLOR.NAME := "Snow"] 

cbPalette <- chromhmm.states$COLOR.NAME
```

```{r}
out.dir.pre      <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/"
```

## Blood

Histone mark enrichment for the GR-induced meCpGs using 15 chromatin states from 18 blood and T- & B-cells types. The first figure denotes the fold enrichment / depletion as compared to baseline meCpGs, the second figure denotes significance based on the empirical P≤0.05 with 1,000 permutations.

```{r}
tissue <- "blood"

enrichment.rslts.fn <- paste0(out.dir.pre, "meqtl_cpgs_chromHMM_", tissue, "/", list.files(paste0(out.dir.pre, "meqtl_cpgs_chromHMM_", tissue)))

pptx.template <- read_pptx("~/bio/code/mpip/dex-stim-human-array/output/plots/enrichment/meqtl_snp_chromHMM_enrichment.pptx")
pptx.out      <- paste0("~/bio/code/mpip/dex-stim-human-array/output/plots/enrichment/meqtl_cpg_chromHMM_", tissue, "_enrichment.pptx")

enrich.plts <- SavePPTXChromHMMEnrichmentPlot(enrichment.rslts.fn, tissue, epigenome, pptx.template, pptx.out, plt.txt.size = plt.axis.size * 1.5)
```

```{r include = T, fig.cap="The figure denotes the fold enrichment / depletion as compared to baseline meCpGs"}
enrich.plts$or
```

```{r include = T, fig.cap="The figure denotes significance based on the empirical P≤0.05 with 1,000 permutations"}
enrich.plts$pval
```

<!-- Next two chuncks prepare data for further analysis -->

```{r read-chromhmm-anno-delta, eval = T, include = F}
out.dir.pre      <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/meqtls/region_wise_independent_snps/"

delta.meqtls.cpg.chromhmm.annotated.df <- fread(paste0(out.dir.pre, "meqtls_cpg_chromhmm_annotated_delta.csv"))

delta.meqtls.cpg.chromhmm.annotated.df[, Model := "Delta"]
```

```{r eval = F}
  enrichment.all.df <- lapply(seq_along(enrichment.rslts.fn), function(i) {
    eid   <- sub(".csv", "", sub(".*_", "", enrichment.rslts.fn[i]))
    ename <- epigenome[EID == eid, NAME] 
    # ename <- paste0(eid, ": ", ename)
    
    df <- data.frame(read.csv2(enrichment.rslts.fn[i]), eid = eid, ename = ename) %>% setDT()
    df <- df[p_val_emp <= 0.05]
    }) %>% bind_rows()

chromhmm.cpg.lst.df <- left_join(enrichment.all.df[, c("eid", "ename", "state")], 
                                 delta.meqtls.cpg.chromhmm.annotated.df[, .(CpG_ID, annot.code, annot.type)],
                                 by = c("eid" = "annot.code", "state" = "annot.type"))

chromhmm.cpg.lst.df  <- split(chromhmm.cpg.lst.df, by = "state")

saveRDS(chromhmm.cpg.lst.df, 
         "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/dex_meqtl_cpgs_by_chromhmm_state_blood_significant.rds")

chromhmm.states.of.interest <- c("2_TssAFlnk", "3_TxFlnk", "4_Tx", "5_TxWk", "6_EnhG", "7_Enh", "10_TssBiv", "14_ReprPCWk")
```

# __IDEAS Enrichment__
***

```{r eval = T, include = F}
ideas.blood.states <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/public_data/ENCODE/IDEAS/all.blood.tissues.ideas.states.rds") 

states.lst <- elementMetadata(ideas.blood.states)[, "name"] %>% unique() %>% sort()

epigenome <- fread("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/epigenomes.tsv",
                   col.names = c("EID", "ORDER", "GROUP", "MNEMONIC", "NAME"))
```

```{r}
out.dir.pre      <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/"
```

## Blood

Histone mark enrichment for the GR-induced meCpGs using condensed annotation of 20 chromatin states from 18 blood and T- & B-cells types. The first figure denotes the fold enrichment / depletion as compared to baseline meCpGs, the second figure denotes significance based on the empirical P≤0.05 with 1,000 permutations.

```{r}
tissue <- "blood"

enrichment.rslts.fn <- paste0(out.dir.pre, "meqtl_cpgs_IDEAS_", tissue, "/", list.files(paste0(out.dir.pre, "meqtl_cpgs_IDEAS_", tissue)))

pptx.template <- read_pptx("~/bio/code/mpip/dex-stim-human-array/output/plots/enrichment/meqtl_snp_chromHMM_enrichment.pptx")
pptx.out      <- paste0("~/bio/code/mpip/dex-stim-human-array/output/plots/enrichment/meqtl_cpg_IDEAS_", tissue, "_enrichment.pptx")

enrich.plts <- SavePPTXChromHMMEnrichmentPlot(enrichment.rslts.fn, tissue, epigenome, pptx.template, pptx.out, plt.txt.size = plt.axis.size * 1.5)
```

```{r include = T, fig.cap="The figure denotes the fold enrichment / depletion as compared to baseline meCpGs"}
enrich.plts$or
```

```{r include = T, fig.cap="The figure denotes significance based on the empirical P≤0.05 with 1,000 permutations"}
enrich.plts$pval
```
