---
title: "cis-meQTLs analysis: genotype with missingness"
subtitle: "Dexamethasone-Stimulated Human Array Project"
author: 
  - name: Anastasiia Hryhorzhevska
    email: anastasiia_hry@psych.mpg.de
    url: https://github.com/ahryho
    affiliation: Max Planck Institute of Psychiatry
    affiliation_url: https://www.psych.mpg.de/2664393/medizinische-genomforschung
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    css: ../../../style.css
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
options(digits = 4, width = 100, stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE,
                      tidy.opts = list(width.cutoff = 100),
                      tidy=TRUE,
                      fig.pos = "H",
                      dpi = 200,
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE, 
                      cache.lazy = FALSE)

# define DT as the standard printing method for data.frame
library(knitr)
library(DT)
knit_print.data.frame <- function(x, ...) {
  knit_print(DT::datatable(x,
                           filter="top",
                           rownames = FALSE,
                           extensions = "FixedColumns",
                           options = list(
                             scrollX = TRUE,
                             pageLength = 5
                             
                           )), ...)
}

registerS3method("knit_print", "data.frame", knit_print.data.frame)

library(data.table)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(scales)
library(corrplot)
library(factoextra)
library(viridis)
library(knitr)
```

```{r include = F, eval = T}
source("~/bio/code/mpip/dex-stim-human-array/code/integrative/util.R")
```

***
# meQTL analysis: SNPs with missingness
***

## Methodology


All meQTLs passed the significance threshold of FDR = 0.05. 

```{r include = F, eval = T}
cbPalette <- c("grey", "#0072B2")
```

```{r, include = F, eval = T}
out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/meqtls/with_missingness/"
setwd(out.dir.pre)

meqtl.veh.fn   <- "me-qtl_cis_result_snps_with_na_veh_beta_fdr_005.csv"
meqtl.delta.fn <- "me-qtl_cis_result_snps_with_na_delta_beta_fdr_005.csv"

col.names <-  c("SNP", "CpG_ID", "beta", "t-stat", "p-value", "fdr")

ind.meqtl.veh.df   <- fread(meqtl.veh.fn, col.names = col.names)
ind.meqtl.delta.df <- fread(meqtl.delta.fn, col.names = col.names)
```

```{r, include = F, eval = T}
meqtl.all.full.df <- rbind(
                           ind.meqtl.veh.df[, treatment := "baseline"],
                           ind.meqtl.delta.df[, treatment := "delta"])

meqtl.all.full.df <- meqtl.all.full.df[!is.na(meqtl.all.full.df$fdr),]
meqtl.all.full.df[["meQTL_ID"]] <- paste(meqtl.all.full.df$SNP, meqtl.all.full.df$CpG_ID, sep = "-")
meqtl.all.full.df <- setDT(meqtl.all.full.df)
```

```{r include = F, eval = T}
freq <- fread("/Users/anastasiia_hry/bio/code/mpip/dex-stim-human-array/data/snps/imputed_qc/from_janine/qc/freq.afreq")
```

## Significant Hits, FDR < 0.05

```{r, include = F, eval = T}
sign.hits.df <- data.table(c(ind.meqtl.delta.df$fdr  %>%  length(), 
                             ind.meqtl.delta.df$CpG_ID %>% unique() %>% length(),
                             ind.meqtl.delta.df$SNP %>% unique() %>% length()),
                           c(ind.meqtl.veh.df$fdr  %>% length(), 
                             ind.meqtl.veh.df$CpG_ID %>% unique() %>% length(),
                             ind.meqtl.veh.df$SNP %>% unique() %>% length()))

sign.hits.df <- as.data.frame(sign.hits.df)
rownames(sign.hits.df) <- c("cis-meQTLs", "CpG", "SNP")
colnames(sign.hits.df) <- c("delta", "baseline")
```

```{r out.width = "95%", include = T, eval = T}
x <- sign.hits.df %>% as.matrix() %>% 
  reshape2::melt(var.names = c("ID", "Treatment"), value.name = "cnt") %>% setDT()
x[Var2 %in% c("veh", "base"), Var2 := "baseline"]
x[Var2 %in% c("delta", "delta GR-induced"), Var2 := "GR-induced"]
x %>%
  ggplot(aes(y = cnt, x = Var2, fill = as.factor(as.numeric(Var2)))) +
    geom_col() +
    geom_text(aes(label = comma(cnt, accuracy = 1L), y = cnt), 
                  stat = "identity", 
                  vjust = -0.2, size = 2.5) + 
    scale_y_continuous(labels = scientific) +
    # scale_fill_viridis(discrete = TRUE, alpha = 0.9, option = "C") +
    labs(title = " ", y = "Count", x = " ") + 
    facet_wrap(~ Var1, ncol = 3) +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank()) +
  scale_fill_manual(values = cbPalette[c(1, 2)])
```

## Upset plots{.tabset .tabset-fade .tabset-pills}

### meQTL level
  
```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

meqtls.meqtls <- list(delta = meqtl.all.full.df[treatment == "delta", meQTL_ID], 
                      baseline = meqtl.all.full.df[treatment == "baseline", meQTL_ID])

upset(fromList(meqtls.meqtls), 
      sets = c("delta", "baseline"),
      nsets = 2, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, meQTL", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = cbPalette[2:1], 
      order.by = "freq")
```

## Scatter plot

```{r out.width = "95%", include = T, eval = T}
fdr.thr <- 0.05
plot.title <- paste0("Significant at FDR <= ", fdr.thr, " meQTLs.",
                     "\nResult of MatrixEQTL")
GetScatterPlot2(meqtl.all.full.df, plot.title = plot.title, cbPalette = cbPalette)
```

## Upset plots{.tabset .tabset-fade .tabset-pills}

### meQTL level
  
```{r out.width = "95%", include = T, eval = T}
meqtls.meqtls <- list(delta = meqtl.all.full.df[treatment == "delta", meQTL_ID], 
                      veh = meqtl.all.full.df[treatment == "baseline", meQTL_ID])

intersect.meqtls         <- intersect(meqtls.meqtls$delta, meqtls.meqtls$veh)
perc.olap.delta.with.veh <- scales::percent(length(intersect.meqtls) / length(meqtls.meqtls$delta), accuracy = 0.1)
perc.olap.veh.with.delta <- scales::percent(length(intersect.meqtls) / length(meqtls.meqtls$veh), accuracy = 0.1)

plot.title <- "Number of intersections between delta and baseline meQTLs"

ggVennDiagram::ggVennDiagram(meqtls.meqtls, 
              category.names = c(paste0("GR-response, ", perc.olap.delta.with.veh), 
                                 paste0("Baseline, ", perc.olap.veh.with.delta)), 
              label_alpha = 0.7,
              edge_size = 0,
              set_geom = "text",
              set_color = "black",
              label = "count") +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10, face="italic")) +
    labs(x = "", y = "",
         title = plot.title) +
    scale_color_manual(values = cbPalette) +
    scale_fill_gradient(low = alpha(cbPalette[2], 0.5), high = alpha(cbPalette[1], 0.5))
```
```{r}
library(eulerr)
plot(euler(meqtls.meqtls),labels =  c(paste0("Delta, ", perc.olap.delta.with.veh), 
                                 paste0("Baseline, ", perc.olap.veh.with.delta)),
      
     fills = list(fill = c(alpha(cbPalette[2], 0.5), alpha(cbPalette[1], 0.5)), alpha = 0.7),
     edges = list(col = "white", lex = 2),
     quantities = list(type = c("counts"), cex = 1))
```

### SNP level
  
```{r out.width = "95%", include = T, eval = T}
meqtls.snps <- list(delta = ind.meqtl.delta.df$SNP, 
                    veh = ind.meqtl.veh.df$SNP)

intersect                <- intersect(meqtls.snps$delta, meqtls.snps$veh)
perc.olap.delta.with.veh <- scales::percent(length(intersect) / length(meqtls.snps$delta), accuracy = 0.1)
perc.olap.veh.with.delta <- scales::percent(length(intersect) / length(meqtls.snps$veh), accuracy = 0.1)

plot.title <- "Number of intersections between delta and baseline meQTL SNPs"

ggVennDiagram::ggVennDiagram(meqtls.snps, 
              category.names = c(paste0("GR-response, ", perc.olap.delta.with.veh), 
                                 paste0("Baseline, ", perc.olap.veh.with.delta)), 
              label_alpha = 0.7,
              edge_size = 0,
              set_geom = "text",
              set_color = "black",
              label = "count") +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10, face="italic")) +
    labs(x = "", y = "",
         title = plot.title) +
    scale_color_manual(values = cbPalette) +
    scale_fill_gradient(low = alpha(cbPalette[2], 0.5), high = alpha(cbPalette[1], 0.5))
```

### CpG level

```{r out.width = "95%", include = T, eval = T}
meqtls.cpg <- list(delta = ind.meqtl.delta.df$CpG_ID, 
                   veh = ind.meqtl.veh.df$CpG_ID)

intersect                <- intersect(meqtls.cpg$delta, meqtls.cpg$veh)
perc.olap.delta.with.veh <- scales::percent(length(intersect) / length(meqtls.cpg$delta), accuracy = 0.1)
perc.olap.veh.with.delta <- scales::percent(length(intersect) / length(meqtls.cpg$veh), accuracy = 0.1)

plot.title <- "Number of intersections between delta and baseline meQTL CpGs"

ggVennDiagram::ggVennDiagram(meqtls.cpg, 
              category.names = c(paste0("GR-response, ", perc.olap.delta.with.veh), 
                                 paste0("Baseline, ", perc.olap.veh.with.delta)), 
              label_alpha = 0.7,
              edge_size = 0,
              set_geom = "text",
              set_color = "black",
              label = "count") +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10, face="italic")) +
    labs(x = "", y = "",
         title = plot.title) +
    scale_color_manual(values = cbPalette) +
    scale_fill_gradient(low = alpha(cbPalette[2], 0.5), high = alpha(cbPalette[1], 0.5))
```
