---
title: "cis-meQTLs analysis: Comparison between approaches. Clumped results"
subtitle: "Dexamethasone-Stimulated Human Array Project"
author: 
  - name: Anastasiia Hryhorzhevska
    email: anastasiia_hry@psych.mpg.de
    url: https://github.com/ahryho
    affiliation: Max Planck Institute of Psychiatry
    affiliation_url: https://www.psych.mpg.de/2664393/medizinische-genomforschung
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    css: ../../../style.css
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
options(digits = 4, width = 100, stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE,
                      tidy.opts = list(width.cutoff = 100),
                      tidy=TRUE,
                      fig.pos = "H",
                      dpi = 200,
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE, 
                      cache.lazy = FALSE)

# define DT as the standard printing method for data.frame
library(knitr)
library(DT)
knit_print.data.frame <- function(x, ...) {
  knit_print(DT::datatable(x,
                           filter="top",
                           rownames = FALSE,
                           extensions = "FixedColumns",
                           options = list(
                             scrollX = TRUE,
                             pageLength = 5
                             
                           )), ...)
}

registerS3method("knit_print", "data.frame", knit_print.data.frame)

library(data.table)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(scales)
library(corrplot)
library(viridis)
library(knitr)
library(eulerr)

library(here)
```

```{r include = F, eval = T}
source("~/mpip/projects/dex-stim-human-array/code/integrative/util.R")
```

# Introduction

This markdown provides comparison between two meQTL analysis after LD clumping. meQTL analysis were done using the R package _MatrixEQTL_. LD clumping was done using PLINK

For all analysis, the same DNAm data were used which underwent following QC steps: 

![DNAm QC](../../../data_overview/img/dnam_qc_roadmap.pdf){width=95%}

The genotype data were different for each analysis:

1. The SNP data were used where the uncertain genotype was replaced with the most likely genotype

2. SNP probabilities, i.e. dosages, were used

For all three analysis, the imputed QC __SNP data__ (no LD pruning, the MHC regions are included) were used. The data includes __3'957'338 SNPs__.

All meQTLs passed the significance threshold of FDR = 0.05. 

LD clumping was done per CpG using the following parameters:

  --clump-p1 = 0.05:           Significance threshold for index SNPs
   
  --clump-p2 = 1:              Secondary significance threshold for clumped SNPs
 
  --clump-r2 = 0.2:            LD threshold for clumping

  --clump-kb = 200:            Physical distance threshold for clumping

# meQTL analysis: using SNPs' probabilities

```{r include = F, eval = T}
cbPalette <- c("grey", "#0072B2")
```

```{r, include = F, eval = T}
out.dir.pre  <- "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/output/data/integrative/matrixEQTL/meqtls/snp_probabilities/clumped/"
setwd(out.dir.pre)

meqtl.veh.fn   <- "me-qtl_cis_result_veh_snp_probabilities_beta_fdr_005.csv"
meqtl.delta.fn <- "me-qtl_cis_result_delta_snp_probabilities_beta_fdr_005.csv"

col.names <-  c("SNP", "CpG_ID", "beta", "t-stat", "p-value", "fdr")

meqtl.veh.df   <- fread(meqtl.veh.fn, col.names = col.names)
meqtl.delta.df <- fread(meqtl.delta.fn, col.names = col.names)
```

```{r, include = F, eval = T}
meqtl.all.prob.df <- rbind(meqtl.veh.df[, treatment := "baseline"],
                           meqtl.delta.df[, treatment := "delta"]) %>% setDT()

meqtl.all.prob.df <- meqtl.all.prob.df[!is.na(fdr)]
meqtl.all.prob.df[, meQTL_ID := paste(SNP, CpG_ID, sep = "-")]
```

## Significant Hits, FDR < 0.05

```{r, include = F, eval = T}
sign.hits.df <- data.table(c(meqtl.all.prob.df[treatment == "delta", meQTL_ID]  %>%  length(), 
                             meqtl.delta.df$CpG_ID %>% unique() %>% length(),
                             meqtl.delta.df$SNP %>% unique() %>% length()),
                           c(meqtl.all.prob.df[treatment == "baseline", meQTL_ID]  %>%  length(), 
                             meqtl.veh.df$CpG_ID %>% unique() %>% length(),
                             meqtl.veh.df$SNP %>% unique() %>% length()))

sign.hits.df <- as.data.frame(sign.hits.df)
rownames(sign.hits.df) <- c("cis-meQTLs", "CpG", "SNP")
colnames(sign.hits.df) <- c("delta", "baseline")
```

```{r out.width = "95%", include = F, eval = T}
x <- sign.hits.df %>% as.matrix() %>% 
  reshape2::melt(var.names = c("ID", "treatment"), value.name = "cnt") %>% setDT()
x[Var2 %in% c("veh", "base"), Var2 := "baseline"]
x[Var2 %in% c("delta", "delta GR-induced"), Var2 := "GR-induced"]

```

```{r}
x %>% 
  ggplot(aes(y = cnt, x = Var2, fill = as.factor(as.numeric(Var2)))) +
    geom_col() +
    geom_text(aes(label = comma(cnt, accuracy = 1L), y = cnt), 
                  stat = "identity", 
                  vjust = -0.2, size = 2.5) + 
    scale_y_continuous(labels = scientific) +
    # scale_fill_viridis(discrete = TRUE, alpha = 0.9, option = "C") +
    labs(title = " ", y = "Count", x = " ") + 
    facet_wrap(~ Var1, ncol = 3) +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank()) +
  scale_fill_manual(values = cbPalette[c(1, 2)])
```

## Scatter plot

```{r out.width = "95%", include = T, eval = T}
fdr.thr <- 0.05
plot.title <- paste0("Significant at FDR <= ", fdr.thr, " meQTLs.",
                     "\nResult of MatrixEQTL")
GetScatterPlot2(meqtl.all.prob.df, plot.title = plot.title, cbPalette = cbPalette)
```

## Upset plots{.tabset .tabset-fade .tabset-pills}

### meQTL level
  
```{r out.width = "95%", include = T, eval = T}
meqtls.meqtls <- list(delta = meqtl.all.prob.df[treatment == "delta", meQTL_ID], 
                      veh = meqtl.all.prob.df[treatment == "baseline", meQTL_ID])

intersect.meqtls         <- intersect(meqtls.meqtls$delta, meqtls.meqtls$veh)
perc.olap.delta.with.veh <- scales::percent(length(intersect.meqtls) / length(meqtls.meqtls$delta), accuracy = 0.1)
perc.olap.veh.with.delta <- scales::percent(length(intersect.meqtls) / length(meqtls.meqtls$veh), accuracy = 0.1)

plot.title <- "Number of intersections between delta and baseline meQTLs"

ggVennDiagram::ggVennDiagram(meqtls.meqtls, 
              category.names = c(paste0("GR-response, ", perc.olap.delta.with.veh), 
                                 paste0("Baseline, ", perc.olap.veh.with.delta)), 
              label_alpha = 0.7,
              edge_size = 0,
              set_geom = "text",
              set_color = "black",
              label = "count") +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10, face="italic")) +
    labs(x = "", y = "",
         title = plot.title) +
    scale_color_manual(values = cbPalette) +
    scale_fill_gradient(low = alpha(cbPalette[2], 0.5), high = alpha(cbPalette[1], 0.5))
```

### SNP level
  
```{r out.width = "95%", include = T, eval = T}
meqtls.snps <- list(delta = meqtl.all.prob.df[treatment == "delta", SNP], 
                      veh = meqtl.all.prob.df[treatment == "baseline", SNP])

intersect                <- intersect(meqtls.snps$delta, meqtls.snps$veh)
perc.olap.delta.with.veh <- scales::percent(length(intersect) / length(meqtls.snps$delta), accuracy = 0.1)
perc.olap.veh.with.delta <- scales::percent(length(intersect) / length(meqtls.snps$veh), accuracy = 0.1)

plot.title <- "Number of intersections between delta and baseline meQTL SNPs"

ggVennDiagram::ggVennDiagram(meqtls.snps, 
              category.names = c(paste0("GR-response, ", perc.olap.delta.with.veh), 
                                 paste0("Baseline, ", perc.olap.veh.with.delta)), 
              label_alpha = 0.7,
              edge_size = 0,
              set_geom = "text",
              set_color = "black",
              label = "count") +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10, face="italic")) +
    labs(x = "", y = "",
         title = plot.title) +
    scale_color_manual(values = cbPalette) +
    scale_fill_gradient(low = alpha(cbPalette[2], 0.5), high = alpha(cbPalette[1], 0.5))
```

### CpG level

```{r out.width = "95%", include = T, eval = T}
meqtls.cpg <- list(delta = meqtl.all.prob.df[treatment == "delta", CpG_ID], 
                      veh = meqtl.all.prob.df[treatment == "baseline", CpG_ID])

intersect                <- intersect(meqtls.cpg$delta, meqtls.cpg$veh)
perc.olap.delta.with.veh <- scales::percent(length(intersect) / length(meqtls.cpg$delta), accuracy = 0.1)
perc.olap.veh.with.delta <- scales::percent(length(intersect) / length(meqtls.cpg$veh), accuracy = 0.1)

plot.title <- "Number of intersections between delta and baseline meQTL CpGs"

ggVennDiagram::ggVennDiagram(meqtls.cpg, 
              category.names = c(paste0("GR-response, ", perc.olap.delta.with.veh), 
                                 paste0("Baseline, ", perc.olap.veh.with.delta)), 
              label_alpha = 0.7,
              edge_size = 0,
              set_geom = "text",
              set_color = "black",
              label = "count") +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10, face="italic")) +
    labs(x = "", y = "",
         title = plot.title) +
    scale_color_manual(values = cbPalette) +
    scale_fill_gradient(low = alpha(cbPalette[2], 0.5), high = alpha(cbPalette[1], 0.5))
```

# meQTLs analysis: using SNPs with missingness

```{r, include = F, eval = T}
out.dir.pre  <- "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/output/data/integrative/matrixEQTL/meqtls/with_missingness/clumped/"
setwd(out.dir.pre)

meqtl.veh.fn   <- "me-qtl_cis_result_snps_with_na_veh_beta_fdr_005.csv"
meqtl.delta.fn <- "me-qtl_cis_result_snps_with_na_delta_beta_fdr_005.csv"

col.names <-  c("SNP", "CpG_ID", "beta", "t-stat", "p-value", "fdr")

meqtl.veh.df   <- fread(meqtl.veh.fn, col.names = col.names)
meqtl.delta.df <- fread(meqtl.delta.fn, col.names = col.names)
```

```{r, include = F, eval = T}
meqtl.all.na.df <- rbind(meqtl.veh.df[, treatment := "baseline"],
                           meqtl.delta.df[, treatment := "delta"]) %>% setDT()

meqtl.all.na.df <- meqtl.all.na.df[!is.na(fdr)]
meqtl.all.na.df[, meQTL_ID := paste(SNP, CpG_ID, sep = "-")]
```

## Significant Hits, FDR < 0.05

```{r, include = F, eval = T}
sign.hits.df <- data.table(c(meqtl.all.na.df[treatment == "delta", meQTL_ID]  %>%  length(), 
                             meqtl.delta.df$CpG_ID %>% unique() %>% length(),
                             meqtl.delta.df$SNP %>% unique() %>% length()),
                           c(meqtl.all.na.df[treatment == "baseline", meQTL_ID]  %>%  length(), 
                             meqtl.veh.df$CpG_ID %>% unique() %>% length(),
                             meqtl.veh.df$SNP %>% unique() %>% length()))

sign.hits.df <- as.data.frame(sign.hits.df)
rownames(sign.hits.df) <- c("cis-meQTLs", "CpG", "SNP")
colnames(sign.hits.df) <- c("delta", "baseline")
```

```{r out.width = "95%", include = F, eval = T}
x <- sign.hits.df %>% as.matrix() %>% 
  reshape2::melt(var.names = c("ID", "treatment"), value.name = "cnt") %>% setDT()
x[Var2 %in% c("veh", "base"), Var2 := "baseline"]
x[Var2 %in% c("delta", "delta GR-induced"), Var2 := "GR-induced"]
```

```{r}
x %>% 
  ggplot(aes(y = cnt, x = Var2, fill = as.factor(as.numeric(Var2)))) +
    geom_col() +
    geom_text(aes(label = comma(cnt, accuracy = 1L), y = cnt), 
                  stat = "identity", 
                  vjust = -0.2, size = 2.5) + 
    scale_y_continuous(labels = scientific) +
    # scale_fill_viridis(discrete = TRUE, alpha = 0.9, option = "C") +
    labs(title = " ", y = "Count", x = " ") + 
    facet_wrap(~ Var1, ncol = 3) +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank()) +
  scale_fill_manual(values = cbPalette[c(1, 2)])
```

# meQTLs analysis: using the most likely genotype (mode)

```{r, include = F, eval = T}
out.dir.pre  <- "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/output/data/integrative/matrixEQTL/meqtls/region_wise_independent_snps/"

setwd(out.dir.pre)

meqtl.veh.fn   <- "me-qtl_cis_indp_rw_veh_fdr_005.csv"
meqtl.delta.fn <- "me-qtl_cis_indp_rw_delta_fdr_005.csv"

col.names <-  c("CpG_ID", "SNP", "beta", "t-stat", "p-value", "fdr")

meqtl.veh.df   <- fread(meqtl.veh.fn, col.names = col.names)
meqtl.delta.df <- fread(meqtl.delta.fn, col.names = col.names)
```

```{r, include = F, eval = T}
meqtl.all.mode.df <- rbind(meqtl.veh.df[, treatment := "baseline"],
                           meqtl.delta.df[, treatment := "delta"]) %>% setDT()

meqtl.all.mode.df <- meqtl.all.mode.df[!is.na(fdr)]
meqtl.all.mode.df[, meQTL_ID := paste(SNP, CpG_ID, sep = "-")]
```

## Significant Hits, FDR < 0.05

```{r, include = F, eval = T}
sign.hits.df <- data.table(c(meqtl.all.mode.df[treatment == "delta", meQTL_ID]  %>%  length(), 
                             meqtl.delta.df$CpG_ID %>% unique() %>% length(),
                             meqtl.delta.df$SNP %>% unique() %>% length()),
                           c(meqtl.all.mode.df[treatment == "baseline", meQTL_ID]  %>%  length(), 
                             meqtl.veh.df$CpG_ID %>% unique() %>% length(),
                             meqtl.veh.df$SNP %>% unique() %>% length()))

sign.hits.df <- as.data.frame(sign.hits.df)
rownames(sign.hits.df) <- c("cis-meQTLs", "CpG", "SNP")
colnames(sign.hits.df) <- c("delta", "baseline")
```

```{r out.width = "95%", include = F, eval = T}
x <- sign.hits.df %>% as.matrix() %>% 
  reshape2::melt(var.names = c("ID", "treatment"), value.name = "cnt") %>% setDT()
x[Var2 %in% c("veh", "base"), Var2 := "baseline"]
x[Var2 %in% c("delta", "delta GR-induced"), Var2 := "GR-induced"]
```

```{r}
x %>% 
  ggplot(aes(y = cnt, x = Var2, fill = as.factor(as.numeric(Var2)))) +
    geom_col() +
    geom_text(aes(label = comma(cnt, accuracy = 1L), y = cnt), 
                  stat = "identity", 
                  vjust = -0.2, size = 2.5) + 
    scale_y_continuous(labels = scientific) +
    # scale_fill_viridis(discrete = TRUE, alpha = 0.9, option = "C") +
    labs(title = " ", y = "Count", x = " ") + 
    facet_wrap(~ Var1, ncol = 3) +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank()) +
  scale_fill_manual(values = cbPalette[c(1, 2)])
```

# Overlap between meQTLs from different approaches

## Probabilities vs missigness

<!-- meQTLs -->

```{r}
meqtls.delta <- list(prob = meqtl.all.prob.df[treatment == "delta", meQTL_ID], 
                     miss = meqtl.all.na.df[treatment == "delta", meQTL_ID])

meqtls.veh   <- list(prob = meqtl.all.prob.df[treatment == "baseline", meQTL_ID], 
                     miss = meqtl.all.na.df[treatment == "baseline", meQTL_ID])
```

### delta GR-induced meQTLs

```{r}
euler.tbl <- euler(meqtls.delta)

plot(euler.tbl,
     quantities = list(type = c("counts", "percent")),
     fills = c("lightgrey", "brown"),
     labels =  c("probabilities", "missigness"))
```

### baseline meQTLs

```{r}
euler.tbl <- euler(meqtls.veh)

plot(euler.tbl,
     quantities = list(type = c("counts", "percent")),
     fills = c("lightgrey", "brown"),
     labels =  c("probabilities", "missigness"))
```

<!-- meCpGs -->

```{r}
mecpgs.delta <- list(prob = meqtl.all.prob.df[treatment == "delta", CpG_ID] %>% unique(), 
                     miss = meqtl.all.na.df[treatment == "delta", CpG_ID] %>% unique())

mecpgs.veh   <- list(prob = meqtl.all.prob.df[treatment == "baseline", CpG_ID] %>% unique(), 
                     miss = meqtl.all.na.df[treatment == "baseline", CpG_ID] %>% unique())
```

### delta GR-induced meCpGs

```{r}
euler.tbl <- euler(mecpgs.delta)

plot(euler.tbl,
     quantities = list(type = c("counts", "percent")),
     fills = c("lightgrey", "brown"),
     labels =  c("probabilities", "missigness"))
```

### baseline meCpGs

```{r}
euler.tbl <- euler(mecpgs.veh)

plot(euler.tbl,
     quantities = list(type = c("counts", "percent")),
     fills = c("lightgrey", "brown"),
     labels =  c("probabilities", "missigness"))
```

## Probabilities vs most likely genotype

```{r}
meqtls.delta <- list(prob = meqtl.all.prob.df[treatment == "delta", meQTL_ID], 
                     mode = meqtl.all.mode.df[treatment == "delta", meQTL_ID])

meqtls.veh   <- list(prob = meqtl.all.prob.df[treatment == "baseline", meQTL_ID], 
                     mode = meqtl.all.mode.df[treatment == "baseline", meQTL_ID])
```

### delta GR-induced meQTLs

```{r}
euler.tbl <- euler(meqtls.delta)

plot(euler.tbl,
     quantities = list(type = c("counts", "percent")),
     fills = c("lightgrey", "brown"),
     labels =  c("probabilities", " most likely genotype"))
```

### baseline meQTLs

```{r}
euler.tbl <- euler(meqtls.veh)

plot(euler.tbl,
     quantities = list(type = c("counts", "percent")),
     fills = c("lightgrey", "brown"),
     labels =  c("probabilities", " most likely genotype"))
```

<!-- meCpGs -->

```{r}
mecpgs.delta <- list(prob = meqtl.all.prob.df[treatment == "delta", CpG_ID] %>% unique(), 
                     mode = meqtl.all.mode.df[treatment == "delta", CpG_ID] %>% unique())

mecpgs.veh   <- list(prob = meqtl.all.prob.df[treatment == "baseline", CpG_ID] %>% unique(), 
                     mode = meqtl.all.mode.df[treatment == "baseline", CpG_ID] %>% unique())
```

### delta GR-induced meCpGs

```{r}
euler.tbl <- euler(mecpgs.delta)

plot(euler.tbl,
     quantities = list(type = c("counts", "percent")),
     fills = c("lightgrey", "brown"),
     labels =  c("probabilities", " most likely genotype"))
```

### baseline meCpGs

```{r}
euler.tbl <- euler(mecpgs.veh)

plot(euler.tbl,
     quantities = list(type = c("counts", "percent")),
     fills = c("lightgrey", "brown"),
     labels =  c("probabilities", " most likely genotype"))
```

<!-- meSNPs -->

```{r}
mesnps.delta <- list(prob = meqtl.all.prob.df[treatment == "delta", SNP] %>% unique(), 
                     mode = meqtl.all.mode.df[treatment == "delta", SNP] %>% unique())

mesnps.veh   <- list(prob = meqtl.all.prob.df[treatment == "baseline", SNP] %>% unique(), 
                     mode = meqtl.all.mode.df[treatment == "baseline", SNP] %>% unique())
```

### delta GR-induced meSNPs

```{r}
euler.tbl <- euler(mesnps.delta)

plot(euler.tbl,
     quantities = list(type = c("counts", "percent")),
     fills = c("lightgrey", "brown"),
     labels =  c("probabilities", " most likely genotype"))
```

### baseline meSNPs

```{r}
euler.tbl <- euler(mesnps.veh)

plot(euler.tbl,
     quantities = list(type = c("counts", "percent")),
     fills = c("lightgrey", "brown"),
     labels =  c("probabilities", " most likely genotype"))
```

# Case visualisation of delta GR-induced meQTL results

```{r, include=F, eval=T}
eqtm.in.pre  <- "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/data/integrative/matrixEQTL/"

snp.df.fn <- paste0(eqtm.in.pre, "snp_mtrx_with_na.csv") # "snp_mtrx.csv")
snp.df    <- fread(snp.df.fn)

methyl.beta.df.fn   <- paste0(eqtm.in.pre, "methyl_beta_mtrx_dex.csv")
methyl.beta.dex.df  <- fread(methyl.beta.df.fn) 

methyl.beta.df.fn   <- paste0(eqtm.in.pre, "methyl_beta_mtrx_veh.csv")
methyl.beta.veh.df  <- fread(methyl.beta.df.fn) 

all(colnames(snp.df)[-1] == colnames(methyl.beta.dex.df)[-1])
all(colnames(snp.df)[-1] == colnames(methyl.beta.veh.df)[-1])
```

```{r}
meqtls.delta <- list(prob = meqtl.all.prob.df[treatment == "delta", meQTL_ID], 
                     miss = NULL, # meqtl.all.na.df[treatment == "delta", meQTL_ID], 
                     mode = meqtl.all.mode.df[treatment == "delta", meQTL_ID])
```

## The example of meQTL which is the most significant only for "the most likely genotype" case

```{r}
unique.mode.meqtl    <- setdiff(meqtls.delta$mode, union(meqtls.delta$prob, meqtls.delta$miss))
unique.mode.meqtl.df <- meqtl.all.mode.df[treatment == "delta"][meQTL_ID %in% unique.mode.meqtl]
```

```{r include = T, eval = T}
selected.meqtl <- unique.mode.meqtl.df[1]

kable(selected.meqtl %>% 
        select(SNP, CpG_ID, FC = beta, `t-stat`, `p-value`, FDR = fdr, Treatment = treatment) %>% 
        mutate_if(is.numeric, funs(as.character(signif(., 3)))))
```

```{r out.width = "95%", include = T, eval = T}
ProcessGetBoxPlot(methyl.beta.veh.df, methyl.beta.dex.df, snp.df, selected.meqtl, fdr.thr = 0.05)
```

## The example of meQTL which is the most significant only for the "dosage" case

```{r}
unique.meqtl    <- setdiff(meqtls.delta$prob, union(meqtls.delta$mode, meqtls.delta$miss))
unique.meqtl.df <- meqtl.all.prob.df[treatment == "delta"][meQTL_ID %in% unique.meqtl]
```

```{r include = T, eval = T}
selected.meqtl <- unique.meqtl.df[1]

kable(selected.meqtl %>% 
        select(SNP, CpG_ID, FC = beta, `t-stat`, `p-value`, FDR = fdr, Treatment = treatment) %>% 
        mutate_if(is.numeric, funs(as.character(signif(., 3)))))
```

```{r out.width = "95%", include = T, eval = T}
ProcessGetBoxPlot(methyl.beta.veh.df, methyl.beta.dex.df, snp.df, selected.meqtl, fdr.thr = 0.05)
```

## The example of meQTL which is significant for both cases

```{r}
meqtl.lst <- intersect(meqtls.delta$mode, meqtls.delta$prob)
```

### The most signififcant from the analysis where uncertain genotypes were replaced with the most likely one

```{r include = T, eval = T}
meqtl.df  <- meqtl.all.mode.df[treatment == "delta"][meQTL_ID %in% meqtl.lst]
selected.meqtl <- meqtl.df[1]

kable(selected.meqtl %>% 
        select(SNP, CpG_ID, FC = beta, `t-stat`, `p-value`, FDR = fdr, Treatment = treatment) %>% 
        mutate_if(is.numeric, funs(as.character(signif(., 3)))))
```

```{r out.width = "95%", include = T, eval = T}
selected.meqtl <- meqtl.df[1]

ProcessGetBoxPlot(methyl.beta.veh.df, methyl.beta.dex.df, snp.df, selected.meqtl, fdr.thr = 0.05)
```

### The most signififcant from the analysis where dosages were used

```{r include = T, eval = T}
meqtl.df  <- meqtl.all.prob.df[treatment == "delta"][meQTL_ID %in% meqtl.lst]
selected.meqtl <- meqtl.df[1]

kable(selected.meqtl %>% 
        select(SNP, CpG_ID, FC = beta, `t-stat`, `p-value`, FDR = fdr, Treatment = treatment) %>% 
        mutate_if(is.numeric, funs(as.character(signif(., 3)))))
```

```{r out.width = "95%", include = T, eval = T}
selected.meqtl <- meqtl.df[1]

ProcessGetBoxPlot(methyl.beta.veh.df, methyl.beta.dex.df, snp.df, selected.meqtl, fdr.thr = 0.05)
```