---
title: "DNAm Analysis"
subtitle: "Dexamethasone-Stimulated Human Array Project"
author: 
  - name: Anastasiia Hryhorzhevska
    email: anastasiia_hry@psych.mpg.de
    url: https://github.com/ahryho
    affiliation: Max Planck Institute of Psychiatry
    affiliation_url: https://www.psych.mpg.de/2664393/medizinische-genomforschung
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    css: style.css
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(digits = 2, width = 100, stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE,
                      tidy.opts = list(width.cutoff = 100),
                      tidy = TRUE,
                      fig.pos = "H",
                      dev = "png",
                      dpi = 72,
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE, 
                      cache.lazy = FALSE)

# define DT as the standard printing method for data.frame
library(knitr)
knit_print.data.frame <- function(x, ...) {
  knit_print(DT::datatable(x,
                           filter="top",
                           rownames = FALSE,
                           extensions = "FixedColumns",
                           options = list(
                             scrollX = TRUE,
                             pageLength = 5
                             
                           )), ...)
}

registerS3method("knit_print", "data.frame", knit_print.data.frame)
```
```{r load-libs, include = F, eval = T}
library(data.table)
library(dplyr)

library(ggplot2)
library(RColorBrewer)
library(corrplot)
library(gridExtra)
library(plotly)
library(viridis)
library(ggrepel)

library(GenomicRanges)
library(GenomicFeatures)
library(ChIPseeker)
library(janitor)
```
# Introduction
***

The introduction will come soon.

***
# DMPs
***

## Overview
***

The current analysis was performed for the linear mixed effects model (lmem) with

  + __DNAm_SV__: first three methylation surrogate variables which represents methylation blood cell counts
  + __SNP_PC__: first two genotype principal components
  + __FDR__: False Discovery Rate was set to 0.01 based on the analysis of 450K 
  + __FC__: Fold Change was set to 0.02 based on the analysis of 450K 

## Roadmap
***

The roadmap is being modified based on the consultations with different people :) 

Final version will be placed later.

### The genotype roadmap:

1. Subset data:
    + MAF >= 5%
    + filter only SNPs
    + exclude MHC region
2. LD pruning based on pairwise correlation with the following options:
    + a window of 50 SNPs: to calculate LD between each pair of SNPs
    + remove one of a pair if the LD is greater than 0.2
    + short the window 5 SNPs forward
3. MDS / PCA in PLINK
4. Analysis in R


## Files location
***

The files' locations will be added pnce the analysis is finished. Currently, the source data and calculated data are stored in the project working directory:

  + __/data__: the source data folder
  + __/output/data__: the folder with analysis results
  
```{sh, include = F, eval = F}
scp -C -r -o ProxyCommand="ssh cluser@biocomp.psych.mpg.de nc slurmgate 22" ahryhorzhevska@slurmgate:/binder/mgp/datasets/2020_DexStim_Array_Human/methylation/20_DMA/01_lme_models/dnam_lmem_svs_pcs_rslt.txt output/data/methylation/01_lmem_dnam
```

```{r plot-param, include = F, eval = T}
standard_textsize <- 12
standard_dpi_fig  <- 600
standard_width_mm_single <- 90
standard_width_mm_double <- 180
standard_height_mm <- 100
```

```{r, include = F, eval = T}
source("~/mpip/projects/dex-stim-human-array/code/methylation/02_dma/01_dmp/funcitons.R")
source("~/mpip/projects/dex-stim-human-array/code/integrative/util.R")
```

## Analysis
***

```{r load-lmem-anno, include = F, eval = T}
lmem.rslt.fn       <- '/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/output/data/methylation/01_lmem_dnam/dnam_lmem_svs_pcs_rslt.txt'
anno.fn            <- '/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/output/data/methylation/02_dmp/dex_cpgs_annotated.csv'
anno.chipseeker.fn <- "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/data/annotation/cpg_annotated_withChIPseeker_all.rds"

lmem.rslt          <- fread(lmem.rslt.fn)
anno.chipseeker    <- readRDS(anno.chipseeker.fn)
```


```{r load-pheno, include = F, eval = T}
pheno.full.fn <- paste0("/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/data/pheno/pheno_full_for_kimono.csv") 
pheno.full    <- read.csv2(pheno.full.fn) %>% setDT()

pheno <- pheno.full[Include == 1]

pheno$Status  <- as.factor(pheno$Status)
pheno$Sex     <- as.factor(pheno$Sex)

levels(pheno$Sex) <- c("Male", "Female")
levels(pheno$Status) <- c("Control", "MDD")
```

# Volcano plot

```{r set-up-params, out.width = "85%", fig.cap = "Model with DNAm BCCs and Genotype PCs", include = T, eval = T}
fdr <- 0.01
fc  <- 0.02

cbPalette <- c( "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888")

volcano.plt <- VolcanoPlot(lmem.rslt, fdr, fc, "DNAm SVs and SNP PCs", cbPalette[c(11, 3, 6)])
ggsave("~/mpip/projects/dex-stim-human-array/output/plots/manuscript/m1_f2_Volcano_plot_DMA.png",
       volcano.plt, width = standard_width_mm_single, height = 60, units = "mm", dpi = 600, scale = 3)
volcano.plt
```

# Significant DMPs

```{r load-sign-dmps, include = F, eval = T}
dmps.sign.anno.fn <- "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/output/data/methylation/02_dmp/dmps_fdr01_fc02_anno_full.csv"
dmps.sign.anno.df <- fread(dmps.sign.anno.fn)
```

## DNAm PCA {.tabset .tabset-pills}
***

  + The PCA was performed to investigate the impact of additional surrogate variables capturing variation in DNA methylation on the association statistics for dex-associated DMPs. 
  + The first ten PCs were derived from the DMPs and compared to the available phenotype data and derived covariates to identify potential sources of additional variation between samples.  


```{r load-dnam-pcs, include = F, eval = T}
pca.obj.fn <- "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/output/data/methylation/02_dmp/dex_methyl_pca_sign_dmps.rds"
pca.obj    <- readRDS(pca.obj.fn)

nr.pcs   <- ncol(pca.obj$rotation)

propvar  <- round(summary(pca.obj)$importance["Proportion of Variance", 1:nr.pcs] * 100, 2 )
cummvar  <- round(summary(pca.obj)$importance["Cumulative Proportion", 1:nr.pcs] * 100, 2) 

dnam.pcs <- data.frame(pca.obj$x)
dnam.pcs <- dnam.pcs[match(pheno$DNAm_ID, as.factor(rownames(dnam.pcs))), ]

dnam.pcs    <- dnam.pcs[match(pheno$DNAm_ID, rownames(dnam.pcs)), paste0("PC", 1:10)] 
colnames(dnam.pcs) <- paste0("DNAm_PC", 1:10)
```

```{r combine-all-data, include = F, eval = T}
bcc.gex  <- pheno[, c("Cort_D1", "ACTH_D1", "Leukos_D1", "Gran_D1", "Mono_D1", "Lymph_D1")]
bcc.gex <- bcc.gex %>% 
  mutate_all(~ifelse(is.na(.), median(., na.rm = TRUE), .))
bcc.meth <- pheno[, c("CD8T", "CD4T", "NK", "Bcell", "Mono", "Gran")]
sv.gex   <- pheno[, paste0("V", 1:10)] # c("V1", "V2", "V3", "V4", "V5")]
sv.meth  <- pheno[, paste0("DNAm_SV", 1:10)]# c("DNAm_SV1", "DNAm_SV2", "DNAm_SV3")]
snp.pcs  <- pheno[, paste0("PC", 1:5)]

tmp.df <- data.frame(sv.meth, meth = bcc.meth, gex = bcc.gex, gex = sv.gex, snp.pcs, dnam.pcs,
                     Dex = as.numeric(pheno$Dex), Status = as.numeric(pheno$Status), Sex = as.numeric(pheno$Sex), BMI = pheno$BMI_D1, 
                     Age = pheno$Age, DNAm_Age = pheno$mAge_Hovath, DNAm_Smoke = pheno$DNAm_SmokingScore) %>% 
  na.omit() %>% 
  scale()

cor.mtrx <- cor(tmp.df)
```

### Corplot

Heat-map of correlations between each PC (1-10) and the available phenotype information:

  + __Dex__: a dexamethasone treatment group
  + __Sex__: a sex group
  + __Status__: a MDD status group 
  + __Age__
  + __BMI__ 

and variables derived from the DNA methylation data:

  + __DNAm BCC__: CD8T, CD4T, NK, Bcell, Mono, Gran - cellular proportion estimates from Houseman algorithm
  + __DNAm Age__: an age estimated from DNAm data using Horvath algorithm
  + __Smoking Score__: smoking score estimated from DNAm data using Elliot's CpGs

```{r out.width = "98%", fig.cap = "Relationship of the top 10 PCs derived from DNAm data with with available and derived covariates", include = T, eval = T}
cor.mtrx.plt <- cor.mtrx[paste0("DNAm_PC", 1:10), c(paste0("DNAm_SV", 1:3),
                                                    paste("meth", colnames(bcc.meth), sep = "." ),
                                                    c("Dex", "Status", "Sex", "BMI", "Age", "DNAm_Age", "DNAm_Smoke"))]
par(cex = 1)
corrplot(cor.mtrx.plt, 
         method = "color", cl.pos = "b", cl.cex = 0.6,
         addgrid.col = "gray90",
         addCoef.col = "black", number.cex = 0.6, number.digits = 1, # Add coefficient of correlation
         tl.col = "black",  tl.cex = 0.6, tl.srt = 45, tl.offset = 0.5, tl.pos = "dt", #Text label color and rotation
         sig.level = 0.01) 
```

### Scree plot

```{r out.width = "90%", fig.cap = "Percentage of variance explained by first 10 eigenvalues"}
propvar.df <- data.frame(PC = 1:10, propvar)
ggplot(propvar.df, aes(PC, propvar)) +
  geom_bar(stat = "identity", width = 0.7, fill = "#00688B") +
  geom_line(colour = "red", linetype = "dashed") +
  geom_point(colour = "red") +
  xlab("DNAm PCs") +
  ylab("Percentage variance explained") +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "none")
```

### Dex separation

```{r out.width = "90%", fig.cap = "PCA plot demonstrating the DEX treatment separation based on significant DMPs", include = T, eval = T}
plt.df <- data.frame((dnam.pcs), Group = as.factor(pheno$Dex), Sex = as.factor(pheno$Sex))

ggplot(data = plt.df, aes(DNAm_PC1, DNAm_PC2, col = Group)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("DNAm PC1 (", signif(propvar[1], 2), "%)")) +
  ylab(paste0("DNAm PC2 (", signif(propvar[2], 2), "%)")) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "none")
```

### Sex separation

```{r out.width = "90%", fig.cap = "PCA plot demonstrating the SEX separation based on significant DMPs", include = T, eval = T}
plt.df <- data.frame((dnam.pcs), Group = as.factor(pheno$Dex), Sex = as.factor(pheno$Sex))

ggplot(data = plt.df[plt.df$Group == 1,], aes(DNAm_PC4, DNAm_PC7, col = Sex)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("DNAm PC4 (", signif(propvar[6], 2), "%)")) +
  ylab(paste0("DNAM PC7 (", signif(propvar[7], 2), "%)")) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "none")
```

## Functional Annotation {.tabset .tabset-fade .tabset-pills}
***

```{r include = F}
anno.chipseeker.df <- as.data.frame(as.GRanges(anno.chipseeker))
anno.chipseeker.df <- anno.chipseeker.df[, c("CpG_ID", "seqnames", "Relation_to_Island", "annotation")]
colnames(anno.chipseeker.df) <- c("PROBE_ID", "chr", "Relation_to_Island", "GeneGroup")
```


```{r filter-data, include = F, eval = T}
fdr <- 0.01
fc  <- 0.02

dmps.sign.anno.df [dmps.sign.anno.df$FC < fc, MethylStatus := "hypomethylated DMPs"]
dmps.sign.anno.df [dmps.sign.anno.df$FC > fc, MethylStatus := "hypermethylated DMPs"]
dmps.sign.anno.df <- left_join(dmps.sign.anno.df, anno.chipseeker.df[, c("PROBE_ID", "GeneGroup")])

interest.cols <- c("PROBE_ID", "chr", "Relation_to_Island", "MethylStatus")

dmps.sign.sub.df <- dmps.sign.anno.df %>% dplyr::select(interest.cols) %>% setDT()
dmps.sign.sub.df <- left_join(dmps.sign.sub.df, anno.chipseeker.df[, c("PROBE_ID", "GeneGroup")]) %>% 
  dplyr::select(PROBE_ID, chr, Relation_to_Island, GeneGroup, MethylStatus)

dmps.merged.df                   <- anno.chipseeker.df[!(anno.chipseeker.df$PROBE_ID %in% dmps.sign.sub.df$PROBE_ID), ]
dmps.merged.df[["MethylStatus"]] <- "TotalEPIC"

dmps.merged.df       <- rbind(dmps.merged.df, dmps.sign.sub.df)
```


```{r, eval = F}
#  Save DMPs sign for manuscript

dmps.out.df <- dmps.sign.anno.df %>% dplyr::select(cpg = PROBE_ID,
                                                  chr = chr,
                                                  cpg_pos = pos,
                                                  Relation_to_Island,
                                                  Gene = UCSC_RefGene_Name,
                                                  GeneGroup,
                                                  FC, PVAL, FDR, 
                                                  MethylStatus)

dmps.out.df$MethylStatus <- word(dmps.out.df$MethylStatus, 1)

dmps.out.df$Gene <- apply(dmps.out.df, 1, function(x) strsplit(x[["Gene"]], ";")[[1]][1])
dmps.out.df$Gene <- ifelse(is.na(dmps.out.df$Gene), "", dmps.out.df$Gene)

write.csv2(dmps.out.df, "~/mpip/projects/dex-stim-human-array/output/data/methylation/01_dmps_sign_anno.csv", 
           quote = F, row.names = F)
```


### Manhattan plot

#### DMPs
```{r}
# Load genomic locations
cpg.loc <- fread("/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/data/integrative/matrixEQTL/cpg_locations.csv", select = c("CpG_ID", "chr", "pos"))

dmps.loc.df <- dmps.sign.anno.df[, c("PROBE_ID", "FDR", "UCSC_RefGene_Name")] %>% mutate(fdr = FDR)
dmps.loc.df <- left_join(dmps.loc.df, cpg.loc, by = c("PROBE_ID" = "CpG_ID"))  
```


```{r out.width = "90%", fig.cap = "Manhattan plot for DMPs", include = T, eval = T}
mhn.plt <- GetManhattanPlot(dmps.loc.df, fdr.thr = 0.01)
# ggsave("~/mpip/projects/dex-stim-human-array/output/plots/manuscript/m1_f3_Manhattan plot for DMPs.png",
#        mhn.plt, width = standard_width_mm_single, height = 60, units = "mm", dpi = 600, scale = 4)
mhn.plt
```

#### All CpGs

```{r}
lmem.loc.df <- lmem.rslt %>% full_join(dmps.loc.df[, c("PROBE_ID", "UCSC_RefGene_Name")]) %>%
  left_join(cpg.loc, by = c("PROBE_ID" = "CpG_ID")) %>% 
  mutate(fdr = FDR)

lmem.loc.df$fdr_bon <- p.adjust(lmem.loc.df$PVAL, method = "bonferroni")
```

```{r out.width = "90%", fig.cap = "Manhattan plot for DMPs", include = T, eval = T}
mhn.lmem.plt <- GetManhattanPlot(lmem.loc.df, fdr.thr = 0.01)
ggsave("~/mpip/projects/dex-stim-human-array/output/plots/manuscript/m1_f3_Manhattan_plot_for_DMA.png",
       mhn.lmem.plt, width = standard_width_mm_double / 1.5, height = 60, units = "mm", dpi = 600, scale = 3)
```


### Chromosomes DMPs

```{r include = F, eval = T}
chr.length.tbl <- getChromInfoFromUCSC("hg19")[, c(1, 2)] %>% setDT()
colnames(chr.length.tbl) <- c("chr", "chr_size")
chr.length.tbl$chr <- as.factor(chr.length.tbl$chr)

chr.order             <- paste("chr", 1:22, sep = "")
dmps.sign.sub.df$chr  <- factor(dmps.sign.sub.df$chr,levels = chr.order)
dmps.sign.sub.df      <- left_join(dmps.sign.sub.df, chr.length.tbl)
chr.cpgs.cnt          <- get_relative_freq_pos_across_chrom(dmps.sign.sub.df, chr.length.tbl)

chr.cpgs.cnt[["MethylStatus"]] <- "DMPs"
```

```{r out.width = "90%", fig.cap = "DMPs distribution across the chromosomes wieghted by the chromosomes' length", include = T, eval = T}
cbPalette <- c( "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888")

ggplot(chr.cpgs.cnt, aes(x = chr, y = freq_cpg_chr, fill = chr)) + 
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_text(aes(label = scales::percent(freq_cpg_chr, accuracy = 0.1), y = freq_cpg_chr), 
            stat = "identity", vjust = -.5, size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(title = "",
       x = " ",
       y = "Relative frequency") +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 8),
         axis.title = element_text(size = 8),
         axis.text.x = element_text(angle = 20, hjust = 0.5),
         axis.text.y = element_blank(),
         axis.ticks.y = element_blank(),
         legend.position = "none") +
  scale_fill_manual(values = rep(cbPalette, 2))
  
```
### Chromosomes All

```{r eval = T, include = F}
chr.hyper.dmps.df  <- left_join(dmps.merged.df[dmps.merged.df$MethylStatus == "hypermethylated DMPs", ], chr.length.tbl)
chr.hyper.dmps.df  <- get_relative_freq_pos_across_chrom(chr.hyper.dmps.df, chr.length.tbl) %>% 
  left_join(chr.cpgs.cnt[, c("chr", "n")], by = "chr") %>% 
  mutate(freq_cpg_chr = freq_cpg_chr * n.x / n.y) %>% 
  dplyr::select(chr, freq_cpg_chr) %>%
  mutate(MethylStatus = "hypermethylated DMPs")

chr.hypo.dmps.df  <- left_join(dmps.merged.df[dmps.merged.df$MethylStatus == "hypomethylated DMPs", ], chr.length.tbl)
chr.hypo.dmps.df  <- get_relative_freq_pos_across_chrom(chr.hypo.dmps.df, chr.length.tbl) %>% 
  left_join(chr.cpgs.cnt[, c("chr", "n")], by = "chr") %>% 
  mutate(freq_cpg_chr = freq_cpg_chr * n.x / n.y) %>% 
  dplyr::select(chr, freq_cpg_chr) %>%
  mutate(MethylStatus = "hypomethylated DMPs")

chr.epic.cpgs.df  <- left_join(dmps.merged.df[dmps.merged.df$MethylStatus == "TotalEPIC", ], chr.length.tbl)
chr.epic.cpgs.df  <- get_relative_freq_pos_across_chrom(chr.epic.cpgs.df, chr.length.tbl) %>% 
  dplyr::select(chr, freq_cpg_chr) %>% 
  mutate(MethylStatus = "TotalEPIC")

chr.cpgs.cnt      <- rbind(chr.hyper.dmps.df, chr.hypo.dmps.df, chr.epic.cpgs.df)
# chr.cpgs.cnt$chr  <- factor(chr.cpgs.cnt$chr, levels = chr.order)
chr.cpgs.cnt$chr  <- factor(substr(chr.cpgs.cnt$chr, 4, length(chr.cpgs.cnt$chr)), level = as.character(1:22))
```

```{r out.width = "90%", fig.cap = "Distribution of the CpGs across the chromosomes wieghted by the chromosomes' length", include = T, eval = T}
# 
cbPalette <- c( "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888")

plt.chr <- ggplot(chr.cpgs.cnt, aes(x = chr, y = freq_cpg_chr, fill = MethylStatus, color = MethylStatus)) + 
  geom_bar(stat = "identity", position = "stack") +
  geom_text(aes(label = scales::percent(freq_cpg_chr, accuracy = 0.1), y = freq_cpg_chr), 
            stat = "identity", position = position_stack(vjust = 0.5), size = standard_textsize / 3,
            color = "black") +
  scale_y_continuous(labels = scales::percent) +
  labs(title = "",
       x = "Chromosomes",
       y = "Relative frequency") +
  theme(legend.title = element_blank(), 
        legend.position = "none",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        axis.title = element_text(size = standard_textsize),
        axis.ticks = element_blank(),
        axis.text.x = element_text(angle = 0, hjust = 0.5, size = standard_textsize),
        axis.text.y = element_blank()) +
  scale_color_manual(values = alpha(cbPalette[c(3, 6, 11)], 0.75)) +
  scale_fill_manual(values = alpha(cbPalette[c(3, 6, 11)], 0.75))

ggsave("~/mpip/projects/dex-stim-human-array/output/plots/manuscript/m1_f4_Distribution of the CpGs across the chromosomes wieghted by the chromosomes length.pdf",
       plt.chr, width = standard_width_mm_double , height = 90, units = "mm", dpi = 600, scale = 2)

plt.chr
```

### CpG Islands

```{r include = F, eval = T}
dmps.cnt.df <- dmps.merged.df %>% 
  group_by(Relation_to_Island, MethylStatus) %>%
  tally() %>% setDT()

dmps.cnt.df[MethylStatus %in% c("TotalEPIC"), freq := n / nrow(lmem.rslt)]
dmps.cnt.df[!(MethylStatus %in% c("TotalEPIC")), freq := n / nrow(dmps.sign.sub.df)]
dmps.island.cnt.df <- dmps.cnt.df %>% mutate(Group = Relation_to_Island) %>% dplyr::select(-Relation_to_Island)
```

```{r out.width = "100%", fig.cap = "The ratio of all assessed EPIC CpGs versus dDMPs in different genomic regions in relation to CpG Island", include = T, eval = T}
plt.cgisland <- ggplot(dmps.cnt.df, aes(x = Relation_to_Island, y = freq, fill = MethylStatus, color = MethylStatus)) + 
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_text(aes(label = scales::percent(freq, accuracy = 0.1)),
            vjust = -.5,  position = position_dodge(1), size = standard_textsize / 3, color = "black") +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "", 
       title = "") +
  theme(legend.title = element_blank(), 
        legend.position = c(0.1, 0.9),
        legend.text = element_text(size = standard_textsize),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        axis.title = element_text(size = standard_textsize),
        axis.ticks = element_blank(),
        axis.text.x = element_text(angle = 0, hjust = 0.5, size = standard_textsize),
        axis.text.y = element_blank()) +
  scale_color_manual(values = alpha(cbPalette[c(3, 6, 11)], 0.75)) +
  scale_fill_manual(values = alpha(cbPalette[c(3, 6, 11)], 0.75))

ggsave("~/mpip/projects/dex-stim-human-array/output/plots/manuscript/m1_f5_dDMPs in different genomic regions in relation to CpG Island.pdf",
       plt.cgisland, width = standard_width_mm_double, height = 90, units = "mm", dpi = 600, scale = 1.5)

plt.cgisland
```

### Gene regions

```{r include = F, eval = T}
dmps.cnt.df <- dmps.merged.df %>% 
  group_by(GeneGroup, MethylStatus) %>%
  tally() %>% setDT()

dmps.cnt.df[MethylStatus %in% c("TotalEPIC"), freq := n / nrow(lmem.rslt)]
dmps.cnt.df[!(MethylStatus %in% c("TotalEPIC")), freq := n / nrow(dmps.sign.sub.df)]
dmps.cnt.df$GeneGroup <- as.factor(dmps.cnt.df$GeneGroup)
levels(dmps.cnt.df$GeneGroup) <- c("1st Exon", "1st Intron", "3' UTR", "5' UTR", "Distal\nIntergenic", "Downstream\n(<=300)", "Other\nExon", "Other\nIntron", "Promoter\n(<=1kb)", "Promoter\n(1-2kb)", "Promoter\n(2-3kb)")
dmps.region.cnt.df <- dmps.cnt.df %>% mutate(Group = GeneGroup) %>% dplyr::select(-GeneGroup)
```

```{r out.width = "100%", fig.cap = "The ratio of all assessed EPIC CpGs versus dDMPs in different gene regions", include = T, eval = T}
plt.generegion <- ggplot(dmps.cnt.df, aes(x = GeneGroup, y = freq, fill = MethylStatus, color = MethylStatus)) + 
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_text(data = dmps.cnt.df[dmps.cnt.df$freq >= 0.004, ],
            aes(label = scales::percent(freq, accuracy = 0.1)),
            vjust = -.5,  position = position_dodge(1), size = standard_textsize / 3, 
            color = "black") +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "", 
       title = "") +
  theme(legend.title = element_blank(), 
        legend.position = "none",
        legend.text = element_text(size = standard_textsize),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        axis.title = element_text(size = standard_textsize),
        axis.ticks = element_blank(),
        axis.text.x = element_text(angle = 0, size = standard_textsize),
        axis.text.y = element_blank()) +
  scale_color_manual(values = alpha(cbPalette[c(3, 6, 11)], 0.75)) +
  scale_fill_manual(values = alpha(cbPalette[c(3, 6, 11)], 0.75))

ggsave("~/mpip/projects/dex-stim-human-array/output/plots/manuscript/m1_f6_dDMPs in different genomic regions in relation to nearby genes.pdf",
       plt.generegion, width = standard_width_mm_double, height = 90, units = "mm", dpi = 600, scale = 1.5)

plt.generegion
```

```{r include = F}
nested <- ((volcano.plt|(mhn.plt/plt.chr))/(plt.cgisland|plt.generegion)) +
  patchwork::plot_annotation(tag_levels = 'A') +
  plot_layout(height = unit(c(standard_height_mm * 2, standard_height_mm), 
                            c('mm', 'mm')),
              widths = unit(c(standard_width_mm_double * 3, standard_width_mm_double * 3),#, standard_width_mm_double + 20), 
                            c('mm', 'mm')))

ggsave("~/mpip/projects/dex-stim-human-array/output/plots/manuscript/m1_f1_dma_panel.png",
       nested, width = standard_width_mm_double * 1.7, height = standard_height_mm * 2, units = "mm", dpi = 600, scale = 2)
```

```{r eval = F, include = F}
dmps.cnt.df <- rbind(dmps.island.cnt.df ,
                     dmps.region.cnt.df, 
                     use.names = F)

plt <- ggplot(dmps.cnt.df, aes(x = Group, y = freq, fill = MethylStatus, color = MethylStatus)) + 
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_text(
            aes(label = scales::percent(freq, accuracy = 0.1)),
            vjust = -.5,  position = position_dodge(1), size = 3, 
            color = "black") +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "", 
       title = "") +
  theme(legend.title = element_blank(), 
        legend.position = "bottom",
        legend.text = element_text(size = standard_textsize),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        axis.title = element_text(size = standard_textsize),
        axis.text.x = element_text(angle = 20, hjust = 0.75, size = standard_textsize),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  scale_color_manual(values = alpha(cbPalette[c(3, 6, 11)], 0.75)) +
  scale_fill_manual(values = alpha(cbPalette[c(3, 6, 11)], 0.75))

ggsave("~/mpip/projects/dex-stim-human-array/output/plots/manuscript/m1_f7_dDMPs_functional_anno.pdf",
       plt, width = standard_width_mm_double, height = 90, units = "mm", dpi = 600, scale = 3)

plt
```

