---
title: "cis-eQTM Analysis"
subtitle: "Dexamethasone-Stimulated Human Array Project"
author: 
  - name: Anastasiia Hryhorzhevska
    email: anastasiia_hry@psych.mpg.de
    url: https://github.com/ahryho
    affiliation: Max Planck Institute of Psychiatry
    affiliation_url: https://www.psych.mpg.de/2664393/medizinische-genomforschung
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    css: style.css
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
options(digits = 4, width = 100, stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE,
                      tidy.opts = list(width.cutoff = 100),
                      tidy=TRUE,
                      fig.pos = "H",
                      dpi = 200,
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE, 
                      cache.lazy = FALSE)

# define DT as the standard printing method for data.frame
library(knitr)
library(DT)
knit_print.data.frame <- function(x, ...) {
  knit_print(DT::datatable(x,
                           filter="top",
                           rownames = FALSE,
                           extensions = "FixedColumns",
                           options = list(
                             scrollX = TRUE,
                             pageLength = 5
                             
                           )), ...)
}

registerS3method("knit_print", "data.frame", knit_print.data.frame)

library(data.table)
library(dplyr)

library(ggplot2)
library(ggrepel)
library(scales)
library(GGally)
library(factoextra)
library(viridis)
```

***
# __MatrixEQTL Primary Results__
***

```{r load-func, include = F, eval = T}
source("~/bio/code/mpip/dex-stim-human-array/code/integrative/util.R")
cbPalette <- c( "#0072B2", "#009E73", "#E69F00", "#F0E442", "#D55E00", "#CC79A7", "#56B4E9", "#999999")
```

```{r, include = F, eval = T}
src.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/data/integrative/matrixEQTL/"
out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/eqtms/"

cpg.loc.fn       <- paste0(src.dir.pre, "cpg_locations.csv")
cpg.loc.colnames <- c("CpG_ID", "chr", "pos_cpg", "pos_end")
cpg.loc          <- fread(cpg.loc.fn, col.names = cpg.loc.colnames)

ensg.loc.fn       <- paste0(src.dir.pre, "ensg_locations.csv")
ensg.loc.colnames <- c("ENSG_ID", "chr", "pos_start_ensg", "pos_end_ensg")
ensg.loc          <- fread(ensg.loc.fn, col.names = ensg.loc.colnames)

meqtl.veh.fn   <- paste0(out.dir.pre, "eqtm_cis_result_veh_beta.csv") 
meqtl.dex.fn   <- paste0(out.dir.pre, "eqtm_cis_result_dex_beta.csv") 
meqtl.delta.fn <- paste0(out.dir.pre, "eqtm_cis_result_delta_no_residuals_or_cov_beta.csv") 
meqtl.delta.res.fn <- paste0(out.dir.pre, "eqtm_cis_result_delta_beta.csv") 

col.names <- c("CpG_ID", "ENSG_ID", "beta", "t-stat", "p-value", "FDR")

dex.df   <- fread(meqtl.dex.fn, col.names = col.names)
veh.df   <- fread(meqtl.veh.fn, col.names = col.names)
delta.df <- fread(meqtl.delta.fn, col.names = col.names)
delta.res.df <- fread(meqtl.delta.res.fn, col.names = col.names)
```

## __Analysis of nominal significant eQTMs, FDR <= 1, p <= 0.05__

```{r, include = F, eval = T}
# head(meqtl.df)
fdr.thr  <- 1

veh.full.df   <- GetFulleQTMdf(veh.df, fdr.thr = fdr.thr, treatment = "baseline")
dex.full.df   <- GetFulleQTMdf(dex.df, fdr.thr = fdr.thr, treatment = "dex")
delta.full.df <- GetFulleQTMdf(delta.df, fdr.thr = fdr.thr, treatment = "delta")
delta.res.full.df <- GetFulleQTMdf(delta.res.df, fdr.thr = fdr.thr, treatment = "delta-res")

all.full.df   <- rbind(dex.full.df, veh.full.df, delta.full.df, delta.res.full.df)
```

### Hits of all eQTMs
<!-- Check the number of found eQTLs, unique CpGs and SNPs in all datasets. -->

```{r, include = F, eval = T}
sign.hits.df <- data.table(c(delta.full.df$fdr  %>% length(), 
                             delta.full.df$CpG_ID %>% unique() %>% length(),
                             delta.full.df$ENSG_ID %>% unique() %>% length()),
                           c(delta.res.full.df$fdr  %>% length(), 
                             delta.res.full.df$CpG_ID %>% unique() %>% length(),
                             delta.res.full.df$ENSG_ID %>% unique() %>% length()),
                           c(dex.full.df$fdr  %>% length(), 
                             dex.full.df$CpG_ID %>% unique() %>% length(),
                             dex.full.df$ENSG_ID %>% unique() %>% length()),
                           c(veh.full.df$fdr  %>% length(), 
                             veh.full.df$CpG_ID %>% unique() %>% length(),
                             veh.full.df$ENSG_ID %>% unique() %>% length()))

sign.hits.df <- as.data.frame(sign.hits.df)
rownames(sign.hits.df) <- c("cis-eQTM", "CpG_ID", "ENSG_ID")
colnames(sign.hits.df) <- c("delta", "delta-res", "dex", "baseline")
```

```{r out.width = "95%", include = T, eval = T}
cbPalette <- c("lightskyblue", "cornflowerblue", "#0072B2", "grey")

sign.hits.df %>% as.matrix() %>% reshape2::melt(var.names = c("ID", "Treatment"), value.name = "cnt") %>%
  ggplot(aes(y = cnt, x = Var2, fill = as.factor(as.numeric(Var2)))) +
    geom_col() +
    geom_text(aes(label = comma(cnt, accuracy = 1L), y = cnt), 
                  stat = "identity", 
                  vjust = -0.2, size = 2.5) + 
    scale_y_continuous(labels = scientific) +
    # scale_fill_viridis(discrete = TRUE, alpha = 0.9, option = "C") +
    labs(title = " ", y = "Count", x = " ") + 
    facet_wrap(~ Var1, ncol = 3) +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank()) +
  scale_fill_manual(values = cbPalette)
```

## Scatter plot

```{r out.width = "95%", include = T, eval = T}
plot.title <- paste0("Independent and significant at FDR <= ", fdr.thr, " eQTMs.",
                     "\nResult of MatrixEQTL")
GetScatterPlot2(all.full.df, plot.title = plot.title, fdr.thr = fdr.thr)
```

## Effect size

```{r out.width = "95%", include = T, eval = T}
ggplot(all.full.df, aes(y = beta, x = treatment, color = treatment)) +
  geom_boxplot() +
  scale_y_continuous(breaks = seq(signif(min(all.full.df$beta), 0) * 1.2, max(all.full.df$beta) * 1.2, by = 10)) +
  theme(legend.position = "none", 
          legend.title = element_blank(),
          # panel.grid.major = element_blank(),
          # panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank()) +
  scale_color_manual(values = cbPalette)
```

## Upset plots{.tabset .tabset-fade .tabset-pills}

### eQTM level
  
```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

eqtms <- list(delta = all.full.df[treatment == "delta", eQTM_ID], 
              delta_res = all.full.df[treatment == "delta_res", eQTM_ID], 
              dex = all.full.df[treatment == "dex", eQTM_ID], 
              veh = all.full.df[treatment == "veh", eQTM_ID])

upset(fromList(eqtms), 
      sets = c("delta", "delta_res", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, eQTM", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c( "#0072B2", "#009E73", "#E69F00", "#F0E442"), 
      order.by = "freq")
```

### Gene level
  
```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

eqtms.ensg <- list(delta = delta.full.df$ENSG_ID, delta_res = delta.res.full.df$ENSG_ID , dex = dex.full.df$ENSG_ID, veh = veh.full.df$ENSG_ID)

upset(fromList(eqtms.ensg), 
      sets = c("delta", "delta_res", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, eGene (ENSG)", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c("#0072B2", "#009E73", "#E69F00", "#F0E442"), 
      order.by = "freq")
```

### CpG level

```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

eqtms.cpg <- list(delta = delta.full.df$CpG_ID, delta_res = delta.res.full.df$CpG_ID, dex = dex.full.df$CpG_ID, veh = veh.full.df$CpG_ID)

upset(fromList(eqtms.cpg), 
      sets = c("delta", "delta_res", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, eCpG", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c("#0072B2", "#009E73", "#E69F00", "#F0E442"), 
      order.by = "freq")
```

## __Analysis of significant eQTMs, FDR <= 0.05__

```{r, include = F, eval = T}
# head(meqtl.df)
fdr.thr  <- 0.05

veh.full.df   <- GetFulleQTMdf(veh.df, fdr.thr = fdr.thr, treatment = "baseline")
dex.full.df   <- GetFulleQTMdf(dex.df, fdr.thr = fdr.thr, treatment = "dex")
delta.full.df <- GetFulleQTMdf(delta.df, fdr.thr = fdr.thr, treatment = "delta")
delta.res.full.df <- GetFulleQTMdf(delta.res.df, fdr.thr = fdr.thr, treatment = "delta_res")

all.full.df   <- rbind(dex.full.df, veh.full.df, delta.full.df, delta.res.full.df)

# fwrite(all.full.df,
#        paste0(out.dir.pre, "eqtm_cis_all_fdr_005.csv"),
#        quote = F, row.names = F, sep   = "\t")
```

### Significant hits of eQTMs with FDR <= 0.05

<!-- Check the number of found eQTLs, unique CpGs and SNPs in all datasets. -->

```{r, include = F, eval = T}
sign.hits.df <- data.table(c(delta.full.df$fdr  %>% length(), 
                             delta.full.df$CpG_ID %>% unique() %>% length(),
                             delta.full.df$ENSG_ID %>% unique() %>% length()),
                           c(delta.res.full.df$fdr  %>% length(), 
                             delta.res.full.df$CpG_ID %>% unique() %>% length(),
                             delta.res.full.df$ENSG_ID %>% unique() %>% length()),
                           c(dex.full.df$fdr  %>% length(), 
                             dex.full.df$CpG_ID %>% unique() %>% length(),
                             dex.full.df$ENSG_ID %>% unique() %>% length()),
                           c(veh.full.df$fdr  %>% length(), 
                             veh.full.df$CpG_ID %>% unique() %>% length(),
                             veh.full.df$ENSG_ID %>% unique() %>% length()))

sign.hits.df <- as.data.frame(sign.hits.df)
rownames(sign.hits.df) <- c("cis-eQTL", "CpG_ID", "ENSG_ID")
colnames(sign.hits.df) <- c("delta", "delta_res", "dex", "baseline")
```

```{r out.width = "95%", include = T, eval = T}
cbPalette <- c("lightskyblue", "cornflowerblue", "#0072B2", "grey")

sign.hits.df %>% as.matrix() %>% reshape2::melt(var.names = c("ID", "Treatment"), value.name = "cnt") %>%
  ggplot(aes(y = cnt, x = Var2, fill = as.factor(as.numeric(Var2)))) +
    geom_col() +
    geom_text(aes(label = comma(cnt, accuracy = 1L), y = cnt), 
                  stat = "identity", 
                  vjust = -0.2, size = 2.5) + 
    scale_y_continuous(labels = scientific) +
    # scale_fill_viridis(discrete = TRUE, alpha = 0.9, option = "C") +
    labs(title = " ", y = "Count", x = " ") + 
    facet_wrap(~ Var1, ncol = 3) +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank()) +
  scale_fill_manual(values = cbPalette)
```

## Scatter plot

```{r out.width = "95%", include = T, eval = T}
plot.title <- paste0("Independent and significant at FDR <= ", fdr.thr, " eQTMs.",
                     "\nResult of MatrixEQTL")
GetScatterPlot2(all.full.df, plot.title = plot.title, fdr.thr = fdr.thr)
```

## Effect size

```{r out.width = "95%", include = T, eval = T}
ggplot(all.full.df, aes(y = beta, x = treatment, color = treatment)) +
  geom_boxplot() +
  scale_y_continuous(breaks = seq(signif(min(all.full.df$beta), 0) * 1.2, max(all.full.df$beta) * 1.2, by = 10)) +
  theme(legend.position = "none", 
          legend.title = element_blank(),
          # panel.grid.major = element_blank(),
          # panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank()) +
  scale_color_manual(values = cbPalette)
```

## Upset plots{.tabset .tabset-fade .tabset-pills}

### eQTM level
  
```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

eqtms <- list(delta = all.full.df[treatment == "delta", eQTM_ID], 
              delta_res = all.full.df[treatment == "delta_res", eQTM_ID], 
              dex = all.full.df[treatment == "dex", eQTM_ID], 
              veh = all.full.df[treatment == "veh", eQTM_ID])

upset(fromList(eqtms), 
      sets = c("delta", "delta_res", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, eQTM", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c( "#0072B2", "#009E73", "#E69F00", "#F0E442"), 
      order.by = "freq")
```

### Gene level
  
```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

eqtms.ensg <- list(delta = delta.full.df$ENSG_ID, delta_res = delta.res.full.df$ENSG_ID , dex = dex.full.df$ENSG_ID, veh = veh.full.df$ENSG_ID)

upset(fromList(eqtms.ensg), 
      sets = c("delta", "delta_res", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, eGene (ENSG)", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c("#0072B2", "#009E73", "#E69F00", "#F0E442"), 
      order.by = "freq")
```

### CpG level

```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

eqtms.cpg <- list(delta = delta.full.df$CpG_ID, delta_res = delta.res.full.df$CpG_ID, dex = dex.full.df$CpG_ID, veh = veh.full.df$CpG_ID)

upset(fromList(eqtms.cpg), 
      sets = c("delta", "delta_res", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, eCpG", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c("#0072B2", "#009E73", "#E69F00", "#F0E442"), 
      order.by = "freq")
```

<!-- ## Distances -->

<!-- ```{r out.width = "95%", include = T, eval = T} -->
<!-- ggplot(meqtl.all.full.df, aes(x = dist, fill = treatment)) +  -->
<!--   geom_density(alpha = 0.2, aes(color = treatment)) + -->
<!--   theme(legend.position = c(.9,.9),  -->
<!--         legend.title = element_blank(), -->
<!--         panel.grid.major = element_blank(), -->
<!--         panel.background = element_blank(), -->
<!--         plot.title = element_text(size = 10), -->
<!--         axis.title = element_text(size = 10)) + -->
<!--   scale_x_continuous(breaks = c(0, 1e2, 1e3, 1e4, 1e5, 1e6), labels = scientific) + -->
<!--   scale_y_continuous(labels = scientific) + -->
<!--   labs(title = paste0("Distribution of cis-meQTLs, FDR = ", fdr.thr),  -->
<!--        x = "Distance of CpG meQTL to SNP",  -->
<!--        y = "Density") + -->
<!--   scale_fill_manual(values = cbPalette) -->
<!-- ``` -->

<!-- ```{r out.width = "100%", include = T, eval = T} -->
<!-- library(mltools) -->
<!-- bins.df <- data.frame(DistanceInterval = bin_data(data.table::data.table(meqtl.all.full.df), binCol = "dist", bins = c(0, 1, 500, 1e3+1, 5e3+1, -->
<!--                                                                                                                        1e4+1, 5e4+1, 1e5+1, 5e5+1, -->
<!--                                                                                                                        1e6+1), returnDT = F)) -->
<!-- levels(bins.df$DistanceInterval) <- c("0", "(0; 500]", "(500; 1e3]", "(1e3; 5e3]", "(5e3; 1e4]", "(1e4; 5e4]", "(5e4; 1e5]", "[1e5; 5e5)", "[5e5; 1e6]") -->
<!-- bins.df <- cbind(bins.df, treatment = meqtl.all.full.df$treatment) -->

<!-- ggplot(bins.df, aes(DistanceInterval, fill = treatment)) +  -->
<!--   geom_bar(position = position_dodge()) + -->
<!--   stat_count(geom = "text",  -->
<!--              aes(label = comma(..count.., accuracy = 1L)), -->
<!--              position = position_dodge(1),  vjust = -1, colour = "black", cex = 2) + -->
<!--   theme(legend.position = c(.1,.9),  -->
<!--         legend.title = element_blank(), -->
<!--         panel.grid.major = element_blank(), -->
<!--         panel.background = element_blank(), -->
<!--         axis.text.y = element_blank(), -->
<!--         plot.title = element_text(size = 10), -->
<!--         axis.title = element_text(size = 8)) + -->
<!--   labs(title = paste0("Distribution of cis-meQTLs, FDR = ", fdr.thr),  -->
<!--        x = "Distance Interval, bp",  -->
<!--        y = "No. cis-meQTLs") +  -->
<!--   scale_fill_manual(values = cbPalette) -->
<!-- ``` -->
<!-- *** -->

<!-- ```{r, eval = F, include = F, warning = F} -->

<!-- ## Groups annotation -->

<!-- ### Read meQTLs with opposite effects -->

<!-- out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/" -->

<!-- meqtls.opposite.fc.grp.with.delta <-  -->
<!--   fread(paste0(out.dir.pre, "dex_veh_meqtl_opposite_beta_group_with_delta.csv")) -->

<!-- meqtls.opposite.fc.grp.no.delta <-  -->
<!--   fread(paste0(out.dir.pre, "dex_veh_meqtl_opposite_beta_9K.csv")) -->

<!-- meqtl.opposite.fc <- c(meqtls.opposite.fc.grp.with.delta$meQTL_ID, meqtls.opposite.fc.grp.no.delta$meQTL_ID) -->
<!-- ``` -->

<!-- ```{r include = F, eval = F} -->
<!-- meqtl.grp.overlaps.all       <- intersect(intersect(meqtls.meqtls$veh, meqtls.meqtls$delta), meqtls.meqtls$dex) -->
<!-- meqtl.grp.overlaps.delta.veh <- setdiff(intersect(meqtls.meqtls$veh, meqtls.meqtls$delta), meqtls.meqtls$dex) -->
<!-- meqtl.grp.overlaps.delta.dex <- setdiff(intersect(meqtls.meqtls$dex, meqtls.meqtls$delta), meqtls.meqtls$veh) -->
<!-- meqtl.grp.only.delta         <- setdiff(setdiff(meqtls.meqtls$delta, meqtls.meqtls$dex), meqtls.meqtls$veh) -->

<!-- meqtl.grp.overlaps.dex.veh <- setdiff(intersect(meqtls.meqtls$veh, meqtls.meqtls$dex), meqtls.meqtls$delta) -->
<!-- meqtl.grp.only.dex         <- setdiff(setdiff(meqtls.meqtls$dex, meqtls.meqtls$veh), meqtls.meqtls$delta) -->
<!-- meqtl.grp.only.veh         <- setdiff(setdiff(meqtls.meqtls$veh, meqtls.meqtls$dex), meqtls.meqtls$delta) -->

<!-- meqtls.opposite.fc.grp.veh.dex <- setdiff(c(meqtl.grp.overlaps.dex.veh, meqtl.grp.only.dex, meqtl.grp.only.veh), meqtls.opposite.fc.grp.no.delta$meQTL_ID) -->
<!-- meqtls.opposite.fc.grp.delta   <- setdiff(c(meqtl.grp.overlaps.all, meqtl.grp.overlaps.delta.veh, meqtl.grp.overlaps.delta.dex,  meqtl.grp.only.delta), meqtls.opposite.fc.grp.with.delta$meQTL_ID) -->
<!-- ``` -->

<!-- <!-- Annotate data --> -->

<!-- ```{r include = F, eval = F} -->
<!-- library(TxDb.Hsapiens.UCSC.hg19.knownGene) -->
<!-- library(ChIPseeker) -->
<!-- library(org.Hs.eg.db) -->

<!-- out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/primary_meqtl_results/" -->

<!-- # anno.snps.all.gr <- readRDS("~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/snp_annotated_withChIPseeker_all.rds") -->

<!-- meqtl.overlap.snp.all.df <- meqtl.all.full.df[meQTL_ID %in% meqtl.grp.overlaps.all] -->
<!-- meqtl.overlap.snp.all.gr <- annotate_with_chipseeker(meqtl.overlap.snp.all.df, "chr", "pos_snp", "pos_snp", "SNP") -->
<!-- saveRDS(meqtl.overlap.snp.all.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_delta_dex_veh_gr.rds")) -->

<!-- meqtl.overlap.snp.delta.veh.df <- meqtl.all.full.df[meQTL_ID %in% meqtl.grp.overlaps.delta.veh] -->
<!-- meqtl.overlap.snp.delta.veh.gr <- annotate_with_chipseeker(meqtl.overlap.snp.delta.veh.df, "chr", "pos_snp", "pos_snp", "SNP") -->
<!-- saveRDS(meqtl.overlap.snp.delta.veh.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_delta_veh_gr.rds")) -->

<!-- meqtl.overlap.snp.delta.dex.df <- meqtl.all.full.df[meQTL_ID %in% meqtl.grp.overlaps.delta.dex] -->
<!-- meqtl.overlap.snp.delta.dex.gr <- annotate_with_chipseeker(meqtl.overlap.snp.delta.dex.df, "chr", "pos_snp", "pos_snp", "SNP") -->
<!-- saveRDS(meqtl.overlap.snp.delta.dex.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_delta_dex_gr.rds")) -->

<!-- meqtl.only.snp.delta.df <- meqtl.all.full.df[meQTL_ID %in% meqtl.grp.only.delta] -->
<!-- meqtl.only.snp.delta.gr <- annotate_with_chipseeker(meqtl.only.snp.delta.df, "chr", "pos_snp", "pos_snp", "SNP") -->
<!-- saveRDS(meqtl.only.snp.delta.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_only_delta_gr.rds")) -->

<!-- meqtl.overlap.snp.dex.veh.df <- meqtl.all.full.df[meQTL_ID %in% meqtl.grp.overlaps.dex.veh] -->
<!-- meqtl.overlap.snp.dex.veh.gr <- annotate_with_chipseeker(meqtl.overlap.snp.dex.veh.df, "chr", "pos_snp", "pos_snp", "SNP") -->
<!-- saveRDS(meqtl.overlap.snp.dex.veh.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_dex_veh_gr.rds")) -->

<!-- meqtl.only.snp.dex.df <- meqtl.all.full.df[meQTL_ID %in% meqtl.grp.only.dex] -->
<!-- meqtl.only.snp.dex.gr <- annotate_with_chipseeker(meqtl.only.snp.dex.df, "chr", "pos_snp", "pos_snp", "SNP") -->
<!-- saveRDS(meqtl.only.snp.dex.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_only_dex_gr.rds")) -->

<!-- meqtl.only.snp.veh.df <- meqtl.all.full.df[meQTL_ID %in% meqtl.grp.only.veh] -->
<!-- meqtl.only.snp.veh.gr <- annotate_with_chipseeker(meqtl.only.snp.veh.df, "chr", "pos_snp", "pos_snp", "SNP") -->
<!-- saveRDS(meqtl.only.snp.veh.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_only_veh_gr.rds")) -->

<!-- meqtl.opposite.fc.grp.with.delta.df <- meqtl.all.full.df[meQTL_ID %in% meqtls.opposite.fc.grp.with.delta$meQTL_ID] -->
<!-- meqtl.opposite.fc.grp.with.delta.gr <- annotate_with_chipseeker(meqtl.opposite.fc.grp.with.delta.df, "chr", "pos_snp", "pos_snp", "SNP") -->
<!-- saveRDS(meqtl.opposite.fc.grp.with.delta.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_opposite_fc_grp_with_delta_gr.rds")) -->

<!-- meqtl.opposite.fc.grp.no.delta.df <- meqtl.all.full.df[meQTL_ID %in% meqtls.opposite.fc.grp.no.delta$meQTL_ID] -->
<!-- meqtl.opposite.fc.grp.no.delta.gr <- annotate_with_chipseeker(meqtl.opposite.fc.grp.no.delta.df, "chr", "pos_snp", "pos_snp", "SNP") -->
<!-- saveRDS(meqtl.opposite.fc.grp.no.delta.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_opposite_fc_grp_no_delta_gr.rds")) -->

<!-- meqtl.opposite.fc.grp.veh.dex.df  <- rbind(meqtl.overlap.snp.dex.veh.df,  -->
<!--                                            meqtl.only.snp.dex.df,  -->
<!--                                            meqtl.only.snp.veh.df)[!(meQTL_ID %in% meqtls.opposite.fc.grp.no.delta$meQTL_ID)][, .(chr, pos_snp, SNP)] %>% unique() -->
<!-- meqtl.opposite.fc.grp.veh.dex.gr <- annotate_with_chipseeker(meqtl.opposite.fc.grp.veh.dex.df, "chr", "pos_snp", "pos_snp", "SNP") -->
<!-- saveRDS(meqtl.opposite.fc.grp.veh.dex.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_opposite_fc_grp_veh_dex_gr.rds")) -->

<!-- meqtl.opposite.fc.grp.delta.df <- rbind(meqtl.overlap.snp.all.df,  -->
<!--                                          meqtl.overlap.snp.delta.veh.df,  -->
<!--                                          meqtl.overlap.snp.delta.dex.df,  -->
<!--                                          meqtl.only.snp.delta.df)[!(meQTL_ID %in% meqtls.opposite.fc.grp.with.delta$meQTL_ID)][, .(chr, pos_snp, SNP)] %>% unique() -->
<!-- meqtl.opposite.fc.grp.delta.gr <- annotate_with_chipseeker(meqtl.opposite.fc.grp.delta.df, "chr", "pos_snp", "pos_snp", "SNP") -->
<!-- saveRDS(meqtl.opposite.fc.grp.delta.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_opposite_fc_grp_delta_gr.rds")) -->

<!-- meqtl.opposite.fc.df <- rbind(meqtl.opposite.fc.grp.with.delta.df, meqtl.opposite.fc.grp.no.delta.df)[, .(chr, pos_snp, SNP)] %>% unique() -->
<!-- meqtl.opposite.fc.gr <- annotate_with_chipseeker(meqtl.opposite.fc.df, "chr", "pos_snp", "pos_snp", "SNP") -->
<!-- saveRDS(meqtl.opposite.fc.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_opposite_fc_grp_gr.rds")) -->

<!-- ``` -->

<!-- # Functional annotation -->

<!-- ## meQTL level -->

<!-- <!-- ### SNP level --> -->

<!-- <!-- ### Manhattan plots{.tabset .tabset-fade .tabset-pills} --> -->

<!-- #### Veh + Dex + Delta -->

<!-- ```{r} -->
<!-- overlaps.all       <- intersect(intersect(meqtls.meqtls$veh, meqtls.meqtls$delta), meqtls.meqtls$dex) -->
<!-- df <- meqtl.all.full.df[meQTL_ID %in% overlaps.all][treatment == "delta"] -->
<!-- df <- df[, .(SNP, chr, pos_snp, fdr)] -->
<!-- colnames(df) <- c("SNP", "CHR", "BP", "P") -->

<!-- GetManhattanPlot(df, ylims = c(-log10(fdr.thr), -log10(min(df$P)) + 1), plot.title = "Intersect(delta, dex, veh), FDR <= 0.05") -->
<!-- ``` -->

<!-- #### Dex + Delta -->

<!-- ```{r} -->
<!-- overlaps.delta.dex <- setdiff(intersect(meqtls.meqtls$dex, meqtls.meqtls$delta), meqtls.meqtls$veh) -->
<!-- df <- meqtl.all.full.df[meQTL_ID %in% overlaps.delta.dex][treatment == "delta"] -->
<!-- df <- df[, .(SNP, chr, pos_snp, fdr)] -->
<!-- colnames(df) <- c("SNP", "CHR", "BP", "P") -->

<!-- GetManhattanPlot(df, ylims = c(-log10(fdr.thr), -log10(min(df$P)) + 1), plot.title = "Intersect(delta, dex), FDR <= 0.05") -->
<!-- ``` -->

<!-- #### Veh + Delta -->

<!-- ```{r} -->
<!-- overlaps.delta.veh <- setdiff(intersect(meqtls.meqtls$veh, meqtls.meqtls$delta), meqtls.meqtls$dex) -->
<!-- df <- meqtl.all.full.df[meQTL_ID %in% overlaps.delta.veh][treatment == "delta"] -->
<!-- df <- df[, .(SNP, chr, pos_snp, fdr)] -->
<!-- colnames(df) <- c("SNP", "CHR", "BP", "P") -->

<!-- GetManhattanPlot(df, ylims = c(-log10(fdr.thr), -log10(min(df$P)) + 1), plot.title = "Intersect(delta, veh), FDR <= 0.05") -->
<!-- ``` -->

<!-- #### Only delta -->

<!-- ```{r} -->
<!-- only.delta <- setdiff(setdiff(meqtls.meqtls$delta, meqtls.meqtls$dex), meqtls.meqtls$veh) -->
<!-- df <- meqtl.all.full.df[meQTL_ID %in% only.delta] -->
<!-- df <- df[, .(SNP, chr, pos_snp, fdr)] -->
<!-- colnames(df) <- c("SNP", "CHR", "BP", "P") -->

<!-- GetManhattanPlot(df, ylims = c(-log10(fdr.thr), -log10(min(df$P)) + 1), plot.title = "Only delta meQTLs, FDR <= 0.05") -->
<!-- ``` -->

<!-- ### Visualisation {.tabset .tabset-fade .tabset-pills} -->

<!-- ```{r, include=F, eval=T} -->
<!-- eqtm.in.pre  <- "~/bio/code/mpip/dex-stim-human-array/data/integrative/matrixEQTL/" -->

<!-- snp.df.fn <- paste0(eqtm.in.pre, "snp_mtrx.csv") -->
<!-- snp.df    <- fread(snp.df.fn) -->

<!-- methyl.beta.df.fn   <- paste0(eqtm.in.pre, "methyl_beta_mtrx_dex.csv") -->
<!-- methyl.beta.dex.df  <- fread(methyl.beta.df.fn)  -->

<!-- methyl.beta.df.fn   <- paste0(eqtm.in.pre, "methyl_beta_mtrx_veh.csv") -->
<!-- methyl.beta.veh.df  <- fread(methyl.beta.df.fn)  -->

<!-- all(colnames(snp.df)[-1] == colnames(methyl.beta.dex.df)[-1]) -->
<!-- all(colnames(snp.df)[-1] == colnames(methyl.beta.veh.df)[-1]) -->
<!-- ``` -->

<!-- ####  Dex + Veh + Delta -->

<!-- ```{r include = T, eval = T} -->
<!-- # paste0("Scatter plot for the independent significant at FDR <= ", fdr.thr, " meQTLs") -->
<!-- meqtl.id       <- meqtl.all.full.df[meQTL_ID %in% overlaps.all][treatment == "delta"][fdr == min(fdr), meQTL_ID] -->
<!-- selected.meqtl <- meqtl.all.full.df[meQTL_ID %in% meqtl.id] # beta == max(beta) -->

<!-- kable(selected.meqtl %>%  -->
<!--         dplyr::select(SNP, CpG_ID, FC = beta, `t-stat`, `p-value`, FDR = fdr, Treatment = treatment) %>%  -->
<!--         mutate_if(is.numeric, funs(as.character(signif(., 3))))) -->
<!-- ``` -->

<!-- ```{r, eval = T, include = F} -->
<!-- plot.title <- paste0(# "Independent meQTL ", selected.meqtl$meQTL_ID,  -->
<!--                      "The most significant across intersections of DELTA, VEH + DEX meQTLs: ", -->
<!--                      "FC = ", signif(selected.meqtl$beta, 2), " and FDR = ", signif(selected.meqtl$fdr, 2)) -->
<!-- ``` -->
<!-- ```{r out.width = "95%", include = F, eval = F} -->
<!-- GetScatterPlot(meqtl.all.full.df, selected.meqtl, fdr.thr = 0.05, plot.title = plot.title) -->
<!-- ``` -->
<!-- ```{r out.width = "95%", include = T, eval = T} -->
<!-- ProcessGetBoxPlot(methyl.beta.veh.df, methyl.beta.dex.df, snp.df, selected.meqtl[1], fdr.thr = 0.05,  -->
<!--                   plot.title = plot.title) -->
<!-- ```   -->

<!-- ####  Veh + Delta -->

<!-- ```{r include = T, eval = T} -->
<!-- # paste0("Scatter plot for the independent significant at FDR <= ", fdr.thr, " meQTLs") -->
<!-- meqtl.id       <- meqtl.all.full.df[meQTL_ID %in% overlaps.delta.veh][treatment == "delta"][fdr == min(fdr), meQTL_ID] -->
<!-- selected.meqtl <- meqtl.all.full.df[meQTL_ID %in% meqtl.id] # beta == max(beta) -->

<!-- kable(selected.meqtl %>%  -->
<!--         dplyr::select(SNP, CpG_ID, FC = beta, `t-stat`, `p-value`, FDR = fdr, Treatment = treatment) %>%  -->
<!--         mutate_if(is.numeric, funs(as.character(signif(., 3))))) -->
<!-- ``` -->

<!-- ```{r, eval = T, include = F} -->
<!-- plot.title <- paste0(# "Independent meQTL ", selected.meqtl$meQTL_ID,  -->
<!--                      "The most significant across intersections of DELTA and VEH meQTLs: ", -->
<!--                      "FC = ", signif(selected.meqtl$beta, 2), " and FDR = ", signif(selected.meqtl$fdr, 2)) -->
<!-- ``` -->
<!-- ```{r out.width = "95%", include = F, eval = F} -->
<!-- GetScatterPlot(meqtl.all.full.df, selected.meqtl, fdr.thr = 0.05, plot.title = plot.title) -->
<!-- ``` -->
<!-- ```{r out.width = "95%", include = T, eval = T} -->
<!-- ProcessGetBoxPlot(methyl.beta.veh.df, methyl.beta.dex.df, snp.df, selected.meqtl[1], fdr.thr = 0.05,  -->
<!--                   plot.title = plot.title) -->
<!-- ```   -->

<!-- ####  Dex + Delta -->

<!-- ```{r include = T, eval = T} -->
<!-- # paste0("Scatter plot for the independent significant at FDR <= ", fdr.thr, " meQTLs") -->
<!-- meqtl.id       <- meqtl.all.full.df[meQTL_ID %in% overlaps.delta.dex][treatment == "delta"][fdr == min(fdr), meQTL_ID] -->
<!-- selected.meqtl <- meqtl.all.full.df[meQTL_ID %in% meqtl.id] # beta == max(beta) -->

<!-- kable(selected.meqtl %>%  -->
<!--         dplyr::select(SNP, CpG_ID, FC = beta, `t-stat`, `p-value`, FDR = fdr, Treatment = treatment) %>%  -->
<!--         mutate_if(is.numeric, funs(as.character(signif(., 3))))) -->
<!-- ``` -->

<!-- ```{r, eval = T, include = F} -->
<!-- plot.title <- paste0(# "Independent meQTL ", selected.meqtl$meQTL_ID,  -->
<!--                      "The most significant across intersections of DELTA and DEX meQTLs: ", -->
<!--                      "FC = ", signif(selected.meqtl$beta, 2), " and FDR = ", signif(selected.meqtl$fdr, 2)) -->
<!-- ``` -->
<!-- ```{r out.width = "95%", include = F, eval = F} -->
<!-- GetScatterPlot(meqtl.all.full.df, selected.meqtl, fdr.thr = 0.05, plot.title = plot.title) -->
<!-- ``` -->
<!-- ```{r out.width = "95%", include = T, eval = T} -->
<!-- ProcessGetBoxPlot(methyl.beta.veh.df, methyl.beta.dex.df, snp.df, selected.meqtl[16], fdr.thr = 0.05,  -->
<!--                   plot.title = plot.title) -->
<!-- ```   -->

<!-- #### Only delta -->

<!-- ```{r include = T, eval = T} -->
<!-- # paste0("Scatter plot for the independent significant at FDR <= ", fdr.thr, " meQTLs") -->
<!-- meqtl.id       <- meqtl.all.full.df[meQTL_ID %in% only.delta][treatment == "delta"][fdr == min(fdr), meQTL_ID] -->
<!-- selected.meqtl <- meqtl.all.full.df[meQTL_ID %in% meqtl.id] # beta == max(beta) -->

<!-- kable(selected.meqtl %>%  -->
<!--         dplyr::select(SNP, CpG_ID, FC = beta, `t-stat`, `p-value`, FDR = fdr, Treatment = treatment) %>%  -->
<!--         mutate_if(is.numeric, funs(as.character(signif(., 3))))) -->
<!-- ``` -->

<!-- ```{r, eval = T, include = F} -->
<!-- plot.title <- paste0(# "Independent meQTL ", selected.meqtl$meQTL_ID,  -->
<!--                      "The most significant across unique DELTA meQTLs: ", -->
<!--                      "FC = ", signif(selected.meqtl$beta, 2), " and FDR = ", signif(selected.meqtl$fdr, 2)) -->
<!-- ``` -->
<!-- ```{r out.width = "95%", include = F, eval = F} -->
<!-- GetScatterPlot(meqtl.all.full.df, selected.meqtl, fdr.thr = 0.05, plot.title = plot.title) -->
<!-- ``` -->
<!-- ```{r out.width = "95%", include = T, eval = T} -->
<!-- ProcessGetBoxPlot(methyl.beta.veh.df, methyl.beta.dex.df, snp.df, selected.meqtl[1], fdr.thr = 0.05,  -->
<!--                   plot.title = plot.title) -->
<!-- ```   -->


<!-- ### SNP annotation  -->

<!-- ```{r eval = T, include = F} -->
<!-- out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/primary_meqtl_results/" -->

<!-- meqtl.overlap.all.gr       <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_delta_dex_veh_gr.rds")) -->
<!-- meqtl.overlap.delta.veh.gr <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_delta_veh_gr.rds")) -->
<!-- meqtl.overlap.delta.dex.gr <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_delta_dex_gr.rds")) -->
<!-- meqtl.only.delta.gr        <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_only_delta_gr.rds")) -->

<!-- meqtl.overlap.snp.dex.veh.gr <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_dex_veh_gr.rds")) -->
<!-- meqtl.only.snp.dex.gr        <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_only_dex_gr.rds")) -->
<!-- meqtl.only.snp.veh.gr        <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_only_veh_gr.rds")) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- delta.anno.stat.df <- meqtl.only.delta.gr@annoStat -->
<!-- delta.anno.stat.df[["Model"]] <- paste0("only delta \n N = ", length(meqtl.only.delta.gr@anno))  -->

<!-- all.anno.stat.df <- meqtl.overlap.all.gr@annoStat -->
<!-- all.anno.stat.df[["Model"]] <- paste0("intersect \n (delta, veh, dex), \n N = ", length(meqtl.overlap.all.gr@anno)) -->

<!-- delta.veh.anno.stat.df <- meqtl.overlap.delta.veh.gr@annoStat -->
<!-- delta.veh.anno.stat.df[["Model"]] <- paste0("intersect \n (delta, veh) \n N = ", length(meqtl.overlap.delta.veh.gr@anno)) -->

<!-- delta.dex.anno.stat.df <- meqtl.overlap.delta.dex.gr@annoStat -->
<!-- delta.dex.anno.stat.df[["Model"]] <-  paste0("intersect \n (delta, dex) \n N = ", length(meqtl.overlap.delta.dex.gr@anno)) -->

<!-- dex.veh.anno.stat.df <- meqtl.overlap.snp.dex.veh.gr@annoStat -->
<!-- dex.veh.anno.stat.df[["Model"]] <-  paste0("intersect \n (dex, veh) \n N = ", length(meqtl.overlap.snp.dex.veh.gr@anno)) -->

<!-- dex.anno.stat.df <- meqtl.only.snp.dex.gr@annoStat -->
<!-- dex.anno.stat.df[["Model"]] <-  paste0("only dex \n N = ", length(meqtl.only.snp.dex.gr@anno)) -->

<!-- veh.anno.stat.df <- meqtl.only.snp.veh.gr@annoStat -->
<!-- veh.anno.stat.df[["Model"]] <-  paste0("only veh \n N = ", length(meqtl.only.snp.veh.gr@anno)) -->

<!-- anno.stat.df <- rbind(all.anno.stat.df, delta.veh.anno.stat.df, delta.dex.anno.stat.df, delta.anno.stat.df, -->
<!--                       dex.veh.anno.stat.df, dex.anno.stat.df, veh.anno.stat.df) -->

<!-- anno.stat.df$Feature   <- factor(anno.stat.df$Feature) -->
<!-- anno.stat.df$Frequency <- signif(anno.stat.df$Frequency, 3) -->
<!-- anno.stat.df$Model     <- factor(anno.stat.df$Model, -->
<!--                                  levels = c(unique(dex.veh.anno.stat.df$Model),  -->
<!--                                             unique(dex.anno.stat.df$Model),  -->
<!--                                             unique(veh.anno.stat.df$Model), -->
<!--                                             unique(all.anno.stat.df$Model),  -->
<!--                                             unique(delta.veh.anno.stat.df$Model),  -->
<!--                                             unique(delta.dex.anno.stat.df$Model),  -->
<!--                                             unique(delta.anno.stat.df$Model)), -->
<!--                                  ordered = T) -->

<!-- cbPalette <- c("#88CCEE", "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888") -->

<!-- ggplot(anno.stat.df, aes(x = Model, y = Frequency, fill = Feature)) +  -->
<!--   geom_bar( position = "fill", stat = "identity") +  -->
<!--   geom_text(data = subset(anno.stat.df, Frequency > 2), -->
<!--             aes(label = Frequency),  -->
<!--             position = position_fill(vjust = .5), size = 2) + -->
<!--   labs(x = "", -->
<!--        y = "Percentage of SNPs",  -->
<!--        title = "meSNP annotation from UCSC for the hg19 genome build using \nTxDb.Hsapiens.UCSC.hg19.knownGene and ChIPseeker Bioconductor R packages \n\nmeQTL level, N is the number of SNPs") + -->
<!--   theme(legend.position = "right", -->
<!--         panel.grid.major = element_blank(), -->
<!--         panel.background = element_blank(), -->
<!--         plot.title = element_text(size = 8), -->
<!--         axis.title = element_text(size = 8), -->
<!--         axis.text.x = element_text(angle = 0, hjust = 0.5, size = 7)) + -->
<!--    scale_fill_manual("", values = cbPalette) -->
<!-- ``` -->

<!-- ### SNP annotation by groups -->

<!-- ```{r eval = T, include = F} -->
<!-- out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/primary_meqtl_results/" -->

<!-- meqtl.opposite.fc.grp.with.delta.gr <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_opposite_fc_grp_with_delta_gr.rds")) -->
<!-- meqtl.opposite.fc.grp.no.delta.gr  <- readRDS(paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_opposite_fc_grp_no_delta_gr.rds")) -->

<!-- meqtl.opposite.fc.grp.veh.dex.gr <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_opposite_fc_grp_veh_dex_gr.rds")) -->
<!-- meqtl.opposite.fc.grp.delta.gr   <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_opposite_fc_grp_delta_gr.rds")) -->
<!-- meqtl.opposite.fc.gr             <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_snp_anno_opposite_fc_grp_gr.rds")) -->

<!-- # meqtl.opposite.fc.df <- data.frame(meqtl.opposite.fc.gr@anno) %>% setDT() -->
<!-- # gene.df <- meqtl.opposite.fc.df[, .(geneId, ENSEMBL, SYMBOL, GENENAME)] %>% unique() -->
<!-- # fwrite(gene.df, -->
<!-- #        paste0(out.dir.pre, "../genes/meqtl_genes_opposite_fc_gr.csv"), -->
<!-- #        quote = F, row.names = F, sep   = "\t") -->

<!-- ``` -->

<!-- ```{r  out.width = "95%", include = T, eval = T} -->
<!-- opposite.fc.grp.veh.dex.anno.stat.df <- meqtl.opposite.fc.grp.veh.dex.gr@annoStat -->
<!-- opposite.fc.grp.veh.dex.anno.stat.df[["Model"]] <- paste0("veh + dex with parallel FC \n N = ", length(meqtl.opposite.fc.grp.veh.dex.gr@anno))  -->

<!-- opposite.fc.grp.delta.anno.stat.df <- meqtl.opposite.fc.grp.delta.gr@annoStat -->
<!-- opposite.fc.grp.delta.anno.stat.df[["Model"]] <- paste0("delta with parallel FC \n N = ", length(meqtl.opposite.fc.grp.delta.gr@anno))  -->

<!-- meqtl.opposite.fc.anno.stat.df <- meqtl.opposite.fc.gr@annoStat -->
<!-- meqtl.opposite.fc.anno.stat.df[["Model"]] <- paste0("opposite FC \n N = ", length(meqtl.opposite.fc.gr@anno))  -->

<!-- anno.stat.df <- rbind(opposite.fc.grp.veh.dex.anno.stat.df, opposite.fc.grp.delta.anno.stat.df, meqtl.opposite.fc.anno.stat.df) -->

<!-- anno.stat.df$Feature   <- factor(anno.stat.df$Feature) -->
<!-- anno.stat.df$Frequency <- signif(anno.stat.df$Frequency, 3) -->
<!-- anno.stat.df$Model     <- factor(anno.stat.df$Model, -->
<!--                                  levels = c(unique(opposite.fc.grp.veh.dex.anno.stat.df$Model),  -->
<!--                                             unique(opposite.fc.grp.delta.anno.stat.df$Model),  -->
<!--                                             unique(meqtl.opposite.fc.anno.stat.df$Model)), -->
<!--                                  ordered = T) -->
<!-- cbPalette <- c("#88CCEE", "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888") -->

<!-- ggplot(anno.stat.df, aes(x = Model, y = Frequency, fill = Feature)) +  -->
<!--   geom_bar( position = "fill", stat = "identity") +  -->
<!--   geom_text(data = subset(anno.stat.df, Frequency > 2), -->
<!--             aes(label = Frequency),  -->
<!--             position = position_fill(vjust = .5), size = 2) + -->
<!--   labs(x = "", -->
<!--        y = "Percentage of SNPs",  -->
<!--        title = "meSNP annotation from UCSC for the hg19 genome build using \nTxDb.Hsapiens.UCSC.hg19.knownGene and ChIPseeker Bioconductor R packages \n\nmeQTL level, N is the number of SNPs") + -->
<!--   theme(legend.position = "right", -->
<!--         panel.grid.major = element_blank(), -->
<!--         panel.background = element_blank(), -->
<!--         plot.title = element_text(size = 8), -->
<!--         axis.title = element_text(size = 8), -->
<!--         axis.text.x = element_text(angle = 0, hjust = 0.5, size = 7)) + -->
<!--    scale_fill_manual("", values = cbPalette) -->
<!-- ``` -->

<!-- #### Wilcoxon test  -->

<!-- ```{r} -->
<!-- anno.stat.df <- rbind(opposite.fc.grp.veh.dex.anno.stat.df, opposite.fc.grp.delta.anno.stat.df, meqtl.opposite.fc.anno.stat.df) -->

<!-- wilcox.test.rslt <- pairwise.wilcox.test(anno.stat.df$Frequency, anno.stat.df$Model, p.adjust.method = "none", paired = F)$p.value %>% -->
<!--   signif(3) -->

<!-- colnames(wilcox.test.rslt) <- c("parallel FC (delta)", "opposite FC") -->
<!-- rownames(wilcox.test.rslt) <- c("opposite FC", "parallel FC (veh&dex)") -->
<!-- wilcox.test.rslt -->

<!-- # wilcox.test(anno.stat.df$Frequency[1:22] ~ anno.stat.df$Model[1:22]) -->
<!-- ``` -->

<!-- #### SNP distribution around TSS -->

<!-- ```{r out.width = "95%", include = T, eval = T}  -->
<!-- plotDistToTSS(list(Parallel_FC = meqtl.opposite.fc.grp.veh.dex.gr,  -->
<!--                    Parallel_FC_Delta = meqtl.opposite.fc.grp.delta.gr,  -->
<!--                    Opposite_FC = meqtl.opposite.fc.gr), -->
<!--               xlab = " ", -->
<!--               ylab = "SNPs (%) (5'->3')",  -->
<!--               title = "SNP distribution around TSS") + -->
<!--     theme(legend.position = "bottom",  -->
<!--           legend.title = element_blank(), -->
<!--           panel.grid.major = element_blank(), -->
<!--           panel.background = element_blank(), -->
<!--           plot.title = element_text(size = 10), -->
<!--           axis.title = element_text(size = 8), -->
<!--           panel.border = element_blank()) + -->
<!--   guides(fill = guide_legend(ncol = 6, byrow = F)) -->

<!-- ``` -->

<!-- ### CpG annotation -->

<!-- <!-- Annotate data --> -->

<!-- ```{r include = F, eval = F} -->
<!-- library(TxDb.Hsapiens.UCSC.hg19.knownGene) -->
<!-- library(ChIPseeker) -->
<!-- library(org.Hs.eg.db) -->

<!-- out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/primary_meqtl_results/" -->

<!-- meqtl.overlap.cpg.all.df <- meqtl.all.full.df[meQTL_ID %in% meqtl.grp.overlaps.all] -->
<!-- meqtl.overlap.cpg.all.gr <- annotate_with_chipseeker(meqtl.overlap.cpg.all.df, "chr", "pos_cpg", "pos_cpg", "CpG_ID") -->
<!-- saveRDS(meqtl.overlap.cpg.all.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_delta_dex_veh_gr.rds")) -->

<!-- meqtl.overlap.cpg.delta.veh.df <- meqtl.all.full.df[meQTL_ID %in% meqtl.grp.overlaps.delta.veh] -->
<!-- meqtl.overlap.cpg.delta.veh.gr <- annotate_with_chipseeker(meqtl.overlap.cpg.delta.veh.df, "chr", "pos_cpg", "pos_cpg", "CpG_ID") -->
<!-- saveRDS(meqtl.overlap.cpg.delta.veh.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_delta_veh_gr.rds")) -->

<!-- meqtl.overlap.cpg.delta.dex.df <- meqtl.all.full.df[meQTL_ID %in% meqtl.grp.overlaps.delta.dex] -->
<!-- meqtl.overlap.cpg.delta.dex.gr <- annotate_with_chipseeker(meqtl.overlap.cpg.delta.dex.df, "chr", "pos_cpg", "pos_cpg", "CpG_ID") -->
<!-- saveRDS(meqtl.overlap.cpg.delta.dex.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_delta_dex_gr.rds")) -->

<!-- meqtl.only.cpg.delta.df <- meqtl.all.full.df[meQTL_ID %in% meqtl.grp.only.delta] -->
<!-- meqtl.only.cpg.delta.gr <- annotate_with_chipseeker(meqtl.only.cpg.delta.df, "chr", "pos_cpg", "pos_cpg", "CpG_ID") -->
<!-- saveRDS(meqtl.only.cpg.delta.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_only_delta_gr.rds")) -->

<!-- meqtl.overlap.snp.dex.veh.df <- meqtl.all.full.df[meQTL_ID %in% meqtl.grp.overlaps.dex.veh] -->
<!-- meqtl.overlap.cpg.dex.veh.gr <- annotate_with_chipseeker(meqtl.overlap.snp.dex.veh.df, "chr", "pos_cpg", "pos_cpg", "CpG_ID") -->
<!-- saveRDS(meqtl.overlap.cpg.dex.veh.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_dex_veh_gr.rds")) -->

<!-- meqtl.only.cpg.dex.df <- meqtl.all.full.df[meQTL_ID %in% meqtl.grp.only.dex] -->
<!-- meqtl.only.cpg.dex.gr <- annotate_with_chipseeker(meqtl.only.cpg.dex.df, "chr", "pos_cpg", "pos_cpg", "CpG_ID") -->
<!-- saveRDS(meqtl.only.cpg.dex.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_only_dex_gr.rds")) -->

<!-- meqtl.only.cpg.veh.df <- meqtl.all.full.df[meQTL_ID %in% meqtl.grp.only.veh] -->
<!-- meqtl.only.cpg.veh.gr <- annotate_with_chipseeker(meqtl.only.cpg.veh.df, "chr", "pos_cpg", "pos_cpg", "CpG_ID") -->
<!-- saveRDS(meqtl.only.cpg.veh.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_only_veh_gr.rds")) -->

<!-- meqtl.opposite.fc.grp.with.delta.df <- meqtl.all.full.df[meQTL_ID %in% meqtls.opposite.fc.grp.with.delta$meQTL_ID] -->
<!-- meqtl.opposite.fc.grp.with.delta.gr <- annotate_with_chipseeker(meqtl.opposite.fc.grp.with.delta.df, "chr", "pos_cpg", "pos_cpg", "CpG_ID") -->
<!-- saveRDS(meqtl.opposite.fc.grp.with.delta.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_opposite_fc_grp_with_delta_gr.rds")) -->

<!-- meqtl.opposite.fc.grp.no.delta.df <- meqtl.all.full.df[meQTL_ID %in% meqtls.opposite.fc.grp.no.delta$meQTL_ID] -->
<!-- meqtl.opposite.fc.grp.no.delta.gr <- annotate_with_chipseeker(meqtl.opposite.fc.grp.no.delta.df, "chr", "pos_cpg", "pos_cpg", "CpG_ID") -->
<!-- saveRDS(meqtl.opposite.fc.grp.no.delta.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_opposite_fc_grp_no_delta_gr.rds")) -->

<!-- meqtl.opposite.fc.grp.veh.dex.df  <- rbind(meqtl.overlap.cpg.dex.veh.df,  -->
<!--                                            meqtl.only.cpg.dex.df,  -->
<!--                                            meqtl.only.cpg.veh.df)[!(meQTL_ID %in% meqtls.opposite.fc.grp.no.delta$meQTL_ID)][, .(chr, pos_cpg, CpG_ID)] %>% unique() -->
<!-- meqtl.opposite.fc.grp.veh.dex.gr <- annotate_with_chipseeker(meqtl.opposite.fc.grp.veh.dex.df, "chr", "pos_cpg", "pos_cpg", "CpG_ID") -->
<!-- saveRDS(meqtl.opposite.fc.grp.veh.dex.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_opposite_fc_grp_veh_dex_gr.rds")) -->

<!-- meqtl.opposite.fc.grp.delta.df <- rbind(meqtl.overlap.cpg.all.df,  -->
<!--                                          meqtl.overlap.cpg.delta.veh.df,  -->
<!--                                          meqtl.overlap.cpg.delta.dex.df,  -->
<!--                                          meqtl.only.cpg.delta.df)[!(meQTL_ID %in% meqtls.opposite.fc.grp.with.delta$meQTL_ID)][, .(chr, pos_cpg, CpG_ID)] %>% unique() -->
<!-- meqtl.opposite.fc.grp.delta.gr <- annotate_with_chipseeker(meqtl.opposite.fc.grp.delta.df, "chr", "pos_cpg", "pos_cpg", "CpG_ID") -->
<!-- saveRDS(meqtl.opposite.fc.grp.delta.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_opposite_fc_grp_delta_gr.rds")) -->

<!-- meqtl.opposite.fc.df <- rbind(meqtl.opposite.fc.grp.with.delta.df, meqtl.opposite.fc.grp.no.delta.df)[, .(chr, pos_cpg, CpG_ID)] %>% unique() -->
<!-- meqtl.opposite.fc.gr <- annotate_with_chipseeker(meqtl.opposite.fc.df, "chr", "pos_cpg", "pos_cpg", "CpG_ID") -->
<!-- saveRDS(meqtl.opposite.fc.gr, file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_opposite_fc_grp_gr.rds")) -->

<!-- ``` -->

<!-- ```{r eval = T, include = F} -->
<!-- out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/primary_meqtl_results/" -->

<!-- meqtl.overlap.all.gr       <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_delta_dex_veh_gr.rds")) -->
<!-- meqtl.overlap.delta.veh.gr <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_delta_veh_gr.rds")) -->
<!-- meqtl.overlap.delta.dex.gr <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_delta_dex_gr.rds")) -->
<!-- meqtl.only.delta.gr        <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_only_delta_gr.rds")) -->

<!-- meqtl.overlap.dex.veh.gr <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_dex_veh_gr.rds")) -->
<!-- meqtl.only.dex.gr        <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_only_dex_gr.rds")) -->
<!-- meqtl.only.veh.gr        <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_only_veh_gr.rds")) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- delta.anno.stat.df <- meqtl.only.delta.gr@annoStat -->
<!-- delta.anno.stat.df[["Model"]] <- paste0("only delta \n N = ", length(meqtl.only.delta.gr@anno))  -->

<!-- all.anno.stat.df <- meqtl.overlap.all.gr@annoStat -->
<!-- all.anno.stat.df[["Model"]] <- paste0("intersect \n (delta, veh, dex), \n N = ", length(meqtl.overlap.all.gr@anno)) -->

<!-- delta.veh.anno.stat.df <- meqtl.overlap.delta.veh.gr@annoStat -->
<!-- delta.veh.anno.stat.df[["Model"]] <- paste0("intersect \n (delta, veh) \n N = ", length(meqtl.overlap.delta.veh.gr@anno)) -->

<!-- delta.dex.anno.stat.df <- meqtl.overlap.delta.dex.gr@annoStat -->
<!-- delta.dex.anno.stat.df[["Model"]] <-  paste0("intersect \n (delta, dex) \n N = ", length(meqtl.overlap.delta.dex.gr@anno)) -->

<!-- dex.veh.anno.stat.df <- meqtl.overlap.dex.veh.gr@annoStat -->
<!-- dex.veh.anno.stat.df[["Model"]] <-  paste0("intersect \n (dex, veh) \n N = ", length(meqtl.overlap.dex.veh.gr@anno)) -->

<!-- dex.anno.stat.df <- meqtl.only.dex.gr@annoStat -->
<!-- dex.anno.stat.df[["Model"]] <-  paste0("only dex \n N = ", length(meqtl.only.dex.gr@anno)) -->

<!-- veh.anno.stat.df <- meqtl.only.veh.gr@annoStat -->
<!-- veh.anno.stat.df[["Model"]] <-  paste0("only veh \n N = ", length(meqtl.only.veh.gr@anno)) -->

<!-- anno.stat.df <- rbind(all.anno.stat.df, delta.veh.anno.stat.df, delta.dex.anno.stat.df, delta.anno.stat.df, -->
<!--                       dex.veh.anno.stat.df, dex.anno.stat.df, veh.anno.stat.df) -->

<!-- anno.stat.df$Feature   <- factor(anno.stat.df$Feature) -->
<!-- anno.stat.df$Frequency <- signif(anno.stat.df$Frequency, 3) -->
<!-- anno.stat.df$Model     <- factor(anno.stat.df$Model, -->
<!--                                  levels = c(unique(dex.veh.anno.stat.df$Model),  -->
<!--                                             unique(dex.anno.stat.df$Model),  -->
<!--                                             unique(veh.anno.stat.df$Model), -->
<!--                                             unique(all.anno.stat.df$Model),  -->
<!--                                             unique(delta.veh.anno.stat.df$Model),  -->
<!--                                             unique(delta.dex.anno.stat.df$Model),  -->
<!--                                             unique(delta.anno.stat.df$Model)), -->
<!--                                  ordered = T) -->
<!-- cbPalette <- c("#88CCEE", "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888") -->

<!-- ggplot(anno.stat.df, aes(x = Model, y = Frequency, fill = Feature)) +  -->
<!--   geom_bar( position = "fill", stat = "identity") +  -->
<!--   geom_text(data = subset(anno.stat.df, Frequency > 2), -->
<!--             aes(label = Frequency),  -->
<!--             position = position_fill(vjust = .5), size = 2) + -->
<!--   labs(x = "", -->
<!--        y = "Percentage of CpGs",  -->
<!--        title = "meCpG annotation from UCSC for the hg19 genome build using \nTxDb.Hsapiens.UCSC.hg19.knownGene and ChIPseeker Bioconductor R packages \n\nmeQTL level, N is the number of CpGs") + -->
<!--   theme(legend.position = "right", -->
<!--         panel.grid.major = element_blank(), -->
<!--         panel.background = element_blank(), -->
<!--         plot.title = element_text(size = 8), -->
<!--         axis.title = element_text(size = 8), -->
<!--         axis.text.x = element_text(angle = 0, hjust = 0.5, size = 7)) + -->
<!--    scale_fill_manual("", values = cbPalette) -->
<!-- ``` -->

<!-- ### CpG annotation by groups -->

<!-- ```{r eval = T, include = F} -->
<!-- out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/primary_meqtl_results/" -->

<!-- meqtl.opposite.fc.grp.with.delta.gr <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_opposite_fc_grp_with_delta_gr.rds")) -->
<!-- meqtl.opposite.fc.grp.no.delta.gr   <- readRDS(paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_opposite_fc_grp_no_delta_gr.rds")) -->

<!-- meqtl.opposite.fc.grp.veh.dex.gr <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_opposite_fc_grp_veh_dex_gr.rds")) -->
<!-- meqtl.opposite.fc.grp.delta.gr   <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_opposite_fc_grp_delta_gr.rds")) -->
<!-- meqtl.opposite.fc.gr             <- readRDS(file =  paste0(out.dir.pre, "meqtl_overlaps_meqtl_level_cpg_anno_opposite_fc_grp_gr.rds")) -->
<!-- ``` -->

<!-- ```{r out.width = "95%", include = T, eval = T} -->
<!-- opposite.fc.grp.veh.dex.anno.stat.df <- meqtl.opposite.fc.grp.veh.dex.gr@annoStat -->
<!-- opposite.fc.grp.veh.dex.anno.stat.df[["Model"]] <- paste0("veh + dex with parallel FC \n N = ", length(meqtl.opposite.fc.grp.veh.dex.gr@anno))  -->

<!-- opposite.fc.grp.delta.anno.stat.df <- meqtl.opposite.fc.grp.delta.gr@annoStat -->
<!-- opposite.fc.grp.delta.anno.stat.df[["Model"]] <- paste0("delta with parallel FC \n N = ", length(meqtl.opposite.fc.grp.delta.gr@anno))  -->

<!-- meqtl.opposite.fc.anno.stat.df <- meqtl.opposite.fc.gr@annoStat -->
<!-- meqtl.opposite.fc.anno.stat.df[["Model"]] <- paste0("opposite FC \n N = ", length(meqtl.opposite.fc.gr@anno))  -->

<!-- anno.stat.df <- rbind(opposite.fc.grp.veh.dex.anno.stat.df, opposite.fc.grp.delta.anno.stat.df, meqtl.opposite.fc.anno.stat.df) -->

<!-- anno.stat.df$Feature   <- factor(anno.stat.df$Feature) -->
<!-- anno.stat.df$Frequency <- signif(anno.stat.df$Frequency, 3) -->
<!-- anno.stat.df$Model     <- factor(anno.stat.df$Model, -->
<!--                                  levels = c(unique(opposite.fc.grp.veh.dex.anno.stat.df$Model),  -->
<!--                                             unique(opposite.fc.grp.delta.anno.stat.df$Model),  -->
<!--                                             unique(meqtl.opposite.fc.anno.stat.df$Model)), -->
<!--                                  ordered = T) -->

<!-- cbPalette <- c("#88CCEE", "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888") -->

<!-- ggplot(anno.stat.df, aes(x = Model, y = Frequency, fill = Feature)) +  -->
<!--   geom_bar( position = "fill", stat = "identity") +  -->
<!--   geom_text(data = subset(anno.stat.df, Frequency > 2), -->
<!--             aes(label = Frequency),  -->
<!--             position = position_fill(vjust = .5), size = 2) + -->
<!--   labs(x = "", -->
<!--        y = "Percentage of CpGs",  -->
<!--        title = "meCpG annotation from UCSC for the hg19 genome build using \nTxDb.Hsapiens.UCSC.hg19.knownGene and ChIPseeker Bioconductor R packages \n\nmeQTL level, N is the number of CpGs") + -->
<!--   theme(legend.position = "right", -->
<!--         panel.grid.major = element_blank(), -->
<!--         panel.background = element_blank(), -->
<!--         plot.title = element_text(size = 8), -->
<!--         axis.title = element_text(size = 8), -->
<!--         axis.text.x = element_text(angle = 0, hjust = 0.5, size = 7)) + -->
<!--    scale_fill_manual("", values = cbPalette) -->
<!-- ``` -->

<!-- #### Wilcoxon test  -->

<!-- ```{r} -->
<!-- anno.stat.df <- rbind(opposite.fc.grp.veh.dex.anno.stat.df, opposite.fc.grp.delta.anno.stat.df, meqtl.opposite.fc.anno.stat.df) -->

<!-- wilcox.test.rslt <- pairwise.wilcox.test(anno.stat.df$Frequency, anno.stat.df$Model, p.adjust.method = "none", paired = F)$p.value %>% -->
<!--   signif(3) -->

<!-- colnames(wilcox.test.rslt) <- c("parallel FC (delta)", "opposite FC") -->
<!-- rownames(wilcox.test.rslt) <- c("opposite FC", "parallel FC (veh&dex)") -->
<!-- wilcox.test.rslt -->

<!-- # wilcox.test(anno.stat.df$Frequency[1:22] ~ anno.stat.df$Model[1:22]) -->
<!-- ``` -->

<!-- #### SNP distribution around TSS -->

<!-- ```{r out.width = "95%", include = T, eval = T}  -->
<!-- plotDistToTSS(list(Parallel_FC = meqtl.opposite.fc.grp.veh.dex.gr,  -->
<!--                    Parallel_FC_Delta = meqtl.opposite.fc.grp.delta.gr,  -->
<!--                    Opposite_FC = meqtl.opposite.fc.gr), -->
<!--               xlab = " ", -->
<!--               ylab = "CpGs (%) (5'->3')",  -->
<!--               title = "CpG distribution around TSS") + -->
<!--     theme(legend.position = "bottom",  -->
<!--           legend.title = element_blank(), -->
<!--           panel.grid.major = element_blank(), -->
<!--           panel.background = element_blank(), -->
<!--           plot.title = element_text(size = 10), -->
<!--           axis.title = element_text(size = 8), -->
<!--           panel.border = element_blank()) + -->
<!--   guides(fill = guide_legend(ncol = 6, byrow = F)) -->

<!-- ``` -->