---
title: "cis-eQTM Analysis"
subtitle: "Dexamethasone-Stimulated Human Array Project"
author: 
  - name: Anastasiia Hryhorzhevska
    email: anastasiia_hry@psych.mpg.de
    url: https://github.com/ahryho
    affiliation: Max Planck Institute of Psychiatry
    affiliation_url: https://www.psych.mpg.de/2664393/medizinische-genomforschung
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    css: style.css
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
options(digits = 4, width = 100, stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE,
                      tidy.opts = list(width.cutoff = 100),
                      tidy=TRUE,
                      fig.pos = "H",
                      dpi = 200,
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE, 
                      cache.lazy = FALSE)

# define DT as the standard printing method for data.frame
library(knitr)
library(DT)
knit_print.data.frame <- function(x, ...) {
  knit_print(DT::datatable(x,
                           filter="top",
                           rownames = FALSE,
                           extensions = "FixedColumns",
                           options = list(
                             scrollX = TRUE,
                             pageLength = 5
                             
                           )), ...)
}

registerS3method("knit_print", "data.frame", knit_print.data.frame)

library(data.table)
library(dplyr)

library(ggplot2)
library(ggrepel)
library(scales)
library(GGally)
library(factoextra)
library(viridis)
```

***
# __MatrixEQTL Primary Results__
***

```{r load-func, include = F, eval = T}
source("~/bio/code/mpip/dex-stim-human-array/code/integrative/util.R")
cbPalette <- c( "#0072B2", "#009E73", "#E69F00", "#F0E442", "#D55E00", "#CC79A7", "#56B4E9", "#999999")
```

```{r, include = F, eval = T}
src.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/data/integrative/matrixEQTL/"
out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/eqtms/"

cpg.loc.fn       <- paste0(src.dir.pre, "cpg_locations.csv")
cpg.loc.colnames <- c("CpG_ID", "chr", "pos_cpg", "pos_end")
cpg.loc          <- fread(cpg.loc.fn, col.names = cpg.loc.colnames)

ensg.loc.fn       <- paste0(src.dir.pre, "ensg_locations.csv")
ensg.loc.colnames <- c("ENSG_ID", "chr", "pos_start_ensg", "pos_end_ensg")
ensg.loc          <- fread(ensg.loc.fn, col.names = ensg.loc.colnames)

meqtl.veh.fn   <- paste0(out.dir.pre, "eqtm_cis_result_veh_beta.csv") 
meqtl.dex.fn   <- paste0(out.dir.pre, "eqtm_cis_result_dex_beta.csv") 
meqtl.delta.fn <- paste0(out.dir.pre, "eqtm_cis_result_delta_no_residuals_or_cov_beta.csv") 
meqtl.delta.res.fn <- paste0(out.dir.pre, "eqtm_cis_result_delta_beta.csv") 

col.names <- c("CpG_ID", "ENSG_ID", "beta", "t-stat", "p-value", "FDR")

dex.df   <- fread(meqtl.dex.fn, col.names = col.names)
veh.df   <- fread(meqtl.veh.fn, col.names = col.names)
delta.df <- fread(meqtl.delta.fn, col.names = col.names)
delta.res.df <- fread(meqtl.delta.res.fn, col.names = col.names)
```

## __Analysis of nominal significant eQTMs, FDR <= 1, p <= 0.05__

```{r, include = F, eval = T}
# head(meqtl.df)
fdr.thr  <- 1

veh.full.df   <- GetFulleQTMdf(veh.df, fdr.thr = fdr.thr, treatment = "baseline")
dex.full.df   <- GetFulleQTMdf(dex.df, fdr.thr = fdr.thr, treatment = "dex")
delta.full.df <- GetFulleQTMdf(delta.df, fdr.thr = fdr.thr, treatment = "delta")
delta.res.full.df <- GetFulleQTMdf(delta.res.df, fdr.thr = fdr.thr, treatment = "delta-res")

all.full.df   <- rbind(dex.full.df, veh.full.df, delta.full.df, delta.res.full.df)
```

### Hits of all eQTMs
<!-- Check the number of found eQTLs, unique CpGs and SNPs in all datasets. -->

```{r, include = F, eval = T}
sign.hits.df <- data.table(c(delta.full.df$fdr  %>% length(), 
                             delta.full.df$CpG_ID %>% unique() %>% length(),
                             delta.full.df$ENSG_ID %>% unique() %>% length()),
                           c(delta.res.full.df$fdr  %>% length(), 
                             delta.res.full.df$CpG_ID %>% unique() %>% length(),
                             delta.res.full.df$ENSG_ID %>% unique() %>% length()),
                           c(dex.full.df$fdr  %>% length(), 
                             dex.full.df$CpG_ID %>% unique() %>% length(),
                             dex.full.df$ENSG_ID %>% unique() %>% length()),
                           c(veh.full.df$fdr  %>% length(), 
                             veh.full.df$CpG_ID %>% unique() %>% length(),
                             veh.full.df$ENSG_ID %>% unique() %>% length()))

sign.hits.df <- as.data.frame(sign.hits.df)
rownames(sign.hits.df) <- c("cis-eQTM", "CpG_ID", "ENSG_ID")
colnames(sign.hits.df) <- c("delta", "delta-res", "dex", "baseline")
```

```{r out.width = "95%", include = T, eval = T}
cbPalette <- c("lightskyblue", "cornflowerblue", "#0072B2", "grey")

sign.hits.df %>% as.matrix() %>% reshape2::melt(var.names = c("ID", "Treatment"), value.name = "cnt") %>%
  ggplot(aes(y = cnt, x = Var2, fill = as.factor(as.numeric(Var2)))) +
    geom_col() +
    geom_text(aes(label = comma(cnt, accuracy = 1L), y = cnt), 
                  stat = "identity", 
                  vjust = -0.2, size = 2.5) + 
    scale_y_continuous(labels = scientific) +
    # scale_fill_viridis(discrete = TRUE, alpha = 0.9, option = "C") +
    labs(title = " ", y = "Count", x = " ") + 
    facet_wrap(~ Var1, ncol = 3) +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank()) +
  scale_fill_manual(values = cbPalette)
```

## Scatter plot

```{r out.width = "95%", include = T, eval = T}
plot.title <- paste0("Independent and significant at FDR <= ", fdr.thr, " eQTMs.",
                     "\nResult of MatrixEQTL")
GetScatterPlot2(all.full.df, plot.title = plot.title, fdr.thr = fdr.thr)
```

## Effect size

```{r out.width = "95%", include = T, eval = T}
ggplot(all.full.df, aes(y = beta, x = treatment, color = treatment)) +
  geom_boxplot() +
  scale_y_continuous(breaks = seq(signif(min(all.full.df$beta), 0) * 1.2, max(all.full.df$beta) * 1.2, by = 10)) +
  theme(legend.position = "none", 
          legend.title = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank()) +
  scale_color_manual(values = cbPalette)
```

## Upset plots{.tabset .tabset-fade .tabset-pills}

### eQTM level
  
```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

eqtms <- list(delta = all.full.df[treatment == "delta", eQTM_ID], 
              delta_res = all.full.df[treatment == "delta_res", eQTM_ID], 
              dex = all.full.df[treatment == "dex", eQTM_ID], 
              veh = all.full.df[treatment == "veh", eQTM_ID])

upset(fromList(eqtms), 
      sets = c("delta", "delta_res", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, eQTM", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c( "#0072B2", "#009E73", "#E69F00", "#F0E442"), 
      order.by = "freq")
```

### Gene level
  
```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

eqtms.ensg <- list(delta = delta.full.df$ENSG_ID, delta_res = delta.res.full.df$ENSG_ID , dex = dex.full.df$ENSG_ID, veh = veh.full.df$ENSG_ID)

upset(fromList(eqtms.ensg), 
      sets = c("delta", "delta_res", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, eGene (ENSG)", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c("#0072B2", "#009E73", "#E69F00", "#F0E442"), 
      order.by = "freq")
```

### CpG level

```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

eqtms.cpg <- list(delta = delta.full.df$CpG_ID, delta_res = delta.res.full.df$CpG_ID, dex = dex.full.df$CpG_ID, veh = veh.full.df$CpG_ID)

upset(fromList(eqtms.cpg), 
      sets = c("delta", "delta_res", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, eCpG", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c("#0072B2", "#009E73", "#E69F00", "#F0E442"), 
      order.by = "freq")
```

## __Analysis of significant eQTMs, FDR <= 0.05__

```{r, include = F, eval = T}
# head(meqtl.df)
fdr.thr  <- 0.05

veh.full.df   <- GetFulleQTMdf(veh.df, fdr.thr = fdr.thr, treatment = "baseline")
dex.full.df   <- GetFulleQTMdf(dex.df, fdr.thr = fdr.thr, treatment = "dex")
delta.full.df <- GetFulleQTMdf(delta.df, fdr.thr = fdr.thr, treatment = "delta")
delta.res.full.df <- GetFulleQTMdf(delta.res.df, fdr.thr = fdr.thr, treatment = "delta_res")

all.full.df   <- rbind(dex.full.df, veh.full.df, delta.full.df, delta.res.full.df)
```

### Significant hits of eQTMs with FDR <= 0.05

<!-- Check the number of found eQTLs, unique CpGs and SNPs in all datasets. -->

```{r, include = F, eval = T}
sign.hits.df <- data.table(c(delta.full.df$fdr  %>% length(), 
                             delta.full.df$CpG_ID %>% unique() %>% length(),
                             delta.full.df$ENSG_ID %>% unique() %>% length()),
                           c(delta.res.full.df$fdr  %>% length(), 
                             delta.res.full.df$CpG_ID %>% unique() %>% length(),
                             delta.res.full.df$ENSG_ID %>% unique() %>% length()),
                           c(dex.full.df$fdr  %>% length(), 
                             dex.full.df$CpG_ID %>% unique() %>% length(),
                             dex.full.df$ENSG_ID %>% unique() %>% length()),
                           c(veh.full.df$fdr  %>% length(), 
                             veh.full.df$CpG_ID %>% unique() %>% length(),
                             veh.full.df$ENSG_ID %>% unique() %>% length()))

sign.hits.df <- as.data.frame(sign.hits.df)
rownames(sign.hits.df) <- c("cis-eQTL", "CpG_ID", "ENSG_ID")
colnames(sign.hits.df) <- c("delta", "delta_res", "dex", "baseline")
```

```{r out.width = "95%", include = T, eval = T}
cbPalette <- c("lightskyblue", "cornflowerblue", "#0072B2", "grey")

sign.hits.df %>% as.matrix() %>% reshape2::melt(var.names = c("ID", "Treatment"), value.name = "cnt") %>%
  ggplot(aes(y = cnt, x = Var2, fill = as.factor(as.numeric(Var2)))) +
    geom_col() +
    geom_text(aes(label = comma(cnt, accuracy = 1L), y = cnt), 
                  stat = "identity", 
                  vjust = -0.2, size = 2.5) + 
    scale_y_continuous(labels = scientific) +
    labs(title = " ", y = "Count", x = " ") + 
    facet_wrap(~ Var1, ncol = 3) +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank()) +
  scale_fill_manual(values = cbPalette)
```

## Scatter plot

```{r out.width = "95%", include = T, eval = T}
plot.title <- paste0("Independent and significant at FDR <= ", fdr.thr, " eQTMs.",
                     "\nResult of MatrixEQTL")
GetScatterPlot2(all.full.df, plot.title = plot.title, fdr.thr = fdr.thr)
```

## Effect size

```{r out.width = "95%", include = T, eval = T}
ggplot(all.full.df, aes(y = beta, x = treatment, color = treatment)) +
  geom_boxplot() +
  scale_y_continuous(breaks = seq(signif(min(all.full.df$beta), 0) * 1.2, max(all.full.df$beta) * 1.2, by = 10)) +
  theme(legend.position = "none", 
          legend.title = element_blank(),
          # panel.grid.major = element_blank(),
          # panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank()) +
  scale_color_manual(values = cbPalette)
```

## Upset plots{.tabset .tabset-fade .tabset-pills}

### eQTM level
  
```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

eqtms <- list(delta = all.full.df[treatment == "delta", eQTM_ID], 
              delta_res = all.full.df[treatment == "delta_res", eQTM_ID], 
              dex = all.full.df[treatment == "dex", eQTM_ID], 
              veh = all.full.df[treatment == "veh", eQTM_ID])

upset(fromList(eqtms), 
      sets = c("delta", "delta_res", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, eQTM", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c( "#0072B2", "#009E73", "#E69F00", "#F0E442"), 
      order.by = "freq")
```

### Gene level
  
```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

eqtms.ensg <- list(delta = delta.full.df$ENSG_ID, delta_res = delta.res.full.df$ENSG_ID , dex = dex.full.df$ENSG_ID, veh = veh.full.df$ENSG_ID)

upset(fromList(eqtms.ensg), 
      sets = c("delta", "delta_res", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, eGene (ENSG)", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c("#0072B2", "#009E73", "#E69F00", "#F0E442"), 
      order.by = "freq")
```

### CpG level

```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

eqtms.cpg <- list(delta = delta.full.df$CpG_ID, delta_res = delta.res.full.df$CpG_ID, dex = dex.full.df$CpG_ID, veh = veh.full.df$CpG_ID)

upset(fromList(eqtms.cpg), 
      sets = c("delta", "delta_res", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, eCpG", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c("#0072B2", "#009E73", "#E69F00", "#F0E442"), 
      order.by = "freq")
```